<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Practical 6: Cell-Cell Communication â€“ Spatial Omics Data Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../assets/elixir_favicon.ico" rel="icon">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-bf0ef3c727e58b0ba1c4a46410c3c68e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Spatial Omics Data Analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../practicals.html"> 
<span class="menu-text">Practicals</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../presentations.html"> 
<span class="menu-text">Presentations</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/elixir-europe-training/ELIXIR-SCO-spatial-omics/"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Practical 6: Cell-Cell Communication</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../../../index.html" class="sidebar-logo-link">
      <img src="../../../assets/ELIXIR1.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#download-data" id="toc-download-data" class="nav-link active" data-scroll-target="#download-data">0. Download data</a></li>
  <li><a href="#introduction-to-liana" id="toc-introduction-to-liana" class="nav-link" data-scroll-target="#introduction-to-liana">0. Introduction to LIANA+</a></li>
  <li><a href="#cellphonedb-non-spatial-ccc" id="toc-cellphonedb-non-spatial-ccc" class="nav-link" data-scroll-target="#cellphonedb-non-spatial-ccc">1. CellPhoneDB: non-spatial CCC</a>
  <ul class="collapse">
  <li><a href="#spatially-variable-gene-selection" id="toc-spatially-variable-gene-selection" class="nav-link" data-scroll-target="#spatially-variable-gene-selection">Spatially-variable gene selection</a></li>
  </ul></li>
  <li><a href="#misty" id="toc-misty" class="nav-link" data-scroll-target="#misty">3. MISTY</a>
  <ul class="collapse">
  <li><a href="#estimate-pathway-activities" id="toc-estimate-pathway-activities" class="nav-link" data-scroll-target="#estimate-pathway-activities">3.1 Estimate pathway activities</a></li>
  <li><a href="#format-misty-object" id="toc-format-misty-object" class="nav-link" data-scroll-target="#format-misty-object">3.2. Format MISTy object</a></li>
  </ul></li>
  <li><a href="#learn-relationship-with-misty" id="toc-learn-relationship-with-misty" class="nav-link" data-scroll-target="#learn-relationship-with-misty">3.3 Learn relationship with MISTy</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Practical 6: Cell-Cell Communication</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Author: Francesca Drummer</p>
<p>In this notebook we will cover different methods to revocer cell-cell communication (CCC) in spatial transcriptomics.</p>
<ol type="1">
<li>non-spatial CCC testing with spatial DE genes or post-processing filter (e.i. spatial distance) using CellPhoneDB</li>
<li>MISTy</li>
</ol>
<p>To reduce the environment dependencies we will use the LIANA+ implementation of the methods. Please notice that the original tools might offer more functionalities. For that reason we will always link to the original publication and GitHub repository.</p>
<div id="cbb815f9-1dc6-48c4-ad86-c630a644a8f2" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> squidpy <span class="im">as</span> sq</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.sparse <span class="im">import</span> issparse, csr_matrix</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> liana.method <span class="im">import</span> singlecellsignalr, connectome, cellphonedb, natmi, logfc, cellchat, geometric_mean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="download-data" class="level2">
<h2 class="anchored" data-anchor-id="download-data">0. Download data</h2>
<div id="770895c8-4e32-42ab-947d-ea646e4c9c97" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>PATH <span class="op">=</span> <span class="st">"/home/icb/francesca.drummer/1-Projects/"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="517e527b-f9bc-46a5-bd35-4bd9e214d684" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load adata</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(Path(PATH, <span class="st">'xenium_mouse_ad_annotated_rotated.h5ad'</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="3">
<pre><code>AnnData object with n_obs Ã— n_vars = 350209 Ã— 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="770ba75f-9475-4cf0-a97e-f39a32f32fe1" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(adata.X[:<span class="dv">2</span>,:<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (1, 0)    1.0</code></pre>
</div>
</div>
<div id="7dd5fd19-95d3-4823-97ed-b2676d2ab21a" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>adata.layers[<span class="st">'counts'</span>] <span class="op">=</span> adata.X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a2d29a4b-1611-47d7-a748-d04364dde456" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalization to the median</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Freeman-Tukey square root transform</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> issparse(adata.X)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>sqrt_X <span class="op">=</span> adata.X.sqrt()</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new sparse matrix for X + 1</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>X_plus_1 <span class="op">=</span> adata.X <span class="op">+</span> csr_matrix(np.ones(adata.X.shape))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the square root of (X + 1)</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>sqrt_X_plus_1 <span class="op">=</span> X_plus_1.sqrt()</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>adata.layers[<span class="st">'median_ft'</span>] <span class="op">=</span> sqrt_X <span class="op">+</span> sqrt_X_plus_1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="introduction-to-liana" class="level2">
<h2 class="anchored" data-anchor-id="introduction-to-liana">0. Introduction to LIANA+</h2>
<p><a href="https://liana-py.readthedocs.io/en/latest/notebooks/basic_usage.html">LIANA+</a> is a toolbox in Python for various dissociated, multimodal and spatially informed cell-cell communication tools [Dimitrov et al., 2024].</p>
<p>First we install the package and observe which methods are implemented in LIANA+. Each method relies on different assumptions and returns a different ligand-receptor score. Usually, one score for the strength of the interaction (<code>magnitude</code>) and a score reflecting the <code>specifivity</code> of a interaction to a pair of cell identities.</p>
<div id="446adbc6-77b2-4001-a500-326888694246" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> liana <span class="im">as</span> li</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>li.mt.show_methods()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Method Name</th>
<th data-quarto-table-cell-role="th">Magnitude Score</th>
<th data-quarto-table-cell-role="th">Specificity Score</th>
<th data-quarto-table-cell-role="th">Reference</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CellPhoneDB</td>
<td>lr_means</td>
<td>cellphone_pvals</td>
<td>Efremova, M., Vento-Tormo, M., Teichmann, S.A....</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0</td>
<td>Connectome</td>
<td>expr_prod</td>
<td>scaled_weight</td>
<td>Raredon, M.S.B., Yang, J., Garritano, J., Wang...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>log2FC</td>
<td>None</td>
<td>lr_logfc</td>
<td>Dimitrov, D., TÃ¼rei, D., Garrido-Rodriguez, M....</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0</td>
<td>NATMI</td>
<td>expr_prod</td>
<td>spec_weight</td>
<td>Hou, R., Denisenko, E., Ong, H.T., Ramilowski,...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>SingleCellSignalR</td>
<td>lrscore</td>
<td>None</td>
<td>Cabello-Aguilar, S., Alame, M., Kon-Sun-Tack, ...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0</td>
<td>Rank_Aggregate</td>
<td>magnitude_rank</td>
<td>specificity_rank</td>
<td>Dimitrov, D., TÃ¼rei, D., Garrido-Rodriguez, M....</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Geometric Mean</td>
<td>lr_gmeans</td>
<td>gmean_pvals</td>
<td>CellPhoneDBv2's permutation approach applied t...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0</td>
<td>scSeqComm</td>
<td>inter_score</td>
<td>None</td>
<td>Baruzzo, G., Cesaro, G., Di Camillo, B. 2022. ...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>CellChat</td>
<td>lr_probs</td>
<td>cellchat_pvals</td>
<td>Jin, S., Guerrero-Juarez, C.F., Zhang, L., Cha...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Most CCC tools identify LR interaction. For this they rely on a extracting LR pairs from databases. There are diverse databases but LIANA+ has a consensus database that uses LR that are overlapping across databases.</p>
<p>First, we need to ensure that there are LR-pairs present in the data to be detected for communication.</p>
<div id="ce70328a-b6f3-462d-9476-d8908ddd0ca3" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(li.resource.show_resources())</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>resource_name <span class="op">=</span> <span class="st">"mouseconsensus"</span>  <span class="co"># Replace with the desired resource name if needed</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>lr_pairs <span class="op">=</span> li.resource.select_resource(resource_name)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>lr_pairs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['baccin2019', 'cellcall', 'cellchatdb', 'cellinker', 'cellphonedb', 'celltalkdb', 'connectomedb2020', 'consensus', 'embrace', 'guide2pharma', 'hpmr', 'icellnet', 'italk', 'kirouac2010', 'lrdb', 'mouseconsensus', 'ramilowski2015']</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="9">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">ligand</th>
<th data-quarto-table-cell-role="th">receptor</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">31371</td>
<td>Dll1</td>
<td>Notch1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">31372</td>
<td>Dll1</td>
<td>Notch2</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">31373</td>
<td>Dll1</td>
<td>Notch4</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">31374</td>
<td>Dll1</td>
<td>Notch3</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">31375</td>
<td>Nrg2</td>
<td>Erbb2_Erbb3</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">35355</td>
<td>Serpina1a</td>
<td>Lrp1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">35356</td>
<td>Serpina1b</td>
<td>Lrp1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">35357</td>
<td>Serpina1c</td>
<td>Lrp1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">35358</td>
<td>Serpina1d</td>
<td>Lrp1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">35359</td>
<td>Serpina1e</td>
<td>Lrp1</td>
</tr>
</tbody>
</table>

<p>3989 rows Ã— 2 columns</p>
</div>
</div>
</div>
<div id="74dec172-3e81-4871-bdff-0f0bc5de6cf4" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> lr_pairs_in_adata(adata):</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>    genes_in_dataset <span class="op">=</span> <span class="bu">set</span>(adata.var_names)  <span class="co"># Replace `adata.var_names` with your dataset's gene names if different</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Filter the ligand-receptor pairs for those present in your dataset</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    filtered_lr_pairs <span class="op">=</span> lr_pairs[</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>        lr_pairs[<span class="st">'ligand'</span>].isin(genes_in_dataset) <span class="op">&amp;</span> lr_pairs[<span class="st">'receptor'</span>].isin(genes_in_dataset)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Count the number of ligand-receptor pairs present in your dataset</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>    num_lr_pairs <span class="op">=</span> <span class="bu">len</span>(filtered_lr_pairs)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Display the count</span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Number of ligand-receptor pairs present in the dataset: </span><span class="sc">{</span>num_lr_pairs<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="7f5c2eec-7bc5-4558-8dcd-9c84c02c6f10" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>lr_pairs_in_adata(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of ligand-receptor pairs present in the dataset: 16</code></pre>
</div>
</div>
<p>In the following chapter, we will work with the CellPhoneDB method from LIANA+.</p>
</section>
<section id="cellphonedb-non-spatial-ccc" class="level2">
<h2 class="anchored" data-anchor-id="cellphonedb-non-spatial-ccc">1. CellPhoneDB: non-spatial CCC</h2>
<p>We will cover two possible appraoches to integrate spatial information into non-spatially aware CCC tools, as shown by <code>li.mt.show_methods()</code>.</p>
<ol type="1">
<li>Select spatially variable genes as input.</li>
<li>Input all genes into the</li>
</ol>
<section id="spatially-variable-gene-selection" class="level3">
<h3 class="anchored" data-anchor-id="spatially-variable-gene-selection">Spatially-variable gene selection</h3>
<p>We use Moranâ€™s I score as a measure of spatial autocorrelation to identify spatially variable genes.</p>
<p>For more information see: <a href="https://www.sc-best-practices.org/spatial/spatially_variable_genes.html">Chapter 29: Spatially variable genes</a> from single-cell best practices.</p>
<div id="e72f0b9c-a077-4ab9-b212-79241ce7275b" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sub_adata <span class="op">=</span> adata[(adata.obs[<span class="st">'time'</span>] <span class="op">==</span> <span class="st">'5_7'</span>) <span class="op">&amp;</span> (adata.obs[<span class="st">'condition'</span>] <span class="op">==</span> <span class="st">'TgCRND8'</span>)]</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>sub_adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>View of AnnData object with n_obs Ã— n_vars = 58681 Ã— 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial'
    layers: 'counts', 'median_ft'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="e4bf5e1b-745d-4612-8a22-e4d15fb0dc7c" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(sub_adata.X[:<span class="dv">5</span>,:<span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  (0, 2)    4.50828742980957
  (1, 0)    2.081632614135742
  (1, 2)    2.081632614135742
  (2, 2)    2.6322579383850098
  (4, 0)    1.2289156913757324
  (4, 2)    2.457831382751465</code></pre>
</div>
</div>
<div id="046a4e1f-30e4-4d0f-8b8f-2c771105e062" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sq.gr.spatial_neighbors(sub_adata, n_neighs<span class="op">=</span><span class="dv">30</span>, coord_type<span class="op">=</span><span class="st">"generic"</span>, key_added <span class="op">=</span> <span class="st">'neighs_based_spatial'</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>sq.gr.spatial_autocorr(sub_adata, connectivity_key <span class="op">=</span> <span class="st">"neighs_based_spatial_connectivities"</span>, mode<span class="op">=</span><span class="st">"moran"</span>, n_perms<span class="op">=</span><span class="dv">50</span>, genes<span class="op">=</span>sub_adata.var_names)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/squidpy/gr/_utils.py:194: ImplicitModificationWarning: Setting element `.obsp['neighs_based_spatial_connectivities']` of view, initializing view as actual.
100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 50/50 [00:48&lt;00:00,  1.03/s]</code></pre>
</div>
</div>
<div id="d6f3c60f-8a18-431e-9218-1f928867d625" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sub_adata.uns[<span class="st">"moranI"</span>].head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">I</th>
<th data-quarto-table-cell-role="th">pval_norm</th>
<th data-quarto-table-cell-role="th">var_norm</th>
<th data-quarto-table-cell-role="th">pval_z_sim</th>
<th data-quarto-table-cell-role="th">pval_sim</th>
<th data-quarto-table-cell-role="th">var_sim</th>
<th data-quarto-table-cell-role="th">pval_norm_fdr_bh</th>
<th data-quarto-table-cell-role="th">pval_z_sim_fdr_bh</th>
<th data-quarto-table-cell-role="th">pval_sim_fdr_bh</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Nwd2</td>
<td>0.637334</td>
<td>0.0</td>
<td>0.000001</td>
<td>0.0</td>
<td>0.019608</td>
<td>0.000014</td>
<td>0.0</td>
<td>0.0</td>
<td>0.019665</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Ppp1r1b</td>
<td>0.602092</td>
<td>0.0</td>
<td>0.000001</td>
<td>0.0</td>
<td>0.019608</td>
<td>0.000014</td>
<td>0.0</td>
<td>0.0</td>
<td>0.019665</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Mbp</td>
<td>0.592759</td>
<td>0.0</td>
<td>0.000001</td>
<td>0.0</td>
<td>0.019608</td>
<td>0.000012</td>
<td>0.0</td>
<td>0.0</td>
<td>0.019665</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Slc17a7</td>
<td>0.575611</td>
<td>0.0</td>
<td>0.000001</td>
<td>0.0</td>
<td>0.019608</td>
<td>0.000012</td>
<td>0.0</td>
<td>0.0</td>
<td>0.019665</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Neurod6</td>
<td>0.573918</td>
<td>0.0</td>
<td>0.000001</td>
<td>0.0</td>
<td>0.019608</td>
<td>0.000012</td>
<td>0.0</td>
<td>0.0</td>
<td>0.019665</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>Subset data to only spatially variable genes and check that there are still LR pairs present.</p>
<div id="8d005bbf-0ee0-428b-ae5c-b9d54ddac290" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sub_adata_svg <span class="op">=</span> sub_adata[:, sub_adata.uns[<span class="st">"moranI"</span>][<span class="st">'I'</span>] <span class="op">&gt;</span> <span class="fl">0.2</span>]</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sub_adata_svg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<pre><code>View of AnnData object with n_obs Ã— n_vars = 58681 Ã— 115
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap', 'neighs_based_spatial_neighbors', 'moranI'
    obsm: 'X_pca', 'X_umap', 'spatial'
    layers: 'counts', 'median_ft'
    obsp: 'connectivities', 'distances', 'neighs_based_spatial_connectivities', 'neighs_based_spatial_distances'</code></pre>
</div>
</div>
<div id="93238851-6685-44ef-bb68-965287bbcb97" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>lr_pairs_in_adata(sub_adata_svg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of ligand-receptor pairs present in the dataset: 5</code></pre>
</div>
</div>
<div id="5695f21f-2b98-4ffd-99da-46b34c625cfd" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(sub_adata, </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                      color<span class="op">=</span>[<span class="st">"Ppp1r1b"</span>, <span class="st">"Neurod6"</span>],</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>                     shape<span class="op">=</span><span class="va">None</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: Please specify a valid `library_id` or set it permanently in `adata.uns['spatial']`</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-18-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="cellphonedb" class="level4">
<h4 class="anchored" data-anchor-id="cellphonedb">CellPhoneDB</h4>
<div id="5005588c-aaec-4f5e-9348-a33e7d9404f9" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>cellphonedb(sub_adata_svg,</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>            groupby<span class="op">=</span><span class="st">'cell_types'</span>,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>            <span class="co"># </span><span class="al">NOTE</span><span class="co"> by default the resource uses HUMAN gene symbols</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>            resource_name<span class="op">=</span><span class="st">'mouseconsensus'</span>,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>            expr_prop<span class="op">=</span><span class="fl">0.1</span>,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>            verbose<span class="op">=</span><span class="va">True</span>, </span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>            use_raw <span class="op">=</span> <span class="va">False</span>,</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>            layer <span class="op">=</span> <span class="st">'counts'</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>            key_added<span class="op">=</span><span class="st">'cpdb_res'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Using resource `mouseconsensus`.
Using the `counts` layer!
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
39 samples of mat are empty, they will be removed.
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/liana/method/_pipe_utils/_pre.py:153: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
0.98 of entities in the resource are missing from the data.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Generating ligand-receptor stats for 58681 samples and 6 features</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 1000/1000 [00:04&lt;00:00, 207.76it/s]
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/liana/method/sc/_Method.py:266: ImplicitModificationWarning: Trying to modify attribute `._uns` of view, initializing view as actual.</code></pre>
</div>
</div>
<div id="319d455b-dbcf-4e3d-bc9f-59d695ac720d" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>sub_adata_svg.uns[<span class="st">'cpdb_res'</span>].head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">ligand</th>
<th data-quarto-table-cell-role="th">ligand_complex</th>
<th data-quarto-table-cell-role="th">ligand_means</th>
<th data-quarto-table-cell-role="th">ligand_props</th>
<th data-quarto-table-cell-role="th">receptor</th>
<th data-quarto-table-cell-role="th">receptor_complex</th>
<th data-quarto-table-cell-role="th">receptor_means</th>
<th data-quarto-table-cell-role="th">receptor_props</th>
<th data-quarto-table-cell-role="th">source</th>
<th data-quarto-table-cell-role="th">target</th>
<th data-quarto-table-cell-role="th">lr_means</th>
<th data-quarto-table-cell-role="th">cellphone_pvals</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>C1qa</td>
<td>C1qa</td>
<td>3.718225</td>
<td>0.903226</td>
<td>Cd93</td>
<td>Cd93</td>
<td>0.207941</td>
<td>0.12677</td>
<td>Microglia</td>
<td>Astrocytes</td>
<td>1.963083</td>
<td>0.000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">0</td>
<td>C1qa</td>
<td>C1qa</td>
<td>0.311218</td>
<td>0.202483</td>
<td>Cd93</td>
<td>Cd93</td>
<td>0.207941</td>
<td>0.12677</td>
<td>Astrocytes</td>
<td>Astrocytes</td>
<td>0.259579</td>
<td>0.000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>C1qa</td>
<td>C1qa</td>
<td>0.194758</td>
<td>0.112398</td>
<td>Cd93</td>
<td>Cd93</td>
<td>0.207941</td>
<td>0.12677</td>
<td>Oligodendrocytes</td>
<td>Astrocytes</td>
<td>0.201350</td>
<td>0.645</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>C1qa</td>
<td>C1qa</td>
<td>0.183544</td>
<td>0.124688</td>
<td>Cd93</td>
<td>Cd93</td>
<td>0.207941</td>
<td>0.12677</td>
<td>OPC</td>
<td>Astrocytes</td>
<td>0.195742</td>
<td>0.657</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>C1qa</td>
<td>C1qa</td>
<td>0.139770</td>
<td>0.135632</td>
<td>Cd93</td>
<td>Cd93</td>
<td>0.207941</td>
<td>0.12677</td>
<td>Inhibitory neurons</td>
<td>Astrocytes</td>
<td>0.173855</td>
<td>1.000</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="09ddf91a-8f47-4147-81c2-92f1add59257" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>my_plot <span class="op">=</span> li.pl.tileplot(adata <span class="op">=</span> sub_adata_svg,</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>                         fill<span class="op">=</span><span class="st">'means'</span>,</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                         label<span class="op">=</span><span class="st">'props'</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                         label_fun<span class="op">=</span><span class="kw">lambda</span> x: <span class="ss">f'</span><span class="sc">{</span>x<span class="sc">:.2f}</span><span class="ss">'</span>,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>                         top_n<span class="op">=</span><span class="dv">20</span>,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>                         orderby<span class="op">=</span><span class="st">'cellphone_pvals'</span>,</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>                         orderby_ascending<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>                         source_labels<span class="op">=</span>[<span class="st">'Astrocytes'</span>, <span class="st">'Excitatory neurons'</span>, <span class="st">'Inhibitory neurons'</span>, <span class="st">'Microglia'</span>, <span class="st">'OPC'</span>, <span class="st">'Oligodendrocytes'</span>],</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>                         target_labels<span class="op">=</span>[<span class="st">'Astrocytes'</span>],</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>                         uns_key<span class="op">=</span><span class="st">'cpdb_res'</span>, <span class="co"># </span><span class="al">NOTE</span><span class="co">: default is 'liana_res'</span></span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>                         source_title<span class="op">=</span><span class="st">'Ligand'</span>,</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>                         target_title<span class="op">=</span><span class="st">'Receptor'</span>,</span>
<span id="cb33-13"><a href="#cb33-13" aria-hidden="true" tabindex="-1"></a>                         figure_size<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">7</span>)</span>
<span id="cb33-14"><a href="#cb33-14" aria-hidden="true" tabindex="-1"></a>                         )</span>
<span id="cb33-15"><a href="#cb33-15" aria-hidden="true" tabindex="-1"></a>my_plot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-21-output-2.png" width="800" height="700" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>TODO: Plot genes in space to see whether they are visually close.</p>
<p><span style="color:red; font-weight:bold">Change the <code>expr_prop</code> in the CellPhoneDB function and try out some other tools like CellChat. How does it effect the results?.</span></p>
<p><span style="color:red; font-weight:bold">Compare the results for the healthy control or different time points. Do the CCC across cell types change?.</span></p>
<p><span style="color:red; font-weight:bold">What could be an alternative post-hoc analysis instead of using spatially variable gene detection to have a spatially informed CCC?</span></p>
</section>
</section>
</section>
<section id="misty" class="level2">
<h2 class="anchored" data-anchor-id="misty">3. MISTY</h2>
<p>MISTy is a framework that helps understand how different features, such as genes or cell types interact with each other in space. For this MISTy uses so called <em>views</em>, each describing a different spatial context.</p>
<p><img src="./figures/MISTy.png" alt="Alt Text" width="500"></p>
<div id="3d4a55b3-ccf7-44d1-bc77-4749b2ec91d4" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> decoupler <span class="im">as</span> dc</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> plotnine <span class="im">as</span> p9</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> liana <span class="im">as</span> li</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Import Helper functions needed to create MISTy objects</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> liana.method <span class="im">import</span> MistyData, genericMistyData, lrMistyData</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="co">#Import predefined single view models</span></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> liana.method.sp <span class="im">import</span> RandomForestModel, LinearModel, RobustLinearModel</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="estimate-pathway-activities" class="level3">
<h3 class="anchored" data-anchor-id="estimate-pathway-activities">3.1 Estimate pathway activities</h3>
<p>Before we run MISTy, letâ€™s estimate pathway activities as a way to make the data a bit more interpretable. We will use <a href="https://academic.oup.com/bioinformaticsadvances/article/2/1/vbac016/6544613">decoupler-py</a> with pathways genesets from <a href="https://www.nature.com/articles/s41467-017-02391-6">PROGENy</a>. See <a href="https://decoupler-py.readthedocs.io/en/latest/notebooks/spatial.html">this</a> tutorial for details.</p>
<div id="34b91419-f21b-4e7f-b1ed-85423d1bee40" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>progeny <span class="op">=</span> dc.get_progeny(organism<span class="op">=</span><span class="st">'mouse'</span>, top<span class="op">=</span><span class="dv">500</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading annotations for all proteins from the following resources: `['PROGENy']`</code></pre>
</div>
</div>
<div id="bc82bd52-7966-436b-b7b3-cb4312ab7970" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>dc.run_mlm(</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    mat<span class="op">=</span>adata,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    net<span class="op">=</span>progeny,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    source<span class="op">=</span><span class="st">'source'</span>,</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    target<span class="op">=</span><span class="st">'target'</span>,</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    weight<span class="op">=</span><span class="st">'weight'</span>,</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    verbose<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    use_raw<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running mlm on mat with 350209 samples and 347 targets for 11 sources.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 36/36 [00:04&lt;00:00,  7.39it/s]</code></pre>
</div>
</div>
<div id="b6dae42c-938a-47fd-95bc-d1270f2db274" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract progeny activities as an AnnData object</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>acts_progeny <span class="op">=</span> li.ut.obsm_to_adata(adata, <span class="st">'mlm_estimate'</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>acts_progeny</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="29">
<pre><code>AnnData object with n_obs Ã— n_vars = 350209 Ã— 11
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial', 'mlm_estimate', 'mlm_pvals'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="fe5a1dcf-1dd7-4b2a-8994-c868589a3b64" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>acts_progeny.var_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="30">
<pre><code>Index(['Androgen', 'Estrogen', 'JAK-STAT', 'MAPK', 'NFkB', 'PI3K', 'TGFb',
       'TNFa', 'VEGF', 'WNT', 'p53'],
      dtype='object')</code></pre>
</div>
</div>
<div id="307db84c-7620-48d9-be6a-9ec6c532e3bc" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check how the pathway activities look like</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> library_id <span class="kw">in</span> acts_progeny.obs[<span class="st">"batch_key"</span>].unique():</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    adata_subset <span class="op">=</span> acts_progeny[acts_progeny.obs[<span class="st">"batch_key"</span>] <span class="op">==</span> library_id]</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Condition: </span><span class="sc">{</span>np<span class="sc">.</span>unique(adata_subset.obs[<span class="st">"condition"</span>])[<span class="dv">0</span>]<span class="sc">}</span><span class="ss"> and time: </span><span class="sc">{</span>np<span class="sc">.</span>unique(adata_subset.obs[<span class="st">"time"</span>])[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>    sc.pl.spatial(</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>        adata_subset, </span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>        color<span class="op">=</span>[<span class="st">'Androgen'</span>, <span class="st">'Estrogen'</span>, <span class="st">'TGFb'</span>, <span class="st">'TNFa'</span>, <span class="st">'VEGF'</span>, <span class="st">'WNT'</span>, <span class="st">'p53'</span>], </span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>        cmap<span class="op">=</span><span class="st">'RdBu_r'</span>, </span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>        spot_size<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Condition: TgCRND8 and time: 2_5</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-27-output-2.png" width="2911" height="960" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Condition: TgCRND8 and time: 5_7</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-27-output-4.png" width="2911" height="960" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Condition: TgCRND8 and time: 17_9</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-27-output-6.png" width="2911" height="960" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Condition: wildtype and time: 2_5</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-27-output-8.png" width="2911" height="960" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Condition: wildtype and time: 5_7</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-27-output-10.png" width="2911" height="960" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Condition: wildtype and time: 13_4</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-27-output-12.png" width="2911" height="960" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="format-misty-object" class="level3">
<h3 class="anchored" data-anchor-id="format-misty-object">3.2. Format MISTy object</h3>
<p>MISTy objects are in the <a href="https://github.com/scverse/mudata">MuData</a> (Bredikhin et al., 2021) object with one modality per view.</p>
<p>The <em>intra</em> view is the target variable</p>
<div id="fddf4239-ce97-49c9-ab96-c0652d62ff1e" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="32">
<pre><code>AnnData object with n_obs Ã— n_vars = 350209 Ã— 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial', 'mlm_estimate', 'mlm_pvals'
    layers: 'counts', 'median_ft'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="f6c68a56-4990-4b44-a6aa-124cb3dab35d" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>cell_assignments <span class="op">=</span> adata.obs[<span class="st">'cell_types'</span>].astype(<span class="bu">str</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5bd4d803-3712-48b2-8fbb-ec760297a360" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>np.unique(cell_assignments)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="34">
<pre><code>array(['Astrocytes', 'Excitatory neurons', 'Inhibitory neurons',
       'Microglia', 'OPC', 'Oligodendrocytes', 'nan'], dtype=object)</code></pre>
</div>
</div>
<div id="52cc670d-b0d1-46d8-832b-a179587809d1" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>one_hot_data <span class="op">=</span> pd.get_dummies(cell_assignments)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="65eb4b4c-ebdf-4822-8e8c-351434ddf088" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 3: Create AnnData object</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>adata_ct <span class="op">=</span> ad.AnnData(</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>    X<span class="op">=</span>one_hot_data.values,  <span class="co"># One-hot encoding matrix</span></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>    obs<span class="op">=</span>pd.DataFrame(index<span class="op">=</span>adata.obs_names),  <span class="co"># Cells as `.obs`</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>    var<span class="op">=</span>pd.DataFrame(index<span class="op">=</span>np.unique(cell_assignments)),  <span class="co"># Cell types as `.var`</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>adata_ct.obsm[<span class="st">'spatial'</span>] <span class="op">=</span> adata.obsm[<span class="st">'spatial'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f582033b-817c-446a-b224-c744dedc4155" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check key cell types</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="co"># sc.pl.spatial(adata_ct,</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="co">#               color=['OPC'],</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="co">#               size=1.3, ncols=2, alpha_img=0,</span></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="co">#               spot_size = 10</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="co">#               )</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9c31ae7f-ef62-419d-956f-e30c87d7bfda" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>misty <span class="op">=</span> genericMistyData(intra<span class="op">=</span>adata_ct, extra<span class="op">=</span>acts_progeny, cutoff<span class="op">=</span><span class="fl">0.05</span>, bandwidth<span class="op">=</span><span class="dv">200</span>, n_neighs<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>misty</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/mudata/_core/mudata.py:491: UserWarning: Cannot join columns with the same name because var_names are intersecting.
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/liana/method/sp/_misty/_Misty.py:80: ImplicitModificationWarning: Setting element `.layers['weighted']` of view, initializing view as actual.
/home/icb/francesca.drummer/miniconda3/envs/ccc_liana/lib/python3.9/site-packages/liana/method/sp/_misty/_Misty.py:80: ImplicitModificationWarning: Setting element `.layers['weighted']` of view, initializing view as actual.</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre>MuData object with n_obs Ã— n_vars = 350209 Ã— 27
  3 modalities
    intra:  350209 x 5
      obsm: 'spatial'
    juxta:  350209 x 11
      obsm: 'spatial'
      layers:   'weighted'
      obsp: 'spatial_connectivities'
    para:   350209 x 11
      obsm: 'spatial'
      layers:   'weighted'
      obsp: 'spatial_connectivities'</pre>
</div>
</div>
</section>
</section>
<section id="learn-relationship-with-misty" class="level2">
<h2 class="anchored" data-anchor-id="learn-relationship-with-misty">3.3 Learn relationship with MISTy</h2>
<p>Now that we have constructed the object, we can learn the relationships across views.</p>
<p>Reationships can be learned by different models (e.i. RandomForrest, LinearModel). The fastest is the Linear model which we will fit here for each target in the intra-view, using the juxta and para views as predictors.</p>
<div id="2c075e48-2918-4b39-8a92-aadf63711de2" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>misty(model<span class="op">=</span>LinearModel, k_cv<span class="op">=</span><span class="dv">10</span>, seed<span class="op">=</span><span class="dv">1337</span>, bypass_intra<span class="op">=</span><span class="va">True</span>, verbose <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Now learning: nan: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 5/5 [00:44&lt;00:00,  8.89s/it]               </code></pre>
</div>
</div>
<p>By default the results are saved in the <code>misty</code> object because <code>inplace = True</code>.</p>
<p>The <code>misty</code> object does now contain two DataFrames:</p>
<ul>
<li><code>target_metrics</code></li>
<li><code>interactions</code></li>
</ul>
<div id="9c4e6e14-4b60-4bf8-bd32-f28918ab7e2a" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>li.pl.target_metrics(misty, stat<span class="op">=</span><span class="st">'gain_R2'</span>, return_fig<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-36-output-1.png" width="500" height="500" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="435717ac-e170-4d91-aec0-8c11c7cee212" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>li.pl.contributions(misty, return_fig<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-37-output-2.png" width="500" height="500" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="24f2006a-ab16-49d3-8972-79fd4b5c082b" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>(</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>    li.pl.interactions(misty, view<span class="op">=</span><span class="st">'juxta'</span>, return_fig<span class="op">=</span><span class="va">True</span>, figure_size<span class="op">=</span>(<span class="dv">7</span>,<span class="dv">5</span>)) <span class="op">+</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>    p9.scale_fill_gradient2(low <span class="op">=</span> <span class="st">"blue"</span>, mid <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"red"</span>, midpoint <span class="op">=</span> <span class="dv">0</span>)</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p6_ccc_files/figure-html/cell-38-output-2.png" width="700" height="500" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><span style="color:red; font-weight:bold">Fit a <code>RandomForestModel</code> instead. How does it effect the results and their interpretability?.</span></p>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>[1] Dimitrov D., SchÃ¤fer P.S.L, Farr E., Rodriguez Mier P., Lobentanzer S., Badia-i-Mompel P., Dugourd A., Tanevski J., Ramirez Flores R.O. and Saez-Rodriguez J. LIANA+ provides an all-in-one framework for cellâ€“cell communication inference. Nat Cell Biol (2024). https://doi.org/10.1038/s41556-024-01469-w</p>
<p>[2] Li, Z., Wang, T., Liu, P. &amp; Huang, Y. SpatialDM for rapid identification of spatially co-expressed ligandâ€“receptor and revealing cellâ€“cell communication patterns. Nat Commun 14, 3995 (2023).</p>
<p>[3] Bredikhin, D., Kats, I. &amp; Stegle, O. MUON: multimodal omics analysis framework. Genome Biol 23, 42 (2022). https://doi.org/10.1186/s13059-021-02577-8</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/elixir-europe-training\.github\.io\/ELIXIR-SCO-spatial-omics\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>