<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Computational analysis of spatial transcriptomics data – ELIXIR-SCO-spatial-transcriptomics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../assets/elixir_favicon.ico" rel="icon">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-c7e0d8a4bb88518e69f3d36b54b6fc10.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">ELIXIR-SCO-spatial-transcriptomics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../practicals.html"> 
<span class="menu-text">Practicals</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../presentations.html"> 
<span class="menu-text">Presentations</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/elixir-europe-training/ELIXIR-SCO-spatial-omics/"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Computational analysis of spatial transcriptomics data</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../../../index.html" class="sidebar-logo-link">
      <img src="../../../assets/ELIXIR1.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#brainomics-2.0---day-3" id="toc-brainomics-2.0---day-3" class="nav-link active" data-scroll-target="#brainomics-2.0---day-3">BrainOmics 2.0 - Day 3</a>
  <ul class="collapse">
  <li><a href="#agenda" id="toc-agenda" class="nav-link" data-scroll-target="#agenda">Agenda:</a></li>
  </ul></li>
  <li><a href="#dataset-merfish-spatial-transcriptomics-dataset-of-a-single-adult-mouse-brain-zhuang-abca-1-xiaowei-zhuang" id="toc-dataset-merfish-spatial-transcriptomics-dataset-of-a-single-adult-mouse-brain-zhuang-abca-1-xiaowei-zhuang" class="nav-link" data-scroll-target="#dataset-merfish-spatial-transcriptomics-dataset-of-a-single-adult-mouse-brain-zhuang-abca-1-xiaowei-zhuang">Dataset: MERFISH spatial transcriptomics dataset of a single adult mouse brain (Zhuang-ABCA-1) (Xiaowei Zhuang)</a>
  <ul class="collapse">
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">Quality Control</a></li>
  <li><a href="#plot-gene-expression-in-space" id="toc-plot-gene-expression-in-space" class="nav-link" data-scroll-target="#plot-gene-expression-in-space">Plot gene expression in space</a></li>
  </ul></li>
  <li><a href="#prediction-of-spatial-patterns-for-unmeasured-genes-using-scrna-seq-data-with-spage" id="toc-prediction-of-spatial-patterns-for-unmeasured-genes-using-scrna-seq-data-with-spage" class="nav-link" data-scroll-target="#prediction-of-spatial-patterns-for-unmeasured-genes-using-scrna-seq-data-with-spage">Prediction of spatial patterns for unmeasured genes using scRNA-seq data with SpaGE</a>
  <ul class="collapse">
  <li><a href="#leave-one-gene-out-cross-validation" id="toc-leave-one-gene-out-cross-validation" class="nav-link" data-scroll-target="#leave-one-gene-out-cross-validation">Leave-one-gene-out cross validation</a></li>
  <li><a href="#predict-unmeasured-genes" id="toc-predict-unmeasured-genes" class="nav-link" data-scroll-target="#predict-unmeasured-genes">Predict unmeasured genes</a></li>
  </ul></li>
  <li><a href="#unmeasured-gene-imputation-with-tangram" id="toc-unmeasured-gene-imputation-with-tangram" class="nav-link" data-scroll-target="#unmeasured-gene-imputation-with-tangram">Unmeasured gene imputation with Tangram</a></li>
  <li><a href="#cell-type-deconvolution-with-cell-2-location" id="toc-cell-type-deconvolution-with-cell-2-location" class="nav-link" data-scroll-target="#cell-type-deconvolution-with-cell-2-location">Cell type deconvolution with cell 2 location</a>
  <ul class="collapse">
  <li><a href="#additional-access-image-in-the-anndata-object" id="toc-additional-access-image-in-the-anndata-object" class="nav-link" data-scroll-target="#additional-access-image-in-the-anndata-object">Additional: access image in the AnnData object</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Computational analysis of spatial transcriptomics data</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="brainomics-2.0---day-3" class="level2">
<h2 class="anchored" data-anchor-id="brainomics-2.0---day-3">BrainOmics 2.0 - Day 3</h2>
<section id="th-november-2024" class="level4">
<h4 class="anchored" data-anchor-id="th-november-2024">20th November 2024</h4>
</section>
<section id="ahmed-mahfouz-benedetta-manzato-lumc" class="level4">
<h4 class="anchored" data-anchor-id="ahmed-mahfouz-benedetta-manzato-lumc">Ahmed Mahfouz, Benedetta Manzato (LUMC)</h4>
</section>
<section id="notebook-with-solutions" class="level4">
<h4 class="anchored" data-anchor-id="notebook-with-solutions"><span style="color: green"> Notebook with solutions</span></h4>
</section>
<section id="agenda" class="level3">
<h3 class="anchored" data-anchor-id="agenda">Agenda:</h3>
<p><font size="3">1. Spatial data download and format</font></p><font size="3">
<p><font size="3">2. Spatial imputation with SpaGE and Tangram</font></p><font size="3">
<p><font size="3">3. Cell deconvolution with Cell2location</font></p><font size="3">
<p><font size="3">4. Access image in AnnData object</font></p><font size="3">
</font></font></font></font></section><font size="3"><font size="3"><font size="3">
</font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="dataset-merfish-spatial-transcriptomics-dataset-of-a-single-adult-mouse-brain-zhuang-abca-1-xiaowei-zhuang" class="level1">
<h1>Dataset: MERFISH spatial transcriptomics dataset of a single adult mouse brain (Zhuang-ABCA-1) (Xiaowei Zhuang)</h1>
<p><font size="3"> This spatially resolved cell atlas of the whole mouse brain provides a systematic characterization of the spatial organization of transcriptomically defined cell types across the entire adult mouses brain using in situ, single-cell transcriptomic profiling with multiplexed error-robust fluorescence in situ hybridization (MERFISH). A gene panel of 1,122 genes was imaged in 9 million brain cells using <strong>MERFISH</strong>, and spatially resolved, single-cell expression profiling was performed at the whole-transcriptome scale by integrating the MERFISH data with the whole-brain single-cell RNA-sequencing (scRNA-seq) dataset previously generated by the Allen Institute. This resulted in the generation of a comprehensive cell atlas of more than 5,000 transcriptionally distinct cell clusters, belonging to ∼300 major cell types (subclasses), across the whole mouse brain, with high molecular and spatial resolution. The authors performed MERFISH imaging on <strong>245 coronal and sagittal sections</strong>, the cell atlas was then spatially registered to the Allen Mouse Brain Common Coordinate Framework (CCFv3), which allows systematic quantifications of the cell composition and organization in individual brain regions as anatomically delineated in the CCFv3. This cell atlas also reveals molecularly-defined brain regions characterized by distinct cell type compositions, spatial gradients featuring gradual changes in the gene-expression profiles of cells, as well as cell-type-specific cell-cell interactions and the molecular basis and functional implications of these cell-cell interactions.</font></p><font size="3">
<p><font size="3"> The collection spans four mouse specimens (two coronal sets and two sagittal sets): - <font size="3"><strong>Zhuang-ABCA-1</strong>: - <strong>Dataset size</strong>: 4.2 million cells spanning 147 coronal sections with a 1,122 gene panel. - <strong>Filtered dataset</strong>: 2.8 million cells passed the cell classification confidence score threshold and are displayed in the ABC atlas.</font></font></p><font size="3"><font size="3">
<ul>
<li><font size="3"><strong>Zhuang-ABCA-2</strong>:
<ul>
<li><strong>Dataset size</strong>: 1.9 million cells spanning 66 coronal sections.</li>
<li><strong>Filtered dataset</strong>: 1.2 million cells passed the cell classification confidence score threshold.</li>
</ul></font></li><font size="3">
<li><font size="3"><strong>Zhuang-ABCA-3</strong>:
<ul>
<li><strong>Dataset size</strong>: 2.1 million cells spanning 23 sagittal sections.</li>
<li><strong>Filtered dataset</strong>: 1.6 million cells passed the confidence score threshold.</li>
</ul></font></li><font size="3">
<li><font size="3"><strong>Zhuang-ABCA-4</strong>:
<ul>
<li><strong>Dataset size</strong>: 0.22 million cells spanning 3 sagittal sections.</li>
<li><strong>Filtered dataset</strong>: 0.16 million cells passed the confidence score threshold.</li>
</ul></font></li><font size="3">
</font></font></font></ul><font size="3"><font size="3"><font size="3">
<p><font size="3"> We will work with the Brain Coronal 1 collection (Zhuang-ABCA-1); the cell-by-gene matrix of the 8.4 millions cells can be downloaded from the AWS bucket of this animal. The cells are then filtered by cell-classification (label transfer) confidence scores calculated during MERFISH-scRNAseq data integration. 7.0 million cells passed the confidence score threshold for cell subclass label transfer and 5.8 million cells further passed the confidence score threshold for cell cluster label transfer. These 5.8 million cells are included in the cell metadata file that can be downloaded from the the AWS bucket and are displayed on the ABC Atlas. The CCF coordinates of the 5.4 million cells that were registered to the 3D Allen-CCF can be downloaded from the CCF coordinate files in the AWB bucket. The collection spans four mouse specimens (2 coronal sets and 2 sagittal sets). Cells are mapped to the whole mouse brain taxonomy (WMB-taxonomy) and Allen Common Coordinate Framework (Allen-CCF-2020). Refer to Zhang et al, 2023 for more details.</font></p><font size="3">
<p>For more details: <a href="https://www.nature.com/articles/s41586-023-06808-9">Molecularly defined and spatially resolved cell atlas of the whole mouse brain (Zhang et al.&nbsp;2023)</a></p>
<p>From https://github.com/AllenInstitute/abc_atlas_access/blob/main/notebooks/zhuang_merfish_tutorial.ipynb</p>
</font><p><font size="3">Data Availablity : https://knowledge.brain-map.org/data/5C0201JSVE04WY6DMVC/collections </font></p>
<div id="13a8305d-204c-4780-8c01-738bbe0eeabd" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Record the start time</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>start_time <span class="op">=</span> time.time()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c80f5d54-37d4-4375-96c9-48f29a1585a9" class="cell" data-tags="[]" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">"ignore"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing necessary libraries for spatial data analysis and visualization</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os, re</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy.external <span class="im">as</span> sce</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the version of important packages for reproducibility</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>sc.logging.print_header()</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the default aesthetics for Scanpy plots</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>sc.set_figure_params(facecolor<span class="op">=</span><span class="st">'white'</span>, figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">8</span>))  <span class="co"># White background, figure size 8x8</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">0</span>  <span class="co"># Set verbosity to show only errors (0) # errors (0), warnings (1), info (2), hints (3)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>scanpy==1.10.2 anndata==0.10.8 umap==0.5.6 numpy==1.26.4 scipy==1.14.1 pandas==2.2.2 scikit-learn==1.5.1 statsmodels==0.14.2 igraph==0.11.6 pynndescent==0.5.13</code></pre>
</div>
</div>
<div id="f01cbf37-960e-46b1-b0e4-e05d32667014" class="cell" data-tags="[]" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing functions to handle file paths and download data from ABC Atlas</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path  <span class="co"># Object-oriented file paths</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> abc_atlas_access.abc_atlas_cache.abc_project_cache <span class="im">import</span> AbcProjectCache  <span class="co"># Cache system for ABC Atlas data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3"> We create an output folder in your local space to save outputs:</font></p><font size="3">
<div id="860e1d4a-b95e-4d5f-84e3-ffee578f8f87" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>folder_path <span class="op">=</span> <span class="st">'./outputs/'</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(folder_path):</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>        os.makedirs(folder_path)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Folder created: </span><span class="sc">{</span>folder_path<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Folder already exists: </span><span class="sc">{</span>folder_path<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Folder already exists: ./outputs/</code></pre>
</div>
</div>
<div id="c6f83355-5d90-48d8-b358-e1b7ffc284d3" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co">#spatial_wd = '/group/brainomics/InputData'</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>spatial_wd <span class="op">=</span> <span class="st">'/exports/humgen/bmanzato/spatial_workshop/'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3">We will use <a href="https://github.com/AllenInstitute/abc_atlas_access/tree/41eb836e41e516ee528c98bf3979d0aff60c0b85">AbcProjectCache</a> to download the data required for the tutorials.</font></p>
<p><font size="3">Data associated with the Allen Brain Cell Atlas is hosted on Amazon Web Services (AWS) in an S3 bucket as a AWS Public Dataset. No account or login is required. The S3 bucket is located here arn:aws:s3:::allen-brain-cell-atlas. You will need to be connected to the internet to run this notebook.</font></p><font size="3">
<p><font size="3">Each data release has an associated manifest.json which lists all the specific version of directories and files that are part of the release.</font></p>
<div id="c2979380-6faf-4c10-81b7-843e208ea699" class="cell" data-tags="[]" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the base path where data will be downloaded</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co">#download_base = Path('./outputs/abc_download_root')  # Path to save downloaded data</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>download_base <span class="op">=</span> Path(<span class="st">'/exports/humgen/bmanzato/abc_download_root'</span>)  <span class="co"># Path to save downloaded data</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize the ABC Project Cache to manage downloaded data from the ABC Atlas</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>abc_cache <span class="op">=</span> AbcProjectCache.from_s3_cache(download_base)  <span class="co"># Fetch data from S3 cache</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># abc_cache.current_manifest show the current cache manifest</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>abc_cache.current_manifest</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>'releases/20241130/manifest.json'</code></pre>
</div>
</div>
<div id="9d8506bf-adc6-4c38-ab16-c0d5b35d81fc" class="cell" data-tags="[]" data-execution_count="7">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>abc_cache.load_latest_manifest()</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># List available directories</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>abc_cache.list_directories</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>['Allen-CCF-2020',
 'MERFISH-C57BL6J-638850',
 'MERFISH-C57BL6J-638850-CCF',
 'MERFISH-C57BL6J-638850-imputed',
 'MERFISH-C57BL6J-638850-sections',
 'WHB-10Xv3',
 'WHB-taxonomy',
 'WMB-10X',
 'WMB-10XMulti',
 'WMB-10Xv2',
 'WMB-10Xv3',
 'WMB-neighborhoods',
 'WMB-taxonomy',
 'Zeng-Aging-Mouse-10Xv3',
 'Zeng-Aging-Mouse-WMB-taxonomy',
 'Zhuang-ABCA-1',
 'Zhuang-ABCA-1-CCF',
 'Zhuang-ABCA-2',
 'Zhuang-ABCA-2-CCF',
 'Zhuang-ABCA-3',
 'Zhuang-ABCA-3-CCF',
 'Zhuang-ABCA-4',
 'Zhuang-ABCA-4-CCF']</code></pre>
</div>
</div>
<p><font size="3">We will use the <strong>Zhuang-ABCA-1 dataset</strong>: 4.2 million cell spatial transcriptomics dataset spanning 147 coronal sections with a 1122 gene panel. 2.8 million cells passed cell classification confidence score threshold and displayed in the ABC atlas</font></p><font size="3">
<p><font size="3">List data files available in the Zhuang-ABCA-1 directory:</font></p><font size="3">
<div id="698f03e8-50c5-4854-8603-86888aac683a" class="cell" data-tags="[]" data-execution_count="8">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>abc_cache.list_data_files(<span class="st">'Zhuang-ABCA-1'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>['Zhuang-ABCA-1/log2', 'Zhuang-ABCA-1/raw']</code></pre>
</div>
</div>
<p><font size="3">List metadata files available in the Zhuang-ABCA-1 directory:</font></p><font size="3">
<div id="c5178bd4-2533-448c-a4c2-6622e4d3846c" class="cell" data-tags="[]" data-execution_count="9">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>abc_cache.list_metadata_files(<span class="st">'Zhuang-ABCA-1'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<pre><code>['cell_metadata', 'cell_metadata_with_cluster_annotation', 'gene']</code></pre>
</div>
</div>
<p><font size="3">Check the size of data and metadata:</font></p><font size="3">
<div id="59b39582-9e06-4269-a825-4c3c540a5eec" class="cell" data-tags="[]" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Size of data: </span><span class="sc">{</span>abc_cache<span class="sc">.</span>get_directory_data_size(<span class="st">'Zhuang-ABCA-1'</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Size of data: 3.09 GB</code></pre>
</div>
</div>
<p><font size="3">Obtain path of data and metadata:</font></p><font size="3">
<div class="sourceCode" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>path_data <span class="op">=</span> abc_cache.get_directory_data(<span class="st">'Zhuang-ABCA-1'</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Zhuang-ABCA-1 data files:</span><span class="ch">\n\t</span><span class="st">"</span>, path_data)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>```python</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>path_metadata <span class="op">=</span> abc_cache.get_directory_metadata(<span class="st">'Zhuang-ABCA-1'</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Zhuang-ABCA-1 metadata files:</span><span class="ch">\n\t</span><span class="st">"</span>, path_metadata)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;</span>We previously downloaded the data <span class="cf">for</span> you, you can find them <span class="kw">in</span> `abc_download_root<span class="op">/</span>`</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="co">### Data format</span></span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;</span> Expression matrices are stored <span class="kw">in</span> the [anndata h5ad <span class="bu">format</span>](https:<span class="op">//</span>anndata.readthedocs.io<span class="op">/</span>en<span class="op">/</span>latest<span class="op">/</span>) <span class="kw">and</span> need to be downloaded to a local <span class="bu">file</span> system <span class="cf">for</span> usage.</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;**</span>Key Components of an AnnData Object:<span class="op">**</span></span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;-</span> <span class="op">**</span>`.X`<span class="op">**</span>: The main data matrix (usually cells by genes), where the rows are cells <span class="kw">and</span> the columns are genes. It typically stores raw <span class="kw">or</span> normalized gene expression data.</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;-</span> <span class="op">**</span>`.obs`<span class="op">**</span>: Observations (metadata about the cells), such <span class="im">as</span> cell <span class="bu">type</span>, experimental conditions, <span class="kw">or</span> quality control metrics.</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;-</span> <span class="op">**</span>`.var`<span class="op">**</span>: Variables (metadata about the genes), such <span class="im">as</span> gene names <span class="kw">or</span> other annotations.</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;-</span> <span class="op">**</span>`.obsm`<span class="op">**</span>: Multi<span class="op">-</span>dimensional annotations of cells, such <span class="im">as</span> dimensionality reduction results (e.g., PCA, UMAP coordinates).</span>
<span id="cb18-23"><a href="#cb18-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-24"><a href="#cb18-24" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;-</span> <span class="op">**</span>`.uns`<span class="op">**</span>: Unstructured annotations, typically storing experimental metadata <span class="kw">or</span> plot settings.</span>
<span id="cb18-25"><a href="#cb18-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-26"><a href="#cb18-26" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;-</span> <span class="op">**</span>`.layers`<span class="op">**</span>: Stores multiple versions of the data matrix (e.g., raw counts, normalized data, etc.).</span>
<span id="cb18-27"><a href="#cb18-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-28"><a href="#cb18-28" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>figure<span class="op">&gt;</span></span>
<span id="cb18-29"><a href="#cb18-29" aria-hidden="true" tabindex="-1"></a>    <span class="op">&lt;</span>img src<span class="op">=</span><span class="st">"img/anndata_format.png"</span> alt<span class="op">=</span><span class="st">"AnnData format"</span> width<span class="op">=</span><span class="st">"500"</span> alt<span class="op">=</span><span class="st">"Description of the image"</span><span class="op">&gt;</span></span>
<span id="cb18-30"><a href="#cb18-30" aria-hidden="true" tabindex="-1"></a>    <span class="op">&lt;</span>figcaption<span class="op">&gt;</span>From <span class="op">&lt;</span>a href<span class="op">=</span><span class="st">"https://joss.theoj.org/papers/10.21105/joss.04371"</span> target<span class="op">=</span><span class="st">"_blank"</span><span class="op">&gt;</span>Virshup et al. (<span class="dv">2024</span>) JOSS<span class="op">&lt;/</span>a<span class="op">&gt;&lt;/</span>figcaption<span class="op">&gt;</span></span>
<span id="cb18-31"><a href="#cb18-31" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;/</span>figure<span class="op">&gt;</span></span>
<span id="cb18-32"><a href="#cb18-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-33"><a href="#cb18-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-34"><a href="#cb18-34" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;</span>Spatial coordinates, being metadata about cells<span class="op">/</span>spots are stored either <span class="kw">in</span> <span class="op">**</span>`.obs`<span class="op">**</span> <span class="kw">or</span> <span class="op">**</span>`.obsm`<span class="op">**</span></span>
<span id="cb18-35"><a href="#cb18-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-36"><a href="#cb18-36" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>span style<span class="op">=</span><span class="st">"color: red; font-size: 15px;"</span><span class="op">&gt;</span> <span class="op">**</span>Read raw count AnnData <span class="bu">object</span> first:<span class="op">**</span></span>
<span id="cb18-37"><a href="#cb18-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-38"><a href="#cb18-38" aria-hidden="true" tabindex="-1"></a>::: {<span class="co">#968b9679-277b-44d7-890e-72d8369987bf .cell tags='[]' execution_count=11}</span></span>
<span id="cb18-39"><a href="#cb18-39" aria-hidden="true" tabindex="-1"></a>``` {.python .cell<span class="op">-</span>code}</span>
<span id="cb18-40"><a href="#cb18-40" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 1</span></span>
<span id="cb18-41"><a href="#cb18-41" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint: use the function ad.read_h5ad </span></span>
<span id="cb18-42"><a href="#cb18-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-43"><a href="#cb18-43" aria-hidden="true" tabindex="-1"></a>path_raw_adata <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/abc_download_root/expression_matrices/Zhuang-ABCA-1/Zhuang-ABCA-1-raw.h5ad"</span></span>
<span id="cb18-44"><a href="#cb18-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-45"><a href="#cb18-45" aria-hidden="true" tabindex="-1"></a><span class="co"># don't change the variable name</span></span>
<span id="cb18-46"><a href="#cb18-46" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> ad.read_h5ad(path_raw_adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>:::</p>
<p><span style="color: red; font-size: 15px;"> <strong>Now read log-transformed AnnData object</strong></span></p>
<div id="86a53987-1024-4801-bbdb-55fb698df137" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 2</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint: use the function ad.read_h5ad </span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>path_log_adata <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/abc_download_root/expression_matrices/Zhuang-ABCA-1/Zhuang-ABCA-1-log2.h5ad"</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="co"># don't change the variable name</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>adata_log <span class="op">=</span> ad.read_h5ad(path_log_adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3">Now we need to add the log-transformed data to the <code>adata</code> object</font></p><font size="3">
<p><span style="color: red; font-size: 15px;"> <strong>To which component should it be added to?</strong></span></p>
<div id="db9276fc-6738-49e5-9588-8f34db172ebc" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 3</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the log-transformed counts to the 'log' layer and the count data to the 'raw' layer</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint: what slot of the adata object is suitable for count matrices?</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>adata.layers[<span class="st">'log2'</span>] <span class="op">=</span> adata_log.X</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>adata.layers[<span class="st">'raw'</span>] <span class="op">=</span> adata.X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3">We delete adata_log since it won’t be used anymore:</font></p><font size="3">
<div id="d467863b-4f3f-4004-9570-27be3deea412" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata_log</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>Read metadata:</strong></span></p>
<div id="9479ab6d-2cd9-4d92-a0b2-ed33f30b50de" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 4</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read genes</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint: use the function pd.read_csv and keep the gene identifier as index</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>path_metadata_genes <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/abc_download_root/metadata/Zhuang-ABCA-1/gene.csv"</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>genes <span class="op">=</span> pd.read_csv(path_metadata_genes,index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>genes.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_symbol</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">mapped_ncbi_identifier</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">gene_identifier</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000024798</td>
<td>Htr7</td>
<td>5-hydroxytryptamine (serotonin) receptor 7</td>
<td>NCBIGene:15566</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSMUSG00000042385</td>
<td>Gzmk</td>
<td>granzyme K</td>
<td>NCBIGene:14945</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000036198</td>
<td>Arhgap36</td>
<td>Rho GTPase activating protein 36</td>
<td>NCBIGene:75404</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSMUSG00000028780</td>
<td>Sema3c</td>
<td>sema domain, immunoglobulin domain (Ig), short...</td>
<td>NCBIGene:20348</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000015843</td>
<td>Rxrg</td>
<td>retinoid X receptor gamma</td>
<td>NCBIGene:20183</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="936c1cb3-d107-483c-a97e-ae3f1b1434ff" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 5</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Read cells</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint: use the function pd.read_csv and keep the cell label as index</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>path_metadata_cell <span class="op">=</span> <span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/abc_download_root/metadata/Zhuang-ABCA-1/views/cell_metadata_with_cluster_annotation.csv"</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>cell_metadata_extended <span class="op">=</span> pd.read_csv(path_metadata_cell,index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>cell_metadata_extended.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">brain_section_label</th>
<th data-quarto-table-cell-role="th">feature_matrix_label</th>
<th data-quarto-table-cell-role="th">donor_label</th>
<th data-quarto-table-cell-role="th">donor_genotype</th>
<th data-quarto-table-cell-role="th">donor_sex</th>
<th data-quarto-table-cell-role="th">cluster_alias</th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">y</th>
<th data-quarto-table-cell-role="th">z</th>
<th data-quarto-table-cell-role="th">subclass_confidence_score</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">neurotransmitter</th>
<th data-quarto-table-cell-role="th">class</th>
<th data-quarto-table-cell-role="th">subclass</th>
<th data-quarto-table-cell-role="th">supertype</th>
<th data-quarto-table-cell-role="th">cluster</th>
<th data-quarto-table-cell-role="th">neurotransmitter_color</th>
<th data-quarto-table-cell-role="th">class_color</th>
<th data-quarto-table-cell-role="th">subclass_color</th>
<th data-quarto-table-cell-role="th">supertype_color</th>
<th data-quarto-table-cell-role="th">cluster_color</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">cell_label</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">182941331246012878296807398333956011710</td>
<td>Zhuang-ABCA-1.089</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>704</td>
<td>0.682522</td>
<td>3.366483</td>
<td>7.82953</td>
<td>0.969933</td>
<td>...</td>
<td>GABA</td>
<td>06 CTX-CGE GABA</td>
<td>049 Lamp5 Gaba</td>
<td>0199 Lamp5 Gaba_1</td>
<td>0709 Lamp5 Gaba_1</td>
<td>#FF3358</td>
<td>#CCFF33</td>
<td>#FF764D</td>
<td>#DC00FF</td>
<td>#998900</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">221260934538535633595532020856387724686</td>
<td>Zhuang-ABCA-1.089</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5243</td>
<td>0.667690</td>
<td>3.442241</td>
<td>7.82953</td>
<td>0.850554</td>
<td>...</td>
<td>NaN</td>
<td>33 Vascular</td>
<td>331 Peri NN</td>
<td>1191 Peri NN_1</td>
<td>5304 Peri NN_1</td>
<td>#666666</td>
<td>#858881</td>
<td>#82992E</td>
<td>#2F00CC</td>
<td>#BB1FCC</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">22228792606814781533240955623030943708</td>
<td>Zhuang-ABCA-1.089</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>14939</td>
<td>0.638731</td>
<td>3.474328</td>
<td>7.82953</td>
<td>0.888285</td>
<td>...</td>
<td>NaN</td>
<td>30 Astro-Epen</td>
<td>319 Astro-TE NN</td>
<td>1163 Astro-TE NN_3</td>
<td>5225 Astro-TE NN_3</td>
<td>#666666</td>
<td>#594a26</td>
<td>#3DCCB1</td>
<td>#a8afa5</td>
<td>#551799</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">272043042552227961220474294517855477150</td>
<td>Zhuang-ABCA-1.089</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>14939</td>
<td>0.653425</td>
<td>3.433218</td>
<td>7.82953</td>
<td>0.900000</td>
<td>...</td>
<td>NaN</td>
<td>30 Astro-Epen</td>
<td>319 Astro-TE NN</td>
<td>1163 Astro-TE NN_3</td>
<td>5225 Astro-TE NN_3</td>
<td>#666666</td>
<td>#594a26</td>
<td>#3DCCB1</td>
<td>#a8afa5</td>
<td>#551799</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">110116287883089187971185374239350249328</td>
<td>Zhuang-ABCA-1.089</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5254</td>
<td>0.623896</td>
<td>3.513574</td>
<td>7.82953</td>
<td>0.999978</td>
<td>...</td>
<td>NaN</td>
<td>33 Vascular</td>
<td>333 Endo NN</td>
<td>1193 Endo NN_1</td>
<td>5310 Endo NN_1</td>
<td>#666666</td>
<td>#858881</td>
<td>#994567</td>
<td>#00992A</td>
<td>#FFB473</td>
</tr>
</tbody>
</table>

<p>5 rows × 22 columns</p>
</div>
</div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>Add gene names to the right component in the <code>adata</code> object</strong></span></p>
<div id="8e11ccc4-adfd-43d9-9395-dcccc2d5673b" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 6</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Where should we store the gene metadata?</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>adata.var <span class="op">=</span> genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3">Let’s show some stats about the dataset:</font></p><font size="3">
<div id="58b18b6f-b176-4fc8-93e9-1722c0ce2d86" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of sections in the dataset: </span><span class="sc">{</span><span class="bu">len</span>(cell_metadata_extended[<span class="st">'brain_section_label'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Total number of cells/spots in the dataset: </span><span class="sc">{</span>cell_metadata_extended<span class="sc">.</span>shape[<span class="dv">0</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Average number of cells/spots per section: </span><span class="sc">{</span>cell_metadata_extended<span class="sc">.</span>shape[<span class="dv">0</span>] <span class="op">/</span> <span class="bu">len</span>(cell_metadata_extended[<span class="st">'brain_section_label'</span>].unique())<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Size of the gene panel: </span><span class="sc">{</span>adata<span class="sc">.</span>shape[<span class="dv">1</span>]<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of sections in the dataset: 147
Total number of cells/spots in the dataset: 2846908
Average number of cells/spots per section: 19366.721088435374
Size of the gene panel: 1122</code></pre>
</div>
</div>
<p><font size="3">We decided to work with <strong>one section only</strong>; let’s select the one that captures most cells</font></p><font size="3">
<p><img src="img/ABCAtlas_Zhuang-ABCA-1_Coronal Grid.png" width="800" height="300"></p>
<p><span style="color: red; font-size: 15px;"> <strong>Which section has captures the highest number of cells?</strong></span></p>
<div id="e2001f8c-4aa9-49e9-bfd7-98dc74f26a6c" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 7</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint : use collections.Counter - https://docs.python.org/3/library/collections.html#collections.Counter</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> collections</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co">### Steps: </span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Identify the brain_section_label of the section with the highest number of cells</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Subset cell_metadata_extended and adata to include only cells from that section</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Identify the section that has the highest number of cells</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>section_counts <span class="op">=</span> collections.Counter(cell_metadata_extended[<span class="st">'brain_section_label'</span>])</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>section_df <span class="op">=</span> pd.DataFrame(section_counts.items(), columns<span class="op">=</span>[<span class="st">'brain_section_label'</span>, <span class="st">'count'</span>])</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the section with the highest number of cells</span></span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>top_section <span class="op">=</span> section_df.loc[section_df[<span class="st">'count'</span>].idxmax()]</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Section with highest number of cells: </span><span class="sc">{</span>top_section[<span class="st">'brain_section_label'</span>]<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Subset the metadata and data to include only spots from that section</span></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>top_section_label <span class="op">=</span> top_section[<span class="st">'brain_section_label'</span>] <span class="co"># extract label name</span></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>cell_metadata_extended_sub <span class="op">=</span> cell_metadata_extended[cell_metadata_extended[<span class="st">'brain_section_label'</span>] <span class="op">==</span> top_section_label] <span class="co"># subset metadata</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> adata[adata.obs[<span class="st">'brain_section_label'</span>] <span class="op">==</span> top_section_label] <span class="co"># subset data</span></span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Output subset info</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Subset contains </span><span class="sc">{</span>adata_section<span class="sc">.</span>shape[<span class="dv">0</span>]<span class="sc">}</span><span class="ss"> cells from section </span><span class="sc">{</span>top_section_label<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Section with highest number of cells: Zhuang-ABCA-1.080
Subset contains 51112 cells from section Zhuang-ABCA-1.080</code></pre>
</div>
</div>
<p><font size="3">Let’s inspect <code>adata_section</code> and <code>cell_metadata_extended_sub</code></font></p><font size="3">
<div id="30971d4e-e230-4fa6-9667-d9212b774adf" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>adata_section.obs.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">brain_section_label</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">cell_label</th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">267344286871071497676342741223031902383</td>
<td>Zhuang-ABCA-1.080</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">299794030110334465835331169565547130392</td>
<td>Zhuang-ABCA-1.080</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">233403047120023583151910226261717321273</td>
<td>Zhuang-ABCA-1.080</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">217976018079736491608380660316279559899</td>
<td>Zhuang-ABCA-1.080</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1869987565945418673548429740100940443</td>
<td>Zhuang-ABCA-1.080</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="48044fd0-212e-4fd1-9cd5-aa4a1efc846b" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>cell_metadata_extended_sub.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">brain_section_label</th>
<th data-quarto-table-cell-role="th">feature_matrix_label</th>
<th data-quarto-table-cell-role="th">donor_label</th>
<th data-quarto-table-cell-role="th">donor_genotype</th>
<th data-quarto-table-cell-role="th">donor_sex</th>
<th data-quarto-table-cell-role="th">cluster_alias</th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">y</th>
<th data-quarto-table-cell-role="th">z</th>
<th data-quarto-table-cell-role="th">subclass_confidence_score</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">neurotransmitter</th>
<th data-quarto-table-cell-role="th">class</th>
<th data-quarto-table-cell-role="th">subclass</th>
<th data-quarto-table-cell-role="th">supertype</th>
<th data-quarto-table-cell-role="th">cluster</th>
<th data-quarto-table-cell-role="th">neurotransmitter_color</th>
<th data-quarto-table-cell-role="th">class_color</th>
<th data-quarto-table-cell-role="th">subclass_color</th>
<th data-quarto-table-cell-role="th">supertype_color</th>
<th data-quarto-table-cell-role="th">cluster_color</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">cell_label</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">247976677407794870437144677714829459713</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>14939</td>
<td>5.897555</td>
<td>2.566162</td>
<td>6.757411</td>
<td>0.967143</td>
<td>...</td>
<td>NaN</td>
<td>30 Astro-Epen</td>
<td>319 Astro-TE NN</td>
<td>1163 Astro-TE NN_3</td>
<td>5225 Astro-TE NN_3</td>
<td>#666666</td>
<td>#594a26</td>
<td>#3DCCB1</td>
<td>#a8afa5</td>
<td>#551799</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">168655970434508722627007022423841385743</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>329</td>
<td>5.896700</td>
<td>2.367992</td>
<td>6.757411</td>
<td>0.999981</td>
<td>...</td>
<td>Glut</td>
<td>01 IT-ET Glut</td>
<td>016 CA1-ProS Glut</td>
<td>0069 CA1-ProS Glut_1</td>
<td>0263 CA1-ProS Glut_1</td>
<td>#2B93DF</td>
<td>#FA0087</td>
<td>#0F3466</td>
<td>#8800CC</td>
<td>#009983</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">306182408260920364328788238077659872267</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>328</td>
<td>5.903615</td>
<td>2.375955</td>
<td>6.757411</td>
<td>0.999961</td>
<td>...</td>
<td>Glut</td>
<td>01 IT-ET Glut</td>
<td>016 CA1-ProS Glut</td>
<td>0069 CA1-ProS Glut_1</td>
<td>0262 CA1-ProS Glut_1</td>
<td>#2B93DF</td>
<td>#FA0087</td>
<td>#0F3466</td>
<td>#8800CC</td>
<td>#FF0096</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">323418747114703574223827093207552818826</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>14939</td>
<td>5.891746</td>
<td>2.252074</td>
<td>6.757411</td>
<td>0.934844</td>
<td>...</td>
<td>NaN</td>
<td>30 Astro-Epen</td>
<td>319 Astro-TE NN</td>
<td>1163 Astro-TE NN_3</td>
<td>5225 Astro-TE NN_3</td>
<td>#666666</td>
<td>#594a26</td>
<td>#3DCCB1</td>
<td>#a8afa5</td>
<td>#551799</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">150648836575778299138347196129291396742</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>14859</td>
<td>5.902247</td>
<td>2.198086</td>
<td>6.757411</td>
<td>0.999988</td>
<td>...</td>
<td>NaN</td>
<td>30 Astro-Epen</td>
<td>319 Astro-TE NN</td>
<td>1164 Astro-TE NN_4</td>
<td>5228 Astro-TE NN_4</td>
<td>#666666</td>
<td>#594a26</td>
<td>#3DCCB1</td>
<td>#604a55</td>
<td>#99FFAA</td>
</tr>
</tbody>
</table>

<p>5 rows × 22 columns</p>
</div>
</div>
</div>
<p><font size="3"> We notice that the indexes don’t match! It’s important to reorder the indexes in cell_metadata_extended_sub to match the ones in adata_section.obs:</font></p><font size="3">
<div id="927535a0-b816-414c-a015-27c9518fba56" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>cell_metadata_extended_sub_reind <span class="op">=</span> cell_metadata_extended_sub.reindex(adata_section.obs.index)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a62ea072-2d64-4622-81e7-e1958be2d164" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>collections.Counter(cell_metadata_extended_sub_reind.index <span class="op">==</span> adata_section.obs.index)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<pre><code>Counter({True: 51112})</code></pre>
</div>
</div>
<p><font size="3">Now we can add cell_metadata_extended_sub as metadata to adata_section</font></p><font size="3">
<div id="febd57ff-bce7-42b1-811d-19179435e221" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>adata_section.obs <span class="op">=</span> cell_metadata_extended_sub_reind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3">Plot the section and color them by all the different classifications avaiable in the metadata (class_color, subclass_color, supertype_color, cluster_color):</font></p><font size="3">
<div id="8470a4f6-e392-42f3-b15f-2a3f95011163" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the color mapping dictionaries</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>class_palette <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(adata_section.obs[<span class="st">'class'</span>].unique(), adata_section.obs[<span class="st">'class_color'</span>].unique()))</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>subclass_palette <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(adata_section.obs[<span class="st">'subclass'</span>].unique(), adata_section.obs[<span class="st">'subclass_color'</span>].unique()))</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>supertype_palette <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(adata_section.obs[<span class="st">'supertype'</span>].unique(), adata_section.obs[<span class="st">'supertype_color'</span>].unique()))</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>cluster_palette <span class="op">=</span> <span class="bu">dict</span>(<span class="bu">zip</span>(adata_section.obs[<span class="st">'cluster'</span>].unique(), adata_section.obs[<span class="st">'cluster_color'</span>].unique()))</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the subplots</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">15</span>))</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>fig.suptitle(<span class="st">'Scatter Plots of Different Categories'</span>, fontsize<span class="op">=</span><span class="dv">16</span>)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot 'class' using class_color</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(ax<span class="op">=</span>axes[<span class="dv">0</span>, <span class="dv">0</span>], x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=-</span>adata_section.obs[<span class="st">'y'</span>], hue<span class="op">=</span><span class="st">'class'</span>, palette<span class="op">=</span>class_palette, data<span class="op">=</span>adata_section.obs, s<span class="op">=</span><span class="dv">8</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">0</span>].set_title(<span class="st">'Class'</span>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">0</span>].set_xlabel(<span class="st">'X Coordinate'</span>)</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">0</span>].set_ylabel(<span class="st">'Y Coordinate'</span>)</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">0</span>].tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">0</span>].grid(<span class="va">False</span>)  <span class="co"># Remove grid</span></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot 'subclass' using subclass_color</span></span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(ax<span class="op">=</span>axes[<span class="dv">0</span>, <span class="dv">1</span>], x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=-</span>adata_section.obs[<span class="st">'y'</span>], hue<span class="op">=</span><span class="st">'subclass'</span>, palette<span class="op">=</span>subclass_palette, data<span class="op">=</span>adata_section.obs, s<span class="op">=</span><span class="dv">8</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">1</span>].set_title(<span class="st">'Subclass'</span>)</span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">1</span>].set_xlabel(<span class="st">'X Coordinate'</span>)</span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">1</span>].set_ylabel(<span class="st">'Y Coordinate'</span>)</span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">1</span>].tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">0</span>, <span class="dv">1</span>].grid(<span class="va">False</span>)  <span class="co"># Remove grid</span></span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot 'supertype' using supertype_color</span></span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(ax<span class="op">=</span>axes[<span class="dv">1</span>, <span class="dv">0</span>], x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=-</span>adata_section.obs[<span class="st">'y'</span>], hue<span class="op">=</span><span class="st">'supertype'</span>, palette<span class="op">=</span>supertype_palette, data<span class="op">=</span>adata_section.obs, s<span class="op">=</span><span class="dv">8</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">0</span>].set_title(<span class="st">'Supertype'</span>)</span>
<span id="cb35-30"><a href="#cb35-30" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">0</span>].set_xlabel(<span class="st">'X Coordinate'</span>)</span>
<span id="cb35-31"><a href="#cb35-31" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">0</span>].set_ylabel(<span class="st">'Y Coordinate'</span>)</span>
<span id="cb35-32"><a href="#cb35-32" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">0</span>].tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-33"><a href="#cb35-33" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">0</span>].grid(<span class="va">False</span>)  <span class="co"># Remove grid</span></span>
<span id="cb35-34"><a href="#cb35-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-35"><a href="#cb35-35" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot 'cluster' using cluster_color</span></span>
<span id="cb35-36"><a href="#cb35-36" aria-hidden="true" tabindex="-1"></a>sns.scatterplot(ax<span class="op">=</span>axes[<span class="dv">1</span>, <span class="dv">1</span>], x<span class="op">=</span><span class="st">'x'</span>, y<span class="op">=-</span>adata_section.obs[<span class="st">'y'</span>], hue<span class="op">=</span><span class="st">'cluster'</span>, palette<span class="op">=</span>cluster_palette, data<span class="op">=</span>adata_section.obs, s<span class="op">=</span><span class="dv">8</span>, legend<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-37"><a href="#cb35-37" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">1</span>].set_title(<span class="st">'Cluster'</span>)</span>
<span id="cb35-38"><a href="#cb35-38" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">1</span>].set_xlabel(<span class="st">'X Coordinate'</span>)</span>
<span id="cb35-39"><a href="#cb35-39" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">1</span>].set_ylabel(<span class="st">'Y Coordinate'</span>)</span>
<span id="cb35-40"><a href="#cb35-40" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">1</span>].tick_params(axis<span class="op">=</span><span class="st">'both'</span>, which<span class="op">=</span><span class="st">'both'</span>, bottom<span class="op">=</span><span class="va">False</span>, left<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb35-41"><a href="#cb35-41" aria-hidden="true" tabindex="-1"></a>axes[<span class="dv">1</span>, <span class="dv">1</span>].grid(<span class="va">False</span>)  <span class="co"># Remove grid</span></span>
<span id="cb35-42"><a href="#cb35-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-43"><a href="#cb35-43" aria-hidden="true" tabindex="-1"></a>plt.tight_layout(rect<span class="op">=</span>[<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.96</span>])  </span>
<span id="cb35-44"><a href="#cb35-44" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-26-output-1.png" width="1182" height="1175" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="quality-control" class="level3">
<h3 class="anchored" data-anchor-id="quality-control">Quality Control</h3>
<p><font size="3">From https://scanpy-tutorials.readthedocs.io/en/latest/spatial/basic-analysis.html</font></p><font size="3">
<div id="4bbfbbab-1a81-4092-92df-445891b6b179" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make gene names unique</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>adata_section.var_names_make_unique()</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="co"># add mt percentage (mention that there's no need for this ds bc it's targeted tr)</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>adata_section.var[<span class="st">"mt"</span>] <span class="op">=</span> adata_section.var_names.<span class="bu">str</span>.startswith(<span class="st">"MT-"</span>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata_section, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="16284989-0551-4c98-b12e-f909aece2eb6" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">4</span>))</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>sns.histplot(adata_section.obs[<span class="st">"total_counts"</span>], kde<span class="op">=</span><span class="va">False</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>])</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>sns.histplot(</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>    adata_section.obs[<span class="st">"total_counts"</span>][adata_section.obs[<span class="st">"total_counts"</span>] <span class="op">&lt;</span> <span class="dv">10000</span>],</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>    kde<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>    bins<span class="op">=</span><span class="dv">40</span>,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>    ax<span class="op">=</span>axs[<span class="dv">1</span>],</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-28-output-1.png" width="707" height="302" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<section id="cell-filtering-and-preprocessing-in-single-cell-rna-seq-analysis" class="level4">
<h4 class="anchored" data-anchor-id="cell-filtering-and-preprocessing-in-single-cell-rna-seq-analysis">Cell Filtering and Preprocessing in Single-Cell RNA-Seq Analysis</h4>
<p><font size="3">In this section, we are performing a series of filtering steps to preprocess the single-cell RNA sequencing (scRNA-seq) data stored in the <code>adata_section</code> object. The filtering aims to retain high-quality cells and genes for downstream analyses. The following steps are applied:</font></p><font size="3">
<ol type="1">
<li><font size="3"><strong>Filter Cells by Minimum Counts</strong>:
<ul>
<li><font size="3">We filter out cells that have fewer than 200 counts. This helps to exclude low-quality cells that may not provide reliable information.</font></li><font size="3">
</font></ul></font></li><font size="3"><font size="3">
</font></font></ol><font size="3"><font size="3">
<div id="e2c147d1-483f-4ee4-85ba-ed310739df71" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata_section, min_counts<span class="op">=</span><span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="1">
<li><font size="3"><strong>Filter Cells by Maximum Counts</strong>:
<ul>
<li><font size="3">We also filter out cells with more than 25,000 counts to remove potential doublets or outlier cells that may artificially inflate the count data</font></li><font size="3">
</font></ul></font></li><font size="3"><font size="3">
</font></font></ol><font size="3"><font size="3">
<div id="67cc3092-4379-4b6b-875d-38edbe3d17bf" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata_section, max_counts<span class="op">=</span><span class="dv">25000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="3" type="1">
<li><p><font size="3"><strong>Filter Based on Mitochondrial Gene Expression</strong>:</font></p><font size="3">
<ul>
<li><font size="3">We further refine our dataset by excluding cells that have more than 10% of their total counts coming from mitochondrial genes.</font></li><font size="3">
</font></ul></font></li><font size="3"><font size="3">
</font></font></ol><font size="3"><font size="3">
<div id="43186dec-2f23-4b81-8ebf-2892979a2d70" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> adata_section[adata_section.obs[<span class="st">"pct_counts_mt"</span>] <span class="op">&lt;</span> <span class="dv">10</span>].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="4" type="1">
<li><font size="3"><strong>Filter Genes by Minimum Cells</strong>:</font></li><font size="3">
</font></ol><font size="3">
<ul>
<li><font size="3">Finally, we filter out genes that are detected in fewer than 20 cells, ensuring that we retain only those genes that have sufficient expression across our dataset.</font></li><font size="3">
</font></ul><font size="3">
<p><span style="color: red; font-size: 15px;"> <strong>Filter out genes that are detected in less than 20 cells</strong></span></p>
<div id="8ba80114-d852-42f6-b700-70b596160a76" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 8</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="co">### Hint: use sc.pp.filter_genes</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_genes(adata_section, min_cells<span class="op">=</span><span class="dv">20</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d1da18a6-f45b-44ed-a4eb-59a8cc7a9902" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of cells after filtering: </span><span class="sc">{</span>adata_section<span class="sc">.</span>n_obs<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of cells after filtering: 36206</code></pre>
</div>
</div>
<p><font size="3"> Perform PCA and UMAP and cluster with Leiden Algorithm</font></p><font size="3">
<div id="77137142-5f01-452a-9015-7dc8b82dfd15" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>sc.pp.pca(adata_section)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata_section)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata_section)</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>    adata_section, </span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>    key_added<span class="op">=</span><span class="st">"clusters"</span>, </span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>    resolution<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    directed<span class="op">=</span><span class="va">False</span>, </span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    n_iterations<span class="op">=</span><span class="dv">2</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="871dc765-9919-444e-ae31-f0c0be0b97c9" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> (<span class="dv">8</span>,<span class="dv">8</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata_section, color<span class="op">=</span>[<span class="st">"total_counts"</span>, <span class="st">"n_genes_by_counts"</span>, <span class="st">"clusters"</span>], wspace<span class="op">=</span><span class="fl">0.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-35-output-1.png" width="1992" height="547" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><font size="3"> Some stats:</font></p><font size="3">
<div id="0b5e5424-5188-49ed-9678-b048bb2c5121" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Number of Neurotransmitters: '</span>, <span class="bu">len</span>(adata_section.obs[<span class="st">'neurotransmitter'</span>].unique()))</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Number of Classes: '</span>, <span class="bu">len</span>(adata_section.obs[<span class="st">'class'</span>].unique()))</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Number of Subclasses: '</span>, <span class="bu">len</span>(adata_section.obs[<span class="st">'subclass'</span>].unique()))</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Number of Supertype: '</span>, <span class="bu">len</span>(adata_section.obs[<span class="st">'supertype'</span>].unique()))</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Number of Cluster: '</span>, <span class="bu">len</span>(adata_section.obs[<span class="st">'cluster'</span>].unique()))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of Neurotransmitters:  7
Number of Classes:  25
Number of Subclasses:  136
Number of Supertype:  420
Number of Cluster:  1000</code></pre>
</div>
</div>
<p><font size="3"> The filtering step performed by the authors of the paper is done differently than the basic one we just performed. The metadata is given only for the filtered object. For this reason we will filter our object to keep only spots that passed the quality control for the Allen Brain Atlas standards.</font></p><font size="3">
<div id="d2c9769e-7819-447b-a13c-968f377167ee" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter spots in adata_section and keep only the ones that are present in metadata</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> adata[adata.obs.index.isin(cell_metadata_extended_sub.index)]</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="co"># reindex cell_metadata_ext_sub to match the index order of adata_section.obs</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>cell_metadata_extended_sub <span class="op">=</span> cell_metadata_extended_sub.reindex(adata_section.obs.index)</span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add metadata to the metadata field in an anndata object : adata.obj</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>adata_section.obs <span class="op">=</span> cell_metadata_extended_sub</span>
<span id="cb48-7"><a href="#cb48-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb48-8"><a href="#cb48-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"The filtered Section 80 dataset has </span><span class="sc">{</span>adata_section<span class="sc">.</span>shape[<span class="dv">0</span>]<span class="sc">}</span><span class="ss"> cells and </span><span class="sc">{</span>adata_section<span class="sc">.</span>shape[<span class="dv">1</span>]<span class="sc">}</span><span class="ss"> genes."</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>The filtered Section 80 dataset has 37068 cells and 1122 genes.</code></pre>
</div>
</div>
<p><font size="3"> To free RAM, we delete the <code>adata</code> object as we will use only the section:</font></p><font size="3">
<div id="7528d536-10e2-4e98-baa9-a7d429deafe0" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3"> Some spatial tools require the spatial coordinates can be stored as <code>spatial</code> or <code>X_spatial</code></font></p><font size="3">
<div id="10c021bf-aa04-4477-a82c-9a9d6c06cdf8" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 9</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Add X_spatial to the right field</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>x_spatial <span class="op">=</span> np.array(adata_section.obs[[<span class="st">'x'</span>,<span class="st">'y'</span>]])</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>adata_section.obsm[<span class="st">'X_spatial'</span>] <span class="op">=</span> x_spatial</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>adata_section.obsm[<span class="st">'X_spatial'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="38">
<pre><code>array([[5.87382373, 1.24637434],
       [5.0668137 , 1.21544436],
       [4.68106119, 1.28715849],
       ...,
       [0.97304079, 4.80513099],
       [0.86566165, 4.29416095],
       [0.84111779, 3.8952338 ]])</code></pre>
</div>
</div>
<p><font size="3"> Once the spatial coordinates are stored in <code>.obsm[X_spatial]</code> you can plot the section with <code>sc.pl.spatial</code></font></p><font size="3">
<div id="e307ded4-983e-48d4-8a7c-42c4f5f6e66a" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate the spatial plot</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata_section, color<span class="op">=</span><span class="st">'class'</span>, color_map<span class="op">=</span><span class="st">'Accent_r'</span>, spot_size<span class="op">=</span><span class="fl">0.04</span>, title<span class="op">=</span><span class="st">'Section 80 colored by Class'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-40-output-1.png" width="809" height="541" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="ff4a5495-ff32-42df-a6e4-76b17d35a6f9" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 10</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the AnnData object to keep only cells belonging to a specific class</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co"># and plot them again with sc.pl.spatial</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>adata_section_one_class <span class="op">=</span> adata_section[adata_section.obs[<span class="st">'class'</span>]<span class="op">==</span><span class="st">'01 IT-ET Glut'</span>]</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata_section_one_class, color<span class="op">=</span><span class="st">'class'</span>, color_map<span class="op">=</span><span class="st">'Accent_r'</span>, spot_size<span class="op">=</span><span class="fl">0.04</span>, title<span class="op">=</span><span class="st">'Section 80 (one Class)'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-41-output-1.png" width="638" height="541" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></section><font size="3"><font size="3"><font size="3">
</font></font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="plot-gene-expression-in-space" class="level3">
<h3 class="anchored" data-anchor-id="plot-gene-expression-in-space">Plot gene expression in space</h3>
<p><font size="3"></font></p><font size="3">
<p>Since each spot in the spatial data corresponds to a specific location and its associated gene expression profile, we can visualize spatial gene expression.</p>
<p>We can do this using the function <code>sc.pl.spatial</code>, it requires the name of the variable index to indicate the genes to plot.</p>
<p>Let’s take a look at <code>adata_section</code>.var:</p>
<div id="d001c6ad-0581-4e54-bcc8-5ea0db2d147c" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>adata_section.var.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_symbol</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">mapped_ncbi_identifier</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">gene_identifier</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000024798</td>
<td>Htr7</td>
<td>5-hydroxytryptamine (serotonin) receptor 7</td>
<td>NCBIGene:15566</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSMUSG00000042385</td>
<td>Gzmk</td>
<td>granzyme K</td>
<td>NCBIGene:14945</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000036198</td>
<td>Arhgap36</td>
<td>Rho GTPase activating protein 36</td>
<td>NCBIGene:75404</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSMUSG00000028780</td>
<td>Sema3c</td>
<td>sema domain, immunoglobulin domain (Ig), short...</td>
<td>NCBIGene:20348</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000015843</td>
<td>Rxrg</td>
<td>retinoid X receptor gamma</td>
<td>NCBIGene:20183</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><font size="3"></font></p><font size="3">
<p>Change the index of the <code>adata_section.var</code> to <code>gene_symbol</code> for clarity</p>
<div id="bf8740b3-6142-42df-8318-1179a739d04b" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reset the index to turn the current index into a column</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>adata_section.var.reset_index(inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>adata_section.var.set_index(<span class="st">'gene_symbol'</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>adata_section.var.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="42">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_identifier</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">mapped_ncbi_identifier</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">gene_symbol</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Htr7</td>
<td>ENSMUSG00000024798</td>
<td>5-hydroxytryptamine (serotonin) receptor 7</td>
<td>NCBIGene:15566</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Gzmk</td>
<td>ENSMUSG00000042385</td>
<td>granzyme K</td>
<td>NCBIGene:14945</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Arhgap36</td>
<td>ENSMUSG00000036198</td>
<td>Rho GTPase activating protein 36</td>
<td>NCBIGene:75404</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Sema3c</td>
<td>ENSMUSG00000028780</td>
<td>sema domain, immunoglobulin domain (Ig), short...</td>
<td>NCBIGene:20348</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Rxrg</td>
<td>ENSMUSG00000015843</td>
<td>retinoid X receptor gamma</td>
<td>NCBIGene:20183</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="11c787c1-911c-4bbb-b621-fee7010bbeb2" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define genes to plot</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>genes_toplot <span class="op">=</span> [<span class="st">'Zic1'</span>,</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a> <span class="st">'Penk'</span>,</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a> <span class="st">'Gfap'</span>,</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a> <span class="st">'Dlx6'</span>,</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a> <span class="st">'Cux2'</span>,</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a> <span class="st">'Fezf2'</span>,</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a> <span class="st">'Nr4a2'</span>,</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a> <span class="st">'Satb2'</span>]</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata_section, color<span class="op">=</span>genes_toplot,spot_size<span class="op">=</span><span class="fl">0.04</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-44-output-1.png" width="2260" height="1104" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>Plot your favourite gene/s</strong></span></p>
<div id="643caa15-8d5d-472e-8668-97fd010b768e" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co">#genes_toplot = [.....]</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="co">#sc.pl.spatial(adata_section, color=genes_toplot,spot_size=0.04)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3bad7970-e5b8-421a-8e14-91c382d2eabc" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>adata_section.write_h5ad(<span class="ss">f"./outputs/adata_section80.h5ad"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c9c35bfe-bcf8-4d3b-8493-ca07c89f5c4e" class="cell" data-execution_count="46">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Record the end time</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>end_time <span class="op">=</span> time.time()</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate elapsed time in minutes</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>elapsed_time <span class="op">=</span> (end_time <span class="op">-</span> start_time) <span class="op">/</span> <span class="dv">60</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Execution Time: </span><span class="sc">{</span>elapsed_time<span class="sc">:.2f}</span><span class="ss"> minutes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Execution Time: 5.45 minutes</code></pre>
</div>
</div>
</font></font></section><font size="3"><font size="3">
</font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="prediction-of-spatial-patterns-for-unmeasured-genes-using-scrna-seq-data-with-spage" class="level1">
<h1>Prediction of spatial patterns for unmeasured genes using scRNA-seq data with SpaGE</h1>
<div id="b24a6480-c29d-4399-97c5-1ceb21638f73" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>start_time <span class="op">=</span> time.time()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="f356ab26-9e1f-4c01-81e4-739efec8e1ff" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> SpaGE.main <span class="im">import</span> SpaGE</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scipy.stats <span class="im">as</span> st</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<figure class="figure">
<img src="img/spage.jpg" width="700" height="280" alt="Description of the image" class="figure-img">
<figcaption>
From <a href="https://academic.oup.com/nar/article/48/18/e107/5909530" target="_blank">Abdelaal et al.&nbsp;(2020) NAR</a>
</figcaption>
</figure>
<p><font size="3"> SpaGE takes as input the (targeted) spatial dataset (query) and a single-cell RNA dataset used as reference. The two datasets should be generated under the same conditions (same species, tissue, region) as much as possible.</font></p><font size="3">
<p><font size="3"> We will use <a href="https://www.nature.com/articles/s41586-023-06812-z">Mouse whole-brain transcriptomic cell type atlas (Hongkui Zeng)</a>.This scRNAseq dataset profiled 7 million cells (approximately 4.0 million cells passing quality control). More in detail:</font></p><font size="3">
<ul>
<li><p><font size="3"> 1.7 million single cell transcriptomes spanning the whole adult mouse brain using 10Xv2 chemistry (WMB-10Xv2)</font></p></li><font size="3">
<li><p><font size="3"> 2.3 million single cell transcriptomes spanning the whole adult mouse brain using 10Xv3 chemistry (WMB-10Xv3)</font></p></li><font size="3">
<li><p><font size="3"> 1687 single cell transcriptomes spanning the whole adult mouse brain using the 10X Multiome chemistry (WMB-10XMulti)</font></p></li><font size="3">
</font></font></font></ul><font size="3"><font size="3"><font size="3">
<figure class="figure">
<img src="img/sc.png" width="700" height="280" alt="Description of the image" class="figure-img">
<figcaption>
From <a href="https://www.nature.com/articles/s41586-023-06812-z" target="_blank">Yao et al.&nbsp;(2023) Nature Methods</a>
</figcaption>
</figure>
<p><font size="3"> Let’s check the names of the files in each of the three single-cell collections:</font></p><font size="3">
<div id="a26d4f58-a9de-461e-bdb6-23ff9a9f3307" class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>abc_cache.list_data_files(<span class="st">'WMB-10Xv2'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="49">
<pre><code>['WMB-10Xv2-CTXsp/log2',
 'WMB-10Xv2-CTXsp/raw',
 'WMB-10Xv2-HPF/log2',
 'WMB-10Xv2-HPF/raw',
 'WMB-10Xv2-HY/log2',
 'WMB-10Xv2-HY/raw',
 'WMB-10Xv2-Isocortex-1/log2',
 'WMB-10Xv2-Isocortex-1/raw',
 'WMB-10Xv2-Isocortex-2/log2',
 'WMB-10Xv2-Isocortex-2/raw',
 'WMB-10Xv2-Isocortex-3/log2',
 'WMB-10Xv2-Isocortex-3/raw',
 'WMB-10Xv2-Isocortex-4/log2',
 'WMB-10Xv2-Isocortex-4/raw',
 'WMB-10Xv2-MB/log2',
 'WMB-10Xv2-MB/raw',
 'WMB-10Xv2-OLF/log2',
 'WMB-10Xv2-OLF/raw',
 'WMB-10Xv2-TH/log2',
 'WMB-10Xv2-TH/raw']</code></pre>
</div>
</div>
<div id="1d51adb7-1297-434e-9b20-fa4fca8cdaf4" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>abc_cache.list_data_files(<span class="st">'WMB-10Xv3'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="50">
<pre><code>['WMB-10Xv3-CB/log2',
 'WMB-10Xv3-CB/raw',
 'WMB-10Xv3-CTXsp/log2',
 'WMB-10Xv3-CTXsp/raw',
 'WMB-10Xv3-HPF/log2',
 'WMB-10Xv3-HPF/raw',
 'WMB-10Xv3-HY/log2',
 'WMB-10Xv3-HY/raw',
 'WMB-10Xv3-Isocortex-1/log2',
 'WMB-10Xv3-Isocortex-1/raw',
 'WMB-10Xv3-Isocortex-2/log2',
 'WMB-10Xv3-Isocortex-2/raw',
 'WMB-10Xv3-MB/log2',
 'WMB-10Xv3-MB/raw',
 'WMB-10Xv3-MY/log2',
 'WMB-10Xv3-MY/raw',
 'WMB-10Xv3-OLF/log2',
 'WMB-10Xv3-OLF/raw',
 'WMB-10Xv3-P/log2',
 'WMB-10Xv3-P/raw',
 'WMB-10Xv3-PAL/log2',
 'WMB-10Xv3-PAL/raw',
 'WMB-10Xv3-STR/log2',
 'WMB-10Xv3-STR/raw',
 'WMB-10Xv3-TH/log2',
 'WMB-10Xv3-TH/raw']</code></pre>
</div>
</div>
<div id="822963e0-8edd-4141-825a-89964e4aff6e" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>abc_cache.list_data_files(<span class="st">'WMB-10XMulti'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="51">
<pre><code>['WMB-10XMulti/log2', 'WMB-10XMulti/raw']</code></pre>
</div>
</div>
<p><font size="3"> Let’s check the size of the files in each of the three single-cell collections:</font></p><font size="3">
<div id="3fbc4886-1bdf-46a2-8c35-6b491dc1e406" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Size of WMB-10Xv2 data: </span><span class="sc">{</span>abc_cache<span class="sc">.</span>get_directory_data_size(<span class="st">'WMB-10Xv2'</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Size of WMB-10Xv2 data: 104.16 GB</code></pre>
</div>
</div>
<div id="a913a0a1-804b-4cb3-812a-9afe06a4ea5d" class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Size of WMB-10Xv3 data: </span><span class="sc">{</span>abc_cache<span class="sc">.</span>get_directory_data_size(<span class="st">'WMB-10Xv3'</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Size of WMB-10Xv3 data: 176.41 GB</code></pre>
</div>
</div>
<div id="1f5454ee-2afe-47cb-874d-9dc7e91f82d9" class="cell" data-execution_count="54">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Size of WMB-10X Multi data: </span><span class="sc">{</span>abc_cache<span class="sc">.</span>get_directory_data_size(<span class="st">'WMB-10XMulti'</span>)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Size of WMB-10X Multi data: 211.28 MB</code></pre>
</div>
</div>
<p><font size="3"> The size is of the combined single-cell object is around 280GB. For this reason, the three objects have been previously downloaded, concatenated, preprocessed and subsetted to obtain NNN cells. For your knowledge, you can find all the steps in <code>obtain_scref.py</code>.</font></p><font size="3">
<p><font size="3"> Let’s read the resulting single-cell AnnData object:</font></p><font size="3">
<div id="926edba6-5733-4042-84f3-7443ed238f0a" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>adata_sc <span class="op">=</span> ad.read_h5ad(<span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/SpaGE_data/sc_adata_1percent.h5ad"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3"> Inspect the object:</font></p><font size="3">
<div id="9a2ca8b8-830e-472b-bbb5-56db7fff393b" class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>adata_sc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="56">
<pre><code>AnnData object with n_obs × n_vars = 28981 × 32285
    obs: 'cell_barcode', 'barcoded_cell_sample_label', 'library_label', 'feature_matrix_label', 'entity', 'brain_section_label', 'library_method', 'region_of_interest_acronym', 'donor_label', 'donor_genotype', 'donor_sex', 'dataset_label', 'x', 'y', 'cluster_alias'</code></pre>
</div>
</div>
<div id="656a5fd7-5b6b-4d14-bcd2-73e2cf64f7d4" class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>adata_sc.obs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="57">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">cell_barcode</th>
<th data-quarto-table-cell-role="th">barcoded_cell_sample_label</th>
<th data-quarto-table-cell-role="th">library_label</th>
<th data-quarto-table-cell-role="th">feature_matrix_label</th>
<th data-quarto-table-cell-role="th">entity</th>
<th data-quarto-table-cell-role="th">brain_section_label</th>
<th data-quarto-table-cell-role="th">library_method</th>
<th data-quarto-table-cell-role="th">region_of_interest_acronym</th>
<th data-quarto-table-cell-role="th">donor_label</th>
<th data-quarto-table-cell-role="th">donor_genotype</th>
<th data-quarto-table-cell-role="th">donor_sex</th>
<th data-quarto-table-cell-role="th">dataset_label</th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">y</th>
<th data-quarto-table-cell-role="th">cluster_alias</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">cell_label</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">GTATTCTCAGCTATTG-045_C01</td>
<td>GTATTCTCAGCTATTG</td>
<td>045_C01</td>
<td>L8TX_180829_01_C10</td>
<td>WMB-10Xv2-TH</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv2</td>
<td>TH</td>
<td>Snap25-IRES2-Cre;Ai14-407905</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv2</td>
<td>12.420248</td>
<td>-10.367616</td>
<td>5076</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GTGTGATCACCAGCGT-468_A02</td>
<td>GTGTGATCACCAGCGT</td>
<td>468_A02</td>
<td>L8TX_201217_01_D07</td>
<td>WMB-10Xv3-MB</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>MB</td>
<td>Snap25-IRES2-Cre;Ai14-556231</td>
<td>Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv3</td>
<td>1.571478</td>
<td>23.592803</td>
<td>14956</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TAACCAGCACCAAATC-204.2_A01</td>
<td>TAACCAGCACCAAATC</td>
<td>204.2_A01</td>
<td>L8TX_200122_01_F12</td>
<td>WMB-10Xv3-MB</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>MB</td>
<td>Snap25-IRES2-Cre;Ai14-504535</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv3</td>
<td>-12.491472</td>
<td>11.322906</td>
<td>5231</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">AGCGTCGGTTCAGCGC-420_C06</td>
<td>AGCGTCGGTTCAGCGC</td>
<td>420_C06</td>
<td>L8TX_201106_01_G03</td>
<td>WMB-10Xv3-MB</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>MB</td>
<td>Snap25-IRES2-Cre;Ai14-551017</td>
<td>Ai14(RCL-tdT)/wt</td>
<td>M</td>
<td>WMB-10Xv3</td>
<td>-10.406498</td>
<td>11.711900</td>
<td>5230</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">AGACAAAAGAACTCCT-278_B03</td>
<td>AGACAAAAGAACTCCT</td>
<td>278_B03</td>
<td>L8TX_200702_01_H01</td>
<td>WMB-10Xv3-TH</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>TH</td>
<td>Snap25-IRES2-Cre;Ai14-529474</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv3</td>
<td>5.455338</td>
<td>-13.476950</td>
<td>5101</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">CGAGTTACAGCACAGA-367_B04</td>
<td>CGAGTTACAGCACAGA</td>
<td>367_B04</td>
<td>L8TX_200924_01_G05</td>
<td>WMB-10Xv3-STR</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>STRd</td>
<td>Snap25-IRES2-Cre;Ai14-544400</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv3</td>
<td>11.337216</td>
<td>10.187603</td>
<td>1092</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GCCCGAAGTGTTAACC-136_A01</td>
<td>GCCCGAAGTGTTAACC</td>
<td>136_A01</td>
<td>L8TX_190716_01_E07</td>
<td>WMB-10Xv3-STR</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>LSX</td>
<td>Snap25-IRES2-Cre;Ai14-471388</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv3</td>
<td>-10.906924</td>
<td>11.981008</td>
<td>5230</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TAGGGTTCACTCGATA-412_B07</td>
<td>TAGGGTTCACTCGATA</td>
<td>412_B07</td>
<td>L8TX_201030_01_D12</td>
<td>WMB-10Xv3-HPF</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>RHP</td>
<td>Snap25-IRES2-Cre;Ai14-550849</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>M</td>
<td>WMB-10Xv3</td>
<td>3.910954</td>
<td>22.416977</td>
<td>14939</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">CATCGGGTCAGTTTGG-033_A01</td>
<td>CATCGGGTCAGTTTGG</td>
<td>033_A01</td>
<td>L8TX_180712_01_H05</td>
<td>WMB-10Xv2-Isocortex-4</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv2</td>
<td>TEa-PERI-ECT</td>
<td>Snap25-IRES2-Cre;Ai14-395345</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>F</td>
<td>WMB-10Xv2</td>
<td>11.004389</td>
<td>16.859750</td>
<td>649</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">CGTGTAACATCGATTG-003_D01</td>
<td>CGTGTAACATCGATTG</td>
<td>003_D01</td>
<td>L8TX_180115_01_F11</td>
<td>WMB-10Xv2-Isocortex-3</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv2</td>
<td>VIS</td>
<td>Snap25-IRES2-Cre;Ai14-366678</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>M</td>
<td>WMB-10Xv2</td>
<td>18.697296</td>
<td>-6.886807</td>
<td>188</td>
</tr>
</tbody>
</table>

<p>28981 rows × 15 columns</p>
</div>
</div>
</div>
<p><font size="3"> Add feature metadata to the AnnData object</font></p><font size="3">
<div id="9f6b28a7-7c33-446a-9294-aa92289833a0" class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>genes <span class="op">=</span> pd.read_csv(<span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/abc_download_root/metadata/WMB-10X/gene.csv"</span>,index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>genes.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="58">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_symbol</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">mapped_ncbi_identifier</th>
<th data-quarto-table-cell-role="th">comment</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">gene_identifier</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000051951</td>
<td>Xkr4</td>
<td>X-linked Kx blood group related 4</td>
<td>NCBIGene:497097</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSMUSG00000089699</td>
<td>Gm1992</td>
<td>predicted gene 1992</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000102331</td>
<td>Gm19938</td>
<td>predicted gene, 19938</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">ENSMUSG00000102343</td>
<td>Gm37381</td>
<td>predicted gene, 37381</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">ENSMUSG00000025900</td>
<td>Rp1</td>
<td>retinitis pigmentosa 1 (human)</td>
<td>NCBIGene:19888</td>
<td>NaN</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="bc782a1c-66b6-4eb4-aa95-ea4a2eb6914f" class="cell" data-execution_count="59">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>adata_sc.var <span class="op">=</span> genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5aec6d0c-dfa4-4586-b887-70330b219f35" class="cell" data-execution_count="60">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reset the index to turn the current index into a column</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>adata_sc.var.reset_index(inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>adata_sc.var.set_index(<span class="st">'gene_symbol'</span>, inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>adata_sc.var.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="60">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_identifier</th>
<th data-quarto-table-cell-role="th">name</th>
<th data-quarto-table-cell-role="th">mapped_ncbi_identifier</th>
<th data-quarto-table-cell-role="th">comment</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">gene_symbol</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Xkr4</td>
<td>ENSMUSG00000051951</td>
<td>X-linked Kx blood group related 4</td>
<td>NCBIGene:497097</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Gm1992</td>
<td>ENSMUSG00000089699</td>
<td>predicted gene 1992</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Gm19938</td>
<td>ENSMUSG00000102331</td>
<td>predicted gene, 19938</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Gm37381</td>
<td>ENSMUSG00000102343</td>
<td>predicted gene, 37381</td>
<td>NaN</td>
<td>NaN</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Rp1</td>
<td>ENSMUSG00000025900</td>
<td>retinitis pigmentosa 1 (human)</td>
<td>NCBIGene:19888</td>
<td>NaN</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><font size="3"> Set up inputs for SpaGE:</font></p><font size="3">
<div id="1d55535e-9547-4d92-a173-81077340893b" class="cell" data-execution_count="61">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the count data and transform them into a Pandas DataFrame</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>RNA_data <span class="op">=</span> pd.DataFrame(adata_sc.X.todense())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="2eb576b9-394b-44c3-ad22-49a40e6c962f" class="cell" data-execution_count="62">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add gene names to the RNA_data columns </span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>RNA_data.columns <span class="op">=</span> adata_sc.var.index</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a3f266e6-82d4-4ced-8a1c-1aa5daa55ef2" class="cell" data-execution_count="63">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter lowely expressed genes</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>Genes_count <span class="op">=</span> np.<span class="bu">sum</span>(RNA_data <span class="op">&gt;</span> <span class="dv">0</span>, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>RNA_data <span class="op">=</span> RNA_data.loc[Genes_count <span class="op">&gt;=</span><span class="dv">10</span>,:]</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> Genes_count</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ecb28736-c54d-404f-b0e3-b4e09c99ac69" class="cell" data-execution_count="64">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalization</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> Log_Norm_cpm(x):</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> np.log(((x<span class="op">/</span>np.<span class="bu">sum</span>(x))<span class="op">*</span><span class="dv">1000000</span>) <span class="op">+</span> <span class="dv">1</span>)</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>RNA_data <span class="op">=</span> RNA_data.<span class="bu">apply</span>(Log_Norm_cpm,axis<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="77339674-5f9d-4ddb-b4ba-7cbf2278e417" class="cell" data-execution_count="65">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transpose RNA_data to have rows = genes and columns = cells</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>RNA_data <span class="op">=</span> RNA_data.T</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="6099b6ab-1c53-4ffa-be3c-c4cda0f5c69f" class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>RNA_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="66">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">0</th>
<th data-quarto-table-cell-role="th">1</th>
<th data-quarto-table-cell-role="th">2</th>
<th data-quarto-table-cell-role="th">3</th>
<th data-quarto-table-cell-role="th">4</th>
<th data-quarto-table-cell-role="th">5</th>
<th data-quarto-table-cell-role="th">6</th>
<th data-quarto-table-cell-role="th">7</th>
<th data-quarto-table-cell-role="th">8</th>
<th data-quarto-table-cell-role="th">9</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">28971</th>
<th data-quarto-table-cell-role="th">28972</th>
<th data-quarto-table-cell-role="th">28973</th>
<th data-quarto-table-cell-role="th">28974</th>
<th data-quarto-table-cell-role="th">28975</th>
<th data-quarto-table-cell-role="th">28976</th>
<th data-quarto-table-cell-role="th">28977</th>
<th data-quarto-table-cell-role="th">28978</th>
<th data-quarto-table-cell-role="th">28979</th>
<th data-quarto-table-cell-role="th">28980</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">gene_symbol</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Xkr4</td>
<td>4.552345</td>
<td>0.0</td>
<td>4.130508</td>
<td>0.0</td>
<td>5.504673</td>
<td>2.214984</td>
<td>2.583386</td>
<td>3.238043</td>
<td>2.214984</td>
<td>3.911375</td>
<td>...</td>
<td>0.0</td>
<td>4.673730</td>
<td>4.253778</td>
<td>2.583386</td>
<td>3.733048</td>
<td>4.673730</td>
<td>0.0</td>
<td>0.0</td>
<td>2.852008</td>
<td>2.583386</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Gm1992</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>5.087625</td>
<td>...</td>
<td>0.0</td>
<td>6.182114</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>5.087625</td>
<td>0.0</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Gm19938</td>
<td>4.665658</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.0</td>
<td>5.354088</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>5.576285</td>
<td>...</td>
<td>0.0</td>
<td>5.067980</td>
<td>3.981880</td>
<td>4.665658</td>
<td>0.000000</td>
<td>3.981880</td>
<td>0.0</td>
<td>0.0</td>
<td>0.000000</td>
<td>3.981880</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Gm37381</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>...</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Rp1</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>...</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
</tbody>
</table>

<p>5 rows × 28981 columns</p>
</div>
</div>
</div>
<div id="d0ac960e-7c2c-471d-8d80-3bf4d82eee2f" class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop any row with at least one NaN value</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>RNA_data <span class="op">=</span> RNA_data.dropna()  <span class="co"># This will drop any row with at least one NaN value</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>How many genes do the spatial and single-cell data set have in common?</strong></span></p>
<p><span style="color: red; font-size: 15px;"> <strong>How many genes are measured in the single-cell dataset that are not measured in the spatial dataset?</strong></span></p>
<div id="2846466a-d702-4e33-a9b3-8e6feafbdc5d" class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co">## EX 11</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co">## Hint: use .intersection</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>gene_intersection <span class="op">=</span> <span class="bu">set</span>(<span class="bu">list</span>(adata_section.var.index)).intersection(<span class="bu">set</span>(<span class="bu">list</span>(adata_sc.var.index)))</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a><span class="co">## Hint: use a set operation</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>genes_not_intersection <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(RNA_data.index) <span class="op">-</span> <span class="bu">set</span>(adata_section.var.index))</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of genes measured both in the sc and spatial datasets: </span><span class="sc">{</span><span class="bu">len</span>(gene_intersection)<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of genes measured in the sc reference only: </span><span class="sc">{</span><span class="bu">len</span>(genes_not_intersection)<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of genes measured both in the sc and spatial datasets: 1122
Number of genes measured in the sc reference only: 27078</code></pre>
</div>
</div>
<div id="d616ce55-41de-420b-afa3-93d50b8a461c" class="cell" data-execution_count="69">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract count matrix from adata_section and rename index and column names</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>spatial_data <span class="op">=</span> pd.DataFrame(adata_section.X)</span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>spatial_data.index <span class="op">=</span> adata_section.obs.index</span>
<span id="cb92-5"><a href="#cb92-5" aria-hidden="true" tabindex="-1"></a>spatial_data.columns <span class="op">=</span> adata_section.var.index</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="a2608515-710a-462c-8ac2-e835eb413c6d" class="cell" data-execution_count="70">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize spatial_data </span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>spatial_data <span class="op">=</span> spatial_data.<span class="bu">apply</span>(Log_Norm_cpm,axis<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3a7b24db-b2b6-4b53-861a-074d6edf7276" class="cell" data-execution_count="71">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>brain_region_specific_genes <span class="op">=</span> [</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Bcl11b"</span>, <span class="st">"Cux2"</span>, <span class="st">"Fezf2"</span>, <span class="st">"Tbr1"</span>, <span class="st">"Satb2"</span>, <span class="co"># isocortex</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Foxp2"</span>,  <span class="co"># Striatum specific</span></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Penk"</span>,   <span class="co"># Basal ganglia specific</span></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Zic1"</span>,   <span class="co"># Cerebellum specific</span></span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Hoxb2"</span>,  <span class="co"># Hindbrain specific</span></span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Dlx6"</span>,   <span class="co"># Subpallium specific</span></span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a>]</span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-10"><a href="#cb94-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Checking for genes from brain_region_specific_genes that are measured by spatial data</span></span>
<span id="cb94-11"><a href="#cb94-11" aria-hidden="true" tabindex="-1"></a>overlap_genes <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(<span class="bu">list</span>(spatial_data.columns)) <span class="op">&amp;</span> <span class="bu">set</span>(brain_region_specific_genes))</span>
<span id="cb94-12"><a href="#cb94-12" aria-hidden="true" tabindex="-1"></a>overlap_genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="71">
<pre><code>['Satb2', 'Cux2', 'Zic1', 'Penk', 'Bcl11b', 'Fezf2', 'Dlx6']</code></pre>
</div>
</div>
<div id="e6720795-f4b1-4770-b965-60e510414ca1" class="cell" data-execution_count="72">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Checking for genes from brain_region_specific_genes that are NOT measured by spatial data</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>non_overlap_genes <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>(brain_region_specific_genes) <span class="op">-</span> <span class="bu">set</span>(overlap_genes))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="leave-one-gene-out-cross-validation" class="level3">
<h3 class="anchored" data-anchor-id="leave-one-gene-out-cross-validation">Leave-one-gene-out cross validation</h3>
<p><font size="3"> Define the function to predict genes measured in the spatial dataset to have a ground truth and plot the measured expression next to the predicted one. <font size="3"> Calculate Spearman correlation between the measured and predicted expression.</font></font></p><font size="3"><font size="3">
<div id="2b94062d-7ddc-40bc-9a49-7bbd1dfd9143" class="cell" data-execution_count="73">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.gridspec <span class="im">as</span> gridspec  <span class="co"># Import for more precise subplot layout control</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>plt.style.use(<span class="st">'dark_background'</span>)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_measured_predicted(Gene_set, corr_df):</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> Gene_set:</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Predict the expression of the current gene using the SpaGE method</span></span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>        Imp_Genes <span class="op">=</span> SpaGE(spatial_data.drop(i, axis<span class="op">=</span><span class="dv">1</span>), RNA_data.T, n_pv<span class="op">=</span><span class="dv">30</span>, genes_to_predict<span class="op">=</span>[i])</span>
<span id="cb97-9"><a href="#cb97-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb97-10"><a href="#cb97-10" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate the Spearman correlation between measured and predicted values</span></span>
<span id="cb97-11"><a href="#cb97-11" aria-hidden="true" tabindex="-1"></a>        corr_df[i] <span class="op">=</span> st.spearmanr(spatial_data[i], Imp_Genes[i])[<span class="dv">0</span>]</span>
<span id="cb97-12"><a href="#cb97-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb97-13"><a href="#cb97-13" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create a figure with two subplots (1 row, 2 columns)</span></span>
<span id="cb97-14"><a href="#cb97-14" aria-hidden="true" tabindex="-1"></a>        fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">8</span>))  <span class="co"># Adjust figure size if needed</span></span>
<span id="cb97-15"><a href="#cb97-15" aria-hidden="true" tabindex="-1"></a>        gs <span class="op">=</span> gridspec.GridSpec(<span class="dv">1</span>, <span class="dv">3</span>, width_ratios<span class="op">=</span>[<span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">0.05</span>], wspace<span class="op">=</span><span class="fl">0.3</span>)  <span class="co"># Create a gridspec for better control</span></span>
<span id="cb97-16"><a href="#cb97-16" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb97-17"><a href="#cb97-17" aria-hidden="true" tabindex="-1"></a>        <span class="co"># --- Measured Gene Expression Plot ---</span></span>
<span id="cb97-18"><a href="#cb97-18" aria-hidden="true" tabindex="-1"></a>        ax0 <span class="op">=</span> fig.add_subplot(gs[<span class="dv">0</span>])  <span class="co"># First plot</span></span>
<span id="cb97-19"><a href="#cb97-19" aria-hidden="true" tabindex="-1"></a>        cmap_measured <span class="op">=</span> spatial_data[i]</span>
<span id="cb97-20"><a href="#cb97-20" aria-hidden="true" tabindex="-1"></a>        cmap_measured[cmap_measured <span class="op">&gt;</span> np.percentile(cmap_measured, <span class="dv">99</span>)] <span class="op">=</span> np.percentile(cmap_measured, <span class="dv">99</span>)</span>
<span id="cb97-21"><a href="#cb97-21" aria-hidden="true" tabindex="-1"></a>        sc1 <span class="op">=</span> ax0.scatter(adata_section.obs[<span class="st">'x'</span>], <span class="op">-</span>adata_section.obs[<span class="st">'y'</span>], s<span class="op">=</span><span class="dv">1</span>, c<span class="op">=</span>cmap_measured, alpha<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb97-22"><a href="#cb97-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb97-23"><a href="#cb97-23" aria-hidden="true" tabindex="-1"></a>        ax0.set_title(<span class="ss">f'Measured Expression: </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb97-24"><a href="#cb97-24" aria-hidden="true" tabindex="-1"></a>        ax0.set_xlabel(<span class="st">'X coordinate'</span>)</span>
<span id="cb97-25"><a href="#cb97-25" aria-hidden="true" tabindex="-1"></a>        ax0.set_ylabel(<span class="st">'Y coordinate'</span>)</span>
<span id="cb97-26"><a href="#cb97-26" aria-hidden="true" tabindex="-1"></a>        ax0.set_aspect(<span class="st">'equal'</span>, adjustable<span class="op">=</span><span class="st">'box'</span>)</span>
<span id="cb97-27"><a href="#cb97-27" aria-hidden="true" tabindex="-1"></a>        ax0.grid(<span class="va">False</span>)</span>
<span id="cb97-28"><a href="#cb97-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb97-29"><a href="#cb97-29" aria-hidden="true" tabindex="-1"></a>        <span class="co"># --- Predicted Gene Expression Plot ---</span></span>
<span id="cb97-30"><a href="#cb97-30" aria-hidden="true" tabindex="-1"></a>        ax1 <span class="op">=</span> fig.add_subplot(gs[<span class="dv">1</span>])  <span class="co"># Second plot</span></span>
<span id="cb97-31"><a href="#cb97-31" aria-hidden="true" tabindex="-1"></a>        cmap_predicted <span class="op">=</span> Imp_Genes[i]</span>
<span id="cb97-32"><a href="#cb97-32" aria-hidden="true" tabindex="-1"></a>        cmap_predicted[cmap_predicted <span class="op">&gt;</span> np.percentile(cmap_predicted, <span class="dv">99</span>)] <span class="op">=</span> np.percentile(cmap_predicted, <span class="dv">99</span>)</span>
<span id="cb97-33"><a href="#cb97-33" aria-hidden="true" tabindex="-1"></a>        sc2 <span class="op">=</span> ax1.scatter(adata_section.obs[<span class="st">'x'</span>], <span class="op">-</span>adata_section.obs[<span class="st">'y'</span>], s<span class="op">=</span><span class="dv">1</span>, c<span class="op">=</span>cmap_predicted, alpha<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb97-34"><a href="#cb97-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb97-35"><a href="#cb97-35" aria-hidden="true" tabindex="-1"></a>        ax1.set_title(<span class="ss">f'Predicted Expression: </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb97-36"><a href="#cb97-36" aria-hidden="true" tabindex="-1"></a>        ax1.set_xlabel(<span class="st">'X coordinate'</span>)</span>
<span id="cb97-37"><a href="#cb97-37" aria-hidden="true" tabindex="-1"></a>        ax1.set_ylabel(<span class="st">'Y coordinate'</span>)</span>
<span id="cb97-38"><a href="#cb97-38" aria-hidden="true" tabindex="-1"></a>        ax1.set_aspect(<span class="st">'equal'</span>, adjustable<span class="op">=</span><span class="st">'box'</span>)</span>
<span id="cb97-39"><a href="#cb97-39" aria-hidden="true" tabindex="-1"></a>        ax1.grid(<span class="va">False</span>)</span>
<span id="cb97-40"><a href="#cb97-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-41"><a href="#cb97-41" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create a colorbar on the far right</span></span>
<span id="cb97-42"><a href="#cb97-42" aria-hidden="true" tabindex="-1"></a>        cbar_ax <span class="op">=</span> fig.add_subplot(gs[<span class="dv">2</span>])  <span class="co"># Third subplot (narrow space for the color bar)</span></span>
<span id="cb97-43"><a href="#cb97-43" aria-hidden="true" tabindex="-1"></a>        cbar <span class="op">=</span> fig.colorbar(sc1, cax<span class="op">=</span>cbar_ax)  <span class="co"># Assign the color bar to the third subplot</span></span>
<span id="cb97-44"><a href="#cb97-44" aria-hidden="true" tabindex="-1"></a>        cbar.set_label(i)</span>
<span id="cb97-45"><a href="#cb97-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-46"><a href="#cb97-46" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Show the plot</span></span>
<span id="cb97-47"><a href="#cb97-47" aria-hidden="true" tabindex="-1"></a>        plt.tight_layout()  <span class="co"># Adjust spacing to avoid overlap</span></span>
<span id="cb97-48"><a href="#cb97-48" aria-hidden="true" tabindex="-1"></a>        plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="95639806-6b66-4151-b829-de3c5a400507" class="cell" data-execution_count="74">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Since we have a ground truth, create a Series to store Spearman correlation results for each gene</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Our gene_set in this case is a set of genes shared between the spatial and the sc datasets</span></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>correlations_spage <span class="op">=</span> pd.Series(index<span class="op">=</span>overlap_genes)</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>plot_measured_predicted(overlap_genes,correlations_spage)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-1.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-2.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-3.png" width="1040" height="517" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-4.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-5.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-6.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-75-output-7.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><font size="3"> Let’s take a look at the Spearman Correlations between measured and predicted gene expression:</font></p><font size="3">
<div id="49dafbec-7b73-4ffc-ad3d-eb1576393a3b" class="cell" data-execution_count="75">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>correlations_spage <span class="op">=</span> pd.DataFrame(correlations_spage, columns<span class="op">=</span>[<span class="st">"correlations_spage"</span>])</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>correlations_spage</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="75">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">correlations_spage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Satb2</td>
<td>0.569667</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Cux2</td>
<td>0.624125</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Zic1</td>
<td>0.560184</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Penk</td>
<td>0.462083</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Bcl11b</td>
<td>0.476352</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Fezf2</td>
<td>0.428165</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Dlx6</td>
<td>0.401532</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>Is there any other gene expression (measured in the spatial dataset) you would like to predict?</strong></span></p>
<div id="c9f32ba9-5e68-4492-9b57-7f7b1af954fb" class="cell" data-execution_count="76">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="bu">list</span>(gene_intersection)[:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="76">
<pre><code>['Adgrf5',
 'Epha1',
 'Ptpru',
 'Scnn1a',
 'Adam18',
 'Prdm8',
 'Hgf',
 'Hoxb4',
 'Ptpn14',
 'Chrna4']</code></pre>
</div>
</div>
<div id="6f894fac-7ead-4104-95d7-6357cf2d1fe9" class="cell" data-execution_count="77">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the set of genes (from the gene_intersection list) where we will perform leave-one-out cross-validation</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>Gene_set <span class="op">=</span> [<span class="st">'Grm3'</span>] <span class="co"># has to be a list</span></span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Since we have a ground truth, create a Series to store Spearman correlation results for each gene</span></span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>correlations_spage_extragenes <span class="op">=</span> pd.Series(index<span class="op">=</span>Gene_set)</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>plot_measured_predicted(Gene_set,correlations_spage_extragenes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-78-output-1.png" width="1040" height="511" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="c389debc-7913-4833-b72a-fb7717edff50" class="cell" data-execution_count="78">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="co"># print Correlations</span></span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>correlations_spage_extragenes <span class="op">=</span> pd.DataFrame(correlations_spage_extragenes, columns<span class="op">=</span>[<span class="st">"correlations_spage"</span>])</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>correlations_spage_extragenes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="78">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">correlations_spage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Grm3</td>
<td>0.637015</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="predict-unmeasured-genes" class="level3">
<h3 class="anchored" data-anchor-id="predict-unmeasured-genes">Predict unmeasured genes</h3>
<p><font size="3"> Define a function to predict unmeasured genes in the spatial dataset.</font></p><font size="3">
<p><font size="3"> In this case correlations cannot be calculated as there is no ground truth available.</font></p><font size="3">
<div id="c5b9df78-120c-43d3-b850-c77537271492" class="cell" data-execution_count="79">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> predict_unmeasured_genes(Gene_set):</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i <span class="kw">in</span> Gene_set:</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Predict the expression of the current gene using the SpaGE method</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>        Imp_Genes <span class="op">=</span> SpaGE(spatial_data, RNA_data.T, n_pv<span class="op">=</span><span class="dv">30</span>, genes_to_predict<span class="op">=</span>[i])</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create a new figure for each gene's predicted expression</span></span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>        plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Get predicted gene expression values and clip both lower and upper bounds</span></span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a>        cmap_predicted <span class="op">=</span> Imp_Genes[i]</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Clip the minimum values at 0 (no negatives) and limit extreme values at 99th percentile</span></span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>        cmap_predicted <span class="op">=</span> np.clip(cmap_predicted, <span class="dv">0</span>, np.percentile(cmap_predicted, <span class="dv">99</span>))</span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-16"><a href="#cb104-16" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Plot scatter plot for the predicted gene expression</span></span>
<span id="cb104-17"><a href="#cb104-17" aria-hidden="true" tabindex="-1"></a>        sc2 <span class="op">=</span> plt.scatter(adata_section.obs[<span class="st">'x'</span>], <span class="op">-</span>adata_section.obs[<span class="st">'y'</span>], s<span class="op">=</span><span class="dv">1</span>, c<span class="op">=</span>cmap_predicted, alpha<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb104-18"><a href="#cb104-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-19"><a href="#cb104-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Set plot title and labels</span></span>
<span id="cb104-20"><a href="#cb104-20" aria-hidden="true" tabindex="-1"></a>        plt.title(<span class="ss">f'Predicted Expression: </span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb104-21"><a href="#cb104-21" aria-hidden="true" tabindex="-1"></a>        plt.xlabel(<span class="st">'X coordinate'</span>)</span>
<span id="cb104-22"><a href="#cb104-22" aria-hidden="true" tabindex="-1"></a>        plt.ylabel(<span class="st">'Y coordinate'</span>)</span>
<span id="cb104-23"><a href="#cb104-23" aria-hidden="true" tabindex="-1"></a>        plt.gca().set_aspect(<span class="st">'equal'</span>, adjustable<span class="op">=</span><span class="st">'box'</span>)</span>
<span id="cb104-24"><a href="#cb104-24" aria-hidden="true" tabindex="-1"></a>        plt.grid(<span class="va">False</span>)</span>
<span id="cb104-25"><a href="#cb104-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-26"><a href="#cb104-26" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create a colorbar on the same plot</span></span>
<span id="cb104-27"><a href="#cb104-27" aria-hidden="true" tabindex="-1"></a>        cbar <span class="op">=</span> plt.colorbar(sc2)</span>
<span id="cb104-28"><a href="#cb104-28" aria-hidden="true" tabindex="-1"></a>        cbar.set_label(i)</span>
<span id="cb104-29"><a href="#cb104-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb104-30"><a href="#cb104-30" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Adjust layout and display the plot</span></span>
<span id="cb104-31"><a href="#cb104-31" aria-hidden="true" tabindex="-1"></a>        plt.tight_layout()</span>
<span id="cb104-32"><a href="#cb104-32" aria-hidden="true" tabindex="-1"></a>        plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ab92cc4a-9eb4-4a73-b54c-cb8cb86edd2e" class="cell" data-execution_count="80">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define the set of genes (from the non_overlap_genes list) to be predicted by SpaGE</span></span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>predict_unmeasured_genes(non_overlap_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-81-output-1.png" width="683" height="622" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-81-output-2.png" width="668" height="622" class="figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-81-output-3.png" width="668" height="622" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><font size="3"> There is no definitive ground truth for the real gene expression profiles of these genes. However, an alternative approach could be to plot region-specific genes and verify whether their expression aligns with the expected brain regions.</font></p><font size="3">
<p><font size="3"> For example, Tbr1 is known to be Isocortex-specific; does its expression show higher levels in the Isocortex? Similarly, Foxp2 is Striatum-specific; does it exhibit a Striatum-specific expression pattern?</font></p><font size="3">
<p><span style="color: red; font-size: 15px;"> <strong>Use <a href="https://mouse.brain-map.org/search/index">this</a> resource to check region-specific gene expression.</strong></span></p>
<p><span style="color: red; font-size: 15px;"> <strong>Is there any other gene expression (NOT measured in the spatial dataset) you would like to predict?</strong></span></p>
<p><span style="color: red; font-size: 15px;"> <strong>Can you think of other genes that we could use to visually validate the accuracy of SpaGE by checking if their expression patterns correspond to their known brain regions? Plot them!</strong></span></p>
<div id="6d62f84d-1144-44f7-8701-55fe5a99bd09" class="cell" data-execution_count="81">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="co">#list(genes_not_intersection)[:10]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="5789f404-810a-4623-8f8f-c0d9b8c5be46" class="cell" data-execution_count="82">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Gene_set = [...]</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="co">#predict_unmeasured_genes(Gene_set)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="371ac2df-aef5-493a-8ff5-38fb8aa29564" class="cell" data-execution_count="83">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Record the end time</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>end_time <span class="op">=</span> time.time()</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate elapsed time in minutes</span></span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>elapsed_time <span class="op">=</span> (end_time <span class="op">-</span> start_time) <span class="op">/</span> <span class="dv">60</span></span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Execution Time SpaGE: </span><span class="sc">{</span>elapsed_time<span class="sc">:.2f}</span><span class="ss"> minutes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Execution Time SpaGE: 12.04 minutes</code></pre>
</div>
</div>
<p><font size="3"> Let’s delete some objects we won’t be using anymore to free some RAM:</font></p><font size="3">
<div id="3e35c884-a934-4070-a0cd-c5d556db8d33" class="cell" data-execution_count="84">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> RNA_data, spatial_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</font></font></font></font></font></section><font size="3"><font size="3"><font size="3">
</font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="unmeasured-gene-imputation-with-tangram" class="level1">
<h1>Unmeasured gene imputation with Tangram</h1>
<div id="518b2c69-c6c0-4ba9-ada7-f7a1709c594b" class="cell" data-execution_count="85">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>start_time <span class="op">=</span> time.time()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="3000981c-012e-404e-b06e-409e7eaaf9cc" class="cell" data-execution_count="86">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tangram <span class="im">as</span> tg</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> MinMaxScaler</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb112-7"><a href="#cb112-7" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>, facecolor<span class="op">=</span><span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="common-gene-set-between-reference-and-spatial-dataset" class="level4">
<h4 class="anchored" data-anchor-id="common-gene-set-between-reference-and-spatial-dataset">Common gene set between reference and spatial dataset</h4>
<p><font size="3"> We want to select our training genes. These genes are shared between the two datasets and should capture the biological variance between cell types. For this, we first compute marker genes on the single-cell data and then use Tangram’s preprocessing function to subset to those genes that are also present in the spatial data.</font></p><font size="3">
<div id="674dbcef-8913-42f0-812a-50d342d679f7" class="cell" data-execution_count="87">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> <span class="bu">list</span>(<span class="bu">set</span>.intersection(<span class="bu">set</span>(adata_sc.var_names), <span class="bu">set</span>(adata_section.var_names)))</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(markers)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="87">
<pre><code>1122</code></pre>
</div>
</div>
<p><font size="3"> <code>tg.pp_adatas</code> computes the overlap between single-cell data and spatial data on the list of genes provided in the genes argument, it then stores the resulting gene set under ‘training_genes’ in both adata objects under the <code>.uns</code> key and enforces consistent ordering of the genes.</font></p><font size="3">
<div id="d12abfa2-3533-4ac3-a16a-55357663742a" class="cell" data-execution_count="88">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>tg.pp_adatas(adata_sc, adata_section, genes<span class="op">=</span>markers) <span class="co"># using the same single-cell reference that we used to run SpaGE</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>filtered out 4067 genes that are detected in less than 1 cells</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO:root:1122 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.
INFO:root:1122 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.
INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.
INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.</code></pre>
</div>
</div>
<p><font size="3"> Check that the function performs as we expect:</font></p><font size="3">
<div id="2ed67270-9beb-48df-8c48-93a59ae24254" class="cell" data-execution_count="89">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="st">"training_genes"</span> <span class="kw">in</span> adata_sc.uns</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a><span class="cf">assert</span> <span class="st">"training_genes"</span> <span class="kw">in</span> adata_section.uns</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb118-4"><a href="#cb118-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Number of training_genes: </span><span class="sc">{</span><span class="bu">len</span>(adata_sc.uns[<span class="st">'training_genes'</span>])<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of training_genes: 1122</code></pre>
</div>
</div>
</font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="computing-the-map-from-single-cells-to-spatial-voxels" class="level4">
<h4 class="anchored" data-anchor-id="computing-the-map-from-single-cells-to-spatial-voxels">Computing the map from single-cells to spatial voxels</h4>
<p><font size="3"> Having specified the training genes, we can now create the map from dissociated single-cell measurements to the spatial locations. For this, we are going to use the <code>map_cells_to_space</code> function from tangram. This function has two different modes, <code>mode='cells'</code> and <code>mode='clusters'</code>. The latter only maps averaged single-cells which makes the mapping computationally faster and more robust when mapping between specimen.</font></p><font size="3">
<div class="sourceCode" id="cb120"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>ad_map <span class="op">=</span> tg.map_cells_to_space(</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>    adata_sc,</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>    adata_section,</span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a>    mode<span class="op">=</span><span class="st">"cells"</span>,</span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a>    density_prior<span class="op">=</span><span class="st">"rna_count_based"</span>,</span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a>    num_epochs<span class="op">=</span><span class="dv">500</span>,</span>
<span id="cb120-7"><a href="#cb120-7" aria-hidden="true" tabindex="-1"></a>    device<span class="op">=</span><span class="st">"cpu"</span>,  <span class="co"># or: cpu</span></span>
<span id="cb120-8"><a href="#cb120-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb120-9"><a href="#cb120-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-10"><a href="#cb120-10" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;</span> I run `tg.map_cells_to_space` <span class="cf">for</span> you (takes a <span class="cf">while</span> to run), let<span class="st">'s just read the resulting AnnData object</span></span>
<span id="cb120-11"><a href="#cb120-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-12"><a href="#cb120-12" aria-hidden="true" tabindex="-1"></a><span class="er">::: {#8849a82d-8176-4cff-93e0-ab07d8e14cda .cell execution_count=90}</span></span>
<span id="cb120-13"><a href="#cb120-13" aria-hidden="true" tabindex="-1"></a>``` {.python .cell<span class="op">-</span>code}</span>
<span id="cb120-14"><a href="#cb120-14" aria-hidden="true" tabindex="-1"></a>ad_map <span class="op">=</span> ad.read_h5ad(<span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/Tangram_data/ad_map_tangram.h5ad"</span>)</span>
<span id="cb120-15"><a href="#cb120-15" aria-hidden="true" tabindex="-1"></a>ad_map</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="90">
<pre><code>AnnData object with n_obs × n_vars = 28981 × 37068
    obs: 'cell_barcode', 'barcoded_cell_sample_label', 'library_label', 'feature_matrix_label', 'entity', 'brain_section_label', 'library_method', 'region_of_interest_acronym', 'donor_label', 'donor_genotype', 'donor_sex', 'dataset_label', 'x', 'y', 'cluster_alias'
    var: 'brain_section_label', 'feature_matrix_label', 'donor_label', 'donor_genotype', 'donor_sex', 'cluster_alias', 'x', 'y', 'z', 'subclass_confidence_score', 'cluster_confidence_score', 'high_quality_transfer', 'neurotransmitter', 'class', 'subclass', 'supertype', 'cluster', 'neurotransmitter_color', 'class_color', 'subclass_color', 'supertype_color', 'cluster_color', 'uniform_density', 'rna_count_based_density'
    uns: 'train_genes_df', 'training_history'</code></pre>
</div>
<p>:::</p>
<p><font size="3"> We observe that Tangram’s mapping from cell i to spatial voxels j is stored in the .X property of ad_map. <font size="3"> The meaning of the .var and .obs also changes:</font></font></p><font size="3"><font size="3">
<ul>
<li><font size="3"> in .var we have the available metadata of the spatial data, adata_section</font></li><font size="3">
<li><font size="3"> in .obs we have the available metadata of the single-cell data, adata_sc</font></li><font size="3">
<li><font size="3"> In addition, the information about the training run is stored in the .uns key, see <code>.uns['training_genes_df']</code> and <code>.uns[training_history]</code>.</font></li><font size="3">
</font></font></font></ul><font size="3"><font size="3"><font size="3">
<p><font size="3"> Now we can <strong>project the genes present in the single-cell data to the spatial locations</strong>. This is easily achieved by multiplying the mapping matrix stored in ad_map with the original single-cell data stored in adata_sc. Tangram already provides a convenience function which takes in a mapping and the corresponding single-cell data. The result is a <strong>spatial voxel by genes matrix</strong> which technically is identical to the original spatial data adata_section but <strong>contains expression values for all genes</strong>.</font></p><font size="3">
<div class="sourceCode" id="cb122"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>ad_ge <span class="op">=</span> tg.project_genes(adata_map<span class="op">=</span>ad_map, adata_sc<span class="op">=</span>adata_sc)</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>ad_ge</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;</span> I run `tg.project_genes` <span class="cf">for</span> you, let<span class="st">'s just read the resulting AnnData object</span></span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a><span class="er">::: {#62f70844-0288-433e-879d-70438a1ae988 .cell execution_count=91}</span></span>
<span id="cb122-7"><a href="#cb122-7" aria-hidden="true" tabindex="-1"></a>``` {.python .cell<span class="op">-</span>code}</span>
<span id="cb122-8"><a href="#cb122-8" aria-hidden="true" tabindex="-1"></a>ad_ge <span class="op">=</span> ad.read_h5ad(<span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/day3/Tangram_data/ad_ge.h5ad"</span>)</span>
<span id="cb122-9"><a href="#cb122-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-10"><a href="#cb122-10" aria-hidden="true" tabindex="-1"></a><span class="co">## add .obsm['X_spatial'] so we can use sc.pl.spatial to plot gene expression</span></span>
<span id="cb122-11"><a href="#cb122-11" aria-hidden="true" tabindex="-1"></a>ad_ge.obsm[<span class="st">'X_spatial'</span>] <span class="op">=</span> np.array(ad_ge.obs[[<span class="st">'x'</span>,<span class="st">'y'</span>]])</span>
<span id="cb122-12"><a href="#cb122-12" aria-hidden="true" tabindex="-1"></a>ad_ge</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="91">
<pre><code>AnnData object with n_obs × n_vars = 37068 × 28218
    obs: 'brain_section_label', 'feature_matrix_label', 'donor_label', 'donor_genotype', 'donor_sex', 'cluster_alias', 'x', 'y', 'z', 'subclass_confidence_score', 'cluster_confidence_score', 'high_quality_transfer', 'neurotransmitter', 'class', 'subclass', 'supertype', 'cluster', 'neurotransmitter_color', 'class_color', 'subclass_color', 'supertype_color', 'cluster_color', 'uniform_density', 'rna_count_based_density'
    var: '_index', 'comment', 'gene_identifier', 'mapped_ncbi_identifier', 'name'
    obsm: 'X_spatial'</code></pre>
</div>
<p>:::</p>
<p><font size="3"> Next, we will compare the new spatial data with the original measurements.</font></p><font size="3">
<p><font size="3"> To do this we will plot the true expression and the predicted expression of the shared genes, as well as the Spearman correlation between the true and predicted expression in all the spots.</font></p><font size="3">
<div id="bd04efe5-5689-409b-bd10-02df368a2563" class="cell" data-execution_count="92">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the same set of genes that we predicted with SpaGE</span></span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a><span class="co"># we need to make the genes all lower-case first</span></span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>overlap_genes <span class="op">=</span> [gene.lower() <span class="cf">for</span> gene <span class="kw">in</span> overlap_genes]</span>
<span id="cb124-4"><a href="#cb124-4" aria-hidden="true" tabindex="-1"></a>overlap_genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="92">
<pre><code>['satb2', 'cux2', 'zic1', 'penk', 'bcl11b', 'fezf2', 'dlx6']</code></pre>
</div>
</div>
<p><font size="3"> Plot the true expression from adata_section</font></p><font size="3">
<div id="ed36a6d2-750f-42af-ae30-b57791bca2b2" class="cell" data-execution_count="93">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata_section, color<span class="op">=</span>overlap_genes, spot_size<span class="op">=</span><span class="fl">0.04</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-94-output-1.png" width="1262" height="579" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><font size="3"> Plot the predicted expression from ad_ge</font></p><font size="3">
<div id="75ac35b9-d0e5-4097-a22f-3b40164cf025" class="cell" data-execution_count="94">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>ad_ge.var.reset_index(inplace<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>ad_ge.var.set_index(<span class="st">'_index'</span>, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="ddf2247b-bdaf-428b-8b18-bfc2d54f77b7" class="cell" data-execution_count="95">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(ad_ge, color<span class="op">=</span>overlap_genes, spot_size<span class="op">=</span><span class="fl">0.04</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-96-output-1.png" width="1252" height="579" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p><font size="3"> Calculate Spearman correlations</font></p><font size="3">
<div id="55bdb832-b668-46ff-8f2a-fe7b87bb0325" class="cell" data-execution_count="96">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the AnnData objects to include only the genes in overlap_genes</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Assuming both AnnData objects have the same gene ordering</span></span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>ad_true  <span class="op">=</span> adata_section[:, adata_section.var_names.isin(overlap_genes)]</span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>ad_ge_predicted <span class="op">=</span> ad_ge[:, ad_ge.var_names.isin(overlap_genes)]</span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-6"><a href="#cb129-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize an empty dictionary to store correlations</span></span>
<span id="cb129-7"><a href="#cb129-7" aria-hidden="true" tabindex="-1"></a>correlation_dict <span class="op">=</span> {}</span>
<span id="cb129-8"><a href="#cb129-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-9"><a href="#cb129-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over each gene to calculate Spearman correlation</span></span>
<span id="cb129-10"><a href="#cb129-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> gene <span class="kw">in</span> overlap_genes:</span>
<span id="cb129-11"><a href="#cb129-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Extract gene expression vectors for the current gene</span></span>
<span id="cb129-12"><a href="#cb129-12" aria-hidden="true" tabindex="-1"></a>    expr1 <span class="op">=</span> ad_true[:, gene].X.flatten()</span>
<span id="cb129-13"><a href="#cb129-13" aria-hidden="true" tabindex="-1"></a>    expr2 <span class="op">=</span> ad_ge_predicted[:, gene].X.flatten()</span>
<span id="cb129-14"><a href="#cb129-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb129-15"><a href="#cb129-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate Spearman correlation for the current gene</span></span>
<span id="cb129-16"><a href="#cb129-16" aria-hidden="true" tabindex="-1"></a>    corr, _ <span class="op">=</span> st.spearmanr(expr1, expr2)</span>
<span id="cb129-17"><a href="#cb129-17" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb129-18"><a href="#cb129-18" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Store the result in the dictionary</span></span>
<span id="cb129-19"><a href="#cb129-19" aria-hidden="true" tabindex="-1"></a>    correlation_dict[gene] <span class="op">=</span> corr</span>
<span id="cb129-20"><a href="#cb129-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb129-21"><a href="#cb129-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the dictionary to a DataFrame</span></span>
<span id="cb129-22"><a href="#cb129-22" aria-hidden="true" tabindex="-1"></a>tangram_correlations <span class="op">=</span> pd.DataFrame.from_dict(correlation_dict, orient<span class="op">=</span><span class="st">'index'</span>, columns<span class="op">=</span>[<span class="st">'correlations_tangram'</span>])</span>
<span id="cb129-23"><a href="#cb129-23" aria-hidden="true" tabindex="-1"></a><span class="co"># capitalize gene names! </span></span>
<span id="cb129-24"><a href="#cb129-24" aria-hidden="true" tabindex="-1"></a>tangram_correlations.index <span class="op">=</span> tangram_correlations.index.<span class="bu">str</span>.capitalize()</span>
<span id="cb129-25"><a href="#cb129-25" aria-hidden="true" tabindex="-1"></a>tangram_correlations</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="96">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">correlations_tangram</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Satb2</td>
<td>0.521339</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Cux2</td>
<td>0.520271</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Zic1</td>
<td>0.438475</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Penk</td>
<td>0.404159</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Bcl11b</td>
<td>0.465106</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Fezf2</td>
<td>0.363526</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Dlx6</td>
<td>0.367113</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><font size="3"> Since we predicted the same genes as SpaGE, let’s compare the Spearman correlations:</font></p><font size="3">
<div id="3def41c2-1a5e-48a6-95af-9a97bb26ddd3" class="cell" data-execution_count="97">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>all_correlations <span class="op">=</span> pd.concat([correlations_spage, tangram_correlations],axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>all_correlations[[<span class="st">'correlations_spage'</span>,<span class="st">'correlations_tangram'</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="97">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">correlations_spage</th>
<th data-quarto-table-cell-role="th">correlations_tangram</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Satb2</td>
<td>0.569667</td>
<td>0.521339</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Cux2</td>
<td>0.624125</td>
<td>0.520271</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Zic1</td>
<td>0.560184</td>
<td>0.438475</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Penk</td>
<td>0.462083</td>
<td>0.404159</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Bcl11b</td>
<td>0.476352</td>
<td>0.465106</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Fezf2</td>
<td>0.428165</td>
<td>0.363526</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">Dlx6</td>
<td>0.401532</td>
<td>0.367113</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>Which tool results in higher correlation between the true-measured gene expression and the predicted expression of the genes of interest?</strong></span></p>
<div id="db86a1c6-50f9-4762-9214-50e9df47082e" class="cell" data-execution_count="98">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Record the end time</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>end_time <span class="op">=</span> time.time()</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate elapsed time in minutes</span></span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>elapsed_time <span class="op">=</span> (end_time <span class="op">-</span> start_time) <span class="op">/</span> <span class="dv">60</span></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Execution Time Tangram: </span><span class="sc">{</span>elapsed_time<span class="sc">:.2f}</span><span class="ss"> minutes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Execution Time Tangram: 2.20 minutes</code></pre>
</div>
</div>
<p><font size="3"> Let’s delete some objects we won’t be using anymore to free some RAM:</font></p><font size="3">
<div id="3ad5b9bb-74e1-40d6-9114-6f89aa370a97" class="cell" data-execution_count="99">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> ad_map, ad_ge</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</font></font></font></font></font></font></font></font></font></font></font></font></font></font></section><font size="3"><font size="3"><font size="3">
</font></font></font></font></font></font></section><font size="3"><font size="3"><font size="3">
<section id="cell-type-deconvolution-with-cell-2-location" class="level1">
<h1>Cell type deconvolution with cell 2 location</h1>
<div id="9dcf3b2c-3b5c-4483-8e45-04e5ef8510f2" class="cell" data-execution_count="101">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>start_time <span class="op">=</span> time.time()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="36bc28b5-9bb2-4fb2-b67d-56b009b92d41" class="cell" data-execution_count="102">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> squidpy <span class="im">as</span> sq</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> cell2location <span class="im">as</span> c2l</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>sc.settings.verbosity <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">80</span>, facecolor<span class="op">=</span><span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e41ce22b-0a76-4aa9-b8c0-417d2ddc210e" class="cell" data-execution_count="103">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set raw layer as X</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>adata_sc.layers[<span class="st">'raw'</span>] <span class="op">=</span> adata_sc.X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><font size="3"> Download single-cell metadata with cell-type annotation with this command. the dataframe is already downloaded and can be read as csv</font></p><font size="3">
<div class="sourceCode" id="cb137"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>cell <span class="op">=</span> abc_cache.get_metadata_dataframe(directory<span class="op">=</span><span class="st">'WMB-10X'</span>, file_name<span class="op">=</span><span class="st">'cell_metadata'</span>).set_index(<span class="st">'cell_label'</span>)</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a><span class="op">&lt;</span>font size<span class="op">=</span><span class="st">"3"</span><span class="op">&gt;</span> Add metadata <span class="cf">with</span> cell <span class="bu">type</span> annotation to the single<span class="op">-</span>cell AnnData <span class="bu">object</span></span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>::: {<span class="co">#5ebc5d6d-8bf3-4dd7-965f-48abeefd0fdd .cell execution_count=104}</span></span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a>``` {.python .cell<span class="op">-</span>code}</span>
<span id="cb137-7"><a href="#cb137-7" aria-hidden="true" tabindex="-1"></a><span class="co"># read csv</span></span>
<span id="cb137-8"><a href="#cb137-8" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> pd.read_csv(<span class="ss">f"</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/cell_metadata_with_cluster_annotation.csv"</span>,index_col<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb137-9"><a href="#cb137-9" aria-hidden="true" tabindex="-1"></a><span class="co"># filter metadata to keep only obervations in the subsetted SC AnnData object</span></span>
<span id="cb137-10"><a href="#cb137-10" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> metadata.loc[metadata.index.intersection(adata_sc.obs.index)]</span>
<span id="cb137-11"><a href="#cb137-11" aria-hidden="true" tabindex="-1"></a><span class="co"># reindex rows</span></span>
<span id="cb137-12"><a href="#cb137-12" aria-hidden="true" tabindex="-1"></a>metadata <span class="op">=</span> metadata.reindex(adata_sc.obs.index)</span>
<span id="cb137-13"><a href="#cb137-13" aria-hidden="true" tabindex="-1"></a><span class="co"># assign metadata to adata_sc.obs (in this way metadata replaces all fields previously in obs)</span></span>
<span id="cb137-14"><a href="#cb137-14" aria-hidden="true" tabindex="-1"></a>adata_sc.obs <span class="op">=</span> metadata</span>
<span id="cb137-15"><a href="#cb137-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-16"><a href="#cb137-16" aria-hidden="true" tabindex="-1"></a>adata_sc.obs.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="104">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">cell_barcode</th>
<th data-quarto-table-cell-role="th">barcoded_cell_sample_label</th>
<th data-quarto-table-cell-role="th">library_label</th>
<th data-quarto-table-cell-role="th">feature_matrix_label</th>
<th data-quarto-table-cell-role="th">entity</th>
<th data-quarto-table-cell-role="th">brain_section_label</th>
<th data-quarto-table-cell-role="th">library_method</th>
<th data-quarto-table-cell-role="th">region_of_interest_acronym</th>
<th data-quarto-table-cell-role="th">donor_label</th>
<th data-quarto-table-cell-role="th">donor_genotype</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">subclass</th>
<th data-quarto-table-cell-role="th">supertype</th>
<th data-quarto-table-cell-role="th">cluster</th>
<th data-quarto-table-cell-role="th">neurotransmitter_color</th>
<th data-quarto-table-cell-role="th">class_color</th>
<th data-quarto-table-cell-role="th">subclass_color</th>
<th data-quarto-table-cell-role="th">supertype_color</th>
<th data-quarto-table-cell-role="th">cluster_color</th>
<th data-quarto-table-cell-role="th">region_of_interest_order</th>
<th data-quarto-table-cell-role="th">region_of_interest_color</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">cell_label</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">GTATTCTCAGCTATTG-045_C01</td>
<td>GTATTCTCAGCTATTG</td>
<td>045_C01</td>
<td>L8TX_180829_01_C10</td>
<td>WMB-10Xv2-TH</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv2</td>
<td>TH</td>
<td>Snap25-IRES2-Cre;Ai14-407905</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>...</td>
<td>151 TH Prkcd Grin2c Glut</td>
<td>0655 TH Prkcd Grin2c Glut_2</td>
<td>2650 TH Prkcd Grin2c Glut_2</td>
<td>#2B93DF</td>
<td>#0D47A1</td>
<td>#68e497</td>
<td>#0F6652</td>
<td>#FFB599</td>
<td>23</td>
<td>#FF00EF</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">GTGTGATCACCAGCGT-468_A02</td>
<td>GTGTGATCACCAGCGT</td>
<td>468_A02</td>
<td>L8TX_201217_01_D07</td>
<td>WMB-10Xv3-MB</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>MB</td>
<td>Snap25-IRES2-Cre;Ai14-556231</td>
<td>Ai14(RCL-tdT)/wt</td>
<td>...</td>
<td>318 Astro-NT NN</td>
<td>1160 Astro-NT NN_2</td>
<td>5214 Astro-NT NN_2</td>
<td>#666666</td>
<td>#594a26</td>
<td>#AD5CCC</td>
<td>#1F8ECC</td>
<td>#2C00FF</td>
<td>25</td>
<td>#CC5CB0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">TAACCAGCACCAAATC-204.2_A01</td>
<td>TAACCAGCACCAAATC</td>
<td>204.2_A01</td>
<td>L8TX_200122_01_F12</td>
<td>WMB-10Xv3-MB</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>MB</td>
<td>Snap25-IRES2-Cre;Ai14-504535</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>...</td>
<td>327 Oligo NN</td>
<td>1184 MOL NN_4</td>
<td>5285 MOL NN_4</td>
<td>#666666</td>
<td>#03045E</td>
<td>#99FFBC</td>
<td>#99CAFF</td>
<td>#5A662E</td>
<td>25</td>
<td>#CC5CB0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">AGCGTCGGTTCAGCGC-420_C06</td>
<td>AGCGTCGGTTCAGCGC</td>
<td>420_C06</td>
<td>L8TX_201106_01_G03</td>
<td>WMB-10Xv3-MB</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>MB</td>
<td>Snap25-IRES2-Cre;Ai14-551017</td>
<td>Ai14(RCL-tdT)/wt</td>
<td>...</td>
<td>327 Oligo NN</td>
<td>1184 MOL NN_4</td>
<td>5284 MOL NN_4</td>
<td>#666666</td>
<td>#03045E</td>
<td>#99FFBC</td>
<td>#99CAFF</td>
<td>#5B2E66</td>
<td>25</td>
<td>#CC5CB0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">AGACAAAAGAACTCCT-278_B03</td>
<td>AGACAAAAGAACTCCT</td>
<td>278_B03</td>
<td>L8TX_200702_01_H01</td>
<td>WMB-10Xv3-TH</td>
<td>cell</td>
<td>NaN</td>
<td>10Xv3</td>
<td>TH</td>
<td>Snap25-IRES2-Cre;Ai14-529474</td>
<td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>
<td>...</td>
<td>150 CM-IAD-CL-PCN Sema5b Glut</td>
<td>0652 CM-IAD-CL-PCN Sema5b Glut_4</td>
<td>2644 CM-IAD-CL-PCN Sema5b Glut_4</td>
<td>#2B93DF</td>
<td>#0D47A1</td>
<td>#076600</td>
<td>#CC7D7A</td>
<td>#7ACC7D</td>
<td>23</td>
<td>#FF00EF</td>
</tr>
</tbody>
</table>

<p>5 rows × 27 columns</p>
</div>
</div>
<p>:::</p>
<div id="b9faae07-e863-4a9f-adcc-abe61138fb09" class="cell" data-execution_count="105">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="co"># identify the shared set of genes between spatial and single-cell </span></span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>shared_features <span class="op">=</span> [feature <span class="cf">for</span> feature <span class="kw">in</span> adata_section.var_names <span class="cf">if</span> feature <span class="kw">in</span> adata_sc.var_names]</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a><span class="co"># filter </span></span>
<span id="cb138-5"><a href="#cb138-5" aria-hidden="true" tabindex="-1"></a>adata_sc <span class="op">=</span> adata_sc[:, adata_sc.var.index.isin(shared_features)].copy()</span>
<span id="cb138-6"><a href="#cb138-6" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> adata_section[:, shared_features].copy() <span class="co"># probably no need to filter the spatial gene set</span></span>
<span id="cb138-7"><a href="#cb138-7" aria-hidden="true" tabindex="-1"></a><span class="co"># reorder genes to have them match between spatial data and sc</span></span>
<span id="cb138-8"><a href="#cb138-8" aria-hidden="true" tabindex="-1"></a>adata_sc <span class="op">=</span> adata_sc[:, shared_features].copy()</span>
<span id="cb138-9"><a href="#cb138-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-10"><a href="#cb138-10" aria-hidden="true" tabindex="-1"></a>adata_section.shape[<span class="dv">1</span>] <span class="op">==</span> adata_sc.shape[<span class="dv">1</span>]</span>
<span id="cb138-11"><a href="#cb138-11" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Number of shared genes between spatial data and reference sc: </span><span class="sc">{</span>adata_section<span class="sc">.</span>shape[<span class="dv">1</span>]<span class="sc">}</span><span class="ss">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of shared genes between spatial data and reference sc: 1122</code></pre>
</div>
</div>
<div id="2afb3635-7b4f-41cd-ad82-b1bed671ec9b" class="cell" data-execution_count="106">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fit the reference model</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>selected <span class="op">=</span> c2l.utils.filtering.filter_genes(</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>    adata_sc, cell_count_cutoff<span class="op">=</span><span class="dv">5</span>, cell_percentage_cutoff2<span class="op">=</span><span class="fl">0.03</span>, nonz_mean_cutoff<span class="op">=</span><span class="fl">1.12</span></span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a>adata_sc <span class="op">=</span> adata_sc[:, selected].copy()</span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> adata_section[:, selected].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-106-output-1.png" width="338" height="321" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="922e8303-ec18-4064-858e-a31dc7d13426" class="cell" data-execution_count="108">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>adata_section.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="108">
<pre><code>(37068, 1089)</code></pre>
</div>
</div>
<div id="7f970565-e3ee-4103-aac0-134ece4858ed" class="cell" data-execution_count="107">
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>adata_sc.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="107">
<pre><code>(28981, 1089)</code></pre>
</div>
</div>
<div id="36e31ad4-286d-4c09-ab76-afd1d4c6c7d6" class="cell" data-execution_count="110">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>c2l.models.RegressionModel.setup_anndata(</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>adata_sc,</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#batch_key="Replicates", # there's no batch </span></span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a>    labels_key<span class="op">=</span><span class="st">"class"</span>, <span class="co"># cell type = class</span></span>
<span id="cb145-5"><a href="#cb145-5" aria-hidden="true" tabindex="-1"></a>    <span class="co">#categorical_covariate_keys=["assay"],</span></span>
<span id="cb145-6"><a href="#cb145-6" aria-hidden="true" tabindex="-1"></a>    layer<span class="op">=</span><span class="st">"raw"</span>,</span>
<span id="cb145-7"><a href="#cb145-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb145-8"><a href="#cb145-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-9"><a href="#cb145-9" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> c2l.models.RegressionModel(adata_sc)</span>
<span id="cb145-10"><a href="#cb145-10" aria-hidden="true" tabindex="-1"></a><span class="co"># default, try on GPU:</span></span>
<span id="cb145-11"><a href="#cb145-11" aria-hidden="true" tabindex="-1"></a>model.train(max_epochs<span class="op">=</span><span class="dv">250</span>, batch_size<span class="op">=</span><span class="dv">2500</span>, train_size<span class="op">=</span><span class="dv">1</span>, lr<span class="op">=</span><span class="fl">0.002</span>)</span>
<span id="cb145-12"><a href="#cb145-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-13"><a href="#cb145-13" aria-hidden="true" tabindex="-1"></a>model.export_posterior(</span>
<span id="cb145-14"><a href="#cb145-14" aria-hidden="true" tabindex="-1"></a>    adata_sc,</span>
<span id="cb145-15"><a href="#cb145-15" aria-hidden="true" tabindex="-1"></a>    sample_kwargs<span class="op">=</span>{<span class="st">"num_samples"</span>: <span class="dv">1000</span>, <span class="st">"batch_size"</span>: <span class="dv">2500</span>},</span>
<span id="cb145-16"><a href="#cb145-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO: GPU available: True (cuda), used: True
INFO:lightning.pytorch.utilities.rank_zero:GPU available: True (cuda), used: True
INFO: TPU available: False, using: 0 TPU cores
INFO:lightning.pytorch.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO: HPU available: False, using: 0 HPUs
INFO:lightning.pytorch.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO: LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:lightning.pytorch.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Epoch 250/250: 100%|██████████| 250/250 [52:11&lt;00:00, 14.47s/it, v_num=1, elbo_train=1.41e+7]</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO: `Trainer.fit` stopped: `max_epochs=250` reached.
INFO:lightning.pytorch.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=250` reached.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Epoch 250/250: 100%|██████████| 250/250 [52:11&lt;00:00, 12.53s/it, v_num=1, elbo_train=1.41e+7]
Sampling local variables, batch:   0%|          | 0/12 [00:01&lt;?, ?it/s]
Sampling global variables, sample: 100%|██████████| 999/999 [00:13&lt;00:00, 73.60it/s]</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="110">
<pre><code>AnnData object with n_obs × n_vars = 28981 × 1089
    obs: 'cell_barcode', 'barcoded_cell_sample_label', 'library_label', 'feature_matrix_label', 'entity', 'brain_section_label', 'library_method', 'region_of_interest_acronym', 'donor_label', 'donor_genotype', 'donor_sex', 'dataset_label', 'x', 'y', 'cluster_alias', 'neurotransmitter', 'class', 'subclass', 'supertype', 'cluster', 'neurotransmitter_color', 'class_color', 'subclass_color', 'supertype_color', 'cluster_color', 'region_of_interest_order', 'region_of_interest_color', '_indices', '_scvi_batch', '_scvi_labels'
    var: 'gene_identifier', 'name', 'mapped_ncbi_identifier', 'comment', 'n_cells', 'nonz_mean'
    uns: 'training_genes', 'overlap_genes', '_scvi_uuid', '_scvi_manager_uuid', 'mod'
    varm: 'means_per_cluster_mu_fg', 'stds_per_cluster_mu_fg', 'q05_per_cluster_mu_fg', 'q95_per_cluster_mu_fg'
    layers: 'raw'</code></pre>
</div>
</div>
<div id="4f775da0-2656-486d-8867-dd584c1c2c09" class="cell" data-execution_count="111">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="co"># export estimated expression in each cluster</span></span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">"means_per_cluster_mu_fg"</span> <span class="kw">in</span> adata_sc.varm.keys():</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>    inf_aver <span class="op">=</span> adata_sc.varm[<span class="st">"means_per_cluster_mu_fg"</span>][</span>
<span id="cb151-4"><a href="#cb151-4" aria-hidden="true" tabindex="-1"></a>        [<span class="ss">f"means_per_cluster_mu_fg_</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span> <span class="cf">for</span> i <span class="kw">in</span> adata_sc.uns[<span class="st">"mod"</span>][<span class="st">"factor_names"</span>]]</span>
<span id="cb151-5"><a href="#cb151-5" aria-hidden="true" tabindex="-1"></a>    ].copy()</span>
<span id="cb151-6"><a href="#cb151-6" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb151-7"><a href="#cb151-7" aria-hidden="true" tabindex="-1"></a>    inf_aver <span class="op">=</span> adata_sc.var[</span>
<span id="cb151-8"><a href="#cb151-8" aria-hidden="true" tabindex="-1"></a>        [<span class="ss">f"means_per_cluster_mu_fg_</span><span class="sc">{</span>i<span class="sc">}</span><span class="ss">"</span> <span class="cf">for</span> i <span class="kw">in</span> adata_sc.uns[<span class="st">"mod"</span>][<span class="st">"factor_names"</span>]]</span>
<span id="cb151-9"><a href="#cb151-9" aria-hidden="true" tabindex="-1"></a>    ].copy()</span>
<span id="cb151-10"><a href="#cb151-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-11"><a href="#cb151-11" aria-hidden="true" tabindex="-1"></a>inf_aver.columns <span class="op">=</span> adata_sc.uns[<span class="st">"mod"</span>][<span class="st">"factor_names"</span>]</span>
<span id="cb151-12"><a href="#cb151-12" aria-hidden="true" tabindex="-1"></a>inf_aver.head()</span>
<span id="cb151-13"><a href="#cb151-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-14"><a href="#cb151-14" aria-hidden="true" tabindex="-1"></a><span class="co"># cell 2 location cell type mapping</span></span>
<span id="cb151-15"><a href="#cb151-15" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'start with Cell2location'</span>)</span>
<span id="cb151-16"><a href="#cb151-16" aria-hidden="true" tabindex="-1"></a>c2l.models.Cell2location.setup_anndata(</span>
<span id="cb151-17"><a href="#cb151-17" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>adata_section,</span>
<span id="cb151-18"><a href="#cb151-18" aria-hidden="true" tabindex="-1"></a>    <span class="co">#batch_key="kidney_n",</span></span>
<span id="cb151-19"><a href="#cb151-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb151-20"><a href="#cb151-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-21"><a href="#cb151-21" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> c2l.models.Cell2location(</span>
<span id="cb151-22"><a href="#cb151-22" aria-hidden="true" tabindex="-1"></a>    adata_section,</span>
<span id="cb151-23"><a href="#cb151-23" aria-hidden="true" tabindex="-1"></a>    cell_state_df<span class="op">=</span>inf_aver,</span>
<span id="cb151-24"><a href="#cb151-24" aria-hidden="true" tabindex="-1"></a>    N_cells_per_location<span class="op">=</span><span class="dv">8</span>,</span>
<span id="cb151-25"><a href="#cb151-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb151-26"><a href="#cb151-26" aria-hidden="true" tabindex="-1"></a>model.view_anndata_setup()</span>
<span id="cb151-27"><a href="#cb151-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-28"><a href="#cb151-28" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'start with training'</span>)</span>
<span id="cb151-29"><a href="#cb151-29" aria-hidden="true" tabindex="-1"></a>model.train(max_epochs<span class="op">=</span><span class="dv">30000</span>, batch_size<span class="op">=</span><span class="va">None</span>, train_size<span class="op">=</span><span class="dv">1</span>)<span class="co">#, use_gpu=use_gpu)</span></span>
<span id="cb151-30"><a href="#cb151-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-31"><a href="#cb151-31" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> model.export_posterior(</span>
<span id="cb151-32"><a href="#cb151-32" aria-hidden="true" tabindex="-1"></a>    adata_section,</span>
<span id="cb151-33"><a href="#cb151-33" aria-hidden="true" tabindex="-1"></a>    sample_kwargs<span class="op">=</span>{</span>
<span id="cb151-34"><a href="#cb151-34" aria-hidden="true" tabindex="-1"></a>        <span class="st">"num_samples"</span>: <span class="dv">1000</span>,</span>
<span id="cb151-35"><a href="#cb151-35" aria-hidden="true" tabindex="-1"></a>        <span class="st">"batch_size"</span>: model.adata.n_obs}<span class="co">#,</span></span>
<span id="cb151-36"><a href="#cb151-36" aria-hidden="true" tabindex="-1"></a>        <span class="co">#"use_gpu": use_gpu,</span></span>
<span id="cb151-37"><a href="#cb151-37" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb151-38"><a href="#cb151-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-39"><a href="#cb151-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb151-40"><a href="#cb151-40" aria-hidden="true" tabindex="-1"></a>adata_section.obs[adata_section.uns[<span class="st">"mod"</span>][<span class="st">"factor_names"</span>]] <span class="op">=</span> adata_section.obsm[</span>
<span id="cb151-41"><a href="#cb151-41" aria-hidden="true" tabindex="-1"></a>    <span class="st">"q05_cell_abundance_w_sf"</span></span>
<span id="cb151-42"><a href="#cb151-42" aria-hidden="true" tabindex="-1"></a>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>start with Cell2location</code></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Anndata setup with scvi-tools version <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1.2</span>.<span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>.
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">Setup via `Cell2location.setup_anndata` with arguments:
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-weight: bold">{</span>
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'layer'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'batch_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'labels_key'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'categorical_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>,
<span style="color: #7fbf7f; text-decoration-color: #7fbf7f">│   </span><span style="color: #008000; text-decoration-color: #008000">'continuous_covariate_keys'</span>: <span style="color: #800080; text-decoration-color: #800080; font-style: italic">None</span>
<span style="font-weight: bold">}</span>
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"></pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">         Summary Statistics         </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┓
┃<span style="font-weight: bold">     Summary Stat Key     </span>┃<span style="font-weight: bold"> Value </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_batch          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   1   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_cells          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 37068 </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_categorical_covs </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff"> n_extra_continuous_covs  </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   0   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">         n_labels         </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   1   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">          n_vars          </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> 1089  </span>│
└──────────────────────────┴───────┘
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">               Data Registry                </span>
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold"> Registry Key </span>┃<span style="font-weight: bold">    scvi-tools Location    </span>┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff">      X       </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          adata.X          </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    batch     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_batch']  </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    ind_x     </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">   adata.obs['_indices']   </span>│
│<span style="color: #0087ff; text-decoration-color: #0087ff">    labels    </span>│<span style="color: #af00d7; text-decoration-color: #af00d7"> adata.obs['_scvi_labels'] </span>│
└──────────────┴───────────────────────────┘
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                     batch State Registry                      </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">     Source Location      </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['_scvi_batch'] </span>│<span style="color: #008000; text-decoration-color: #008000">     0      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
└──────────────────────────┴────────────┴─────────────────────┘
</pre>
</div>
<div class="cell-output cell-output-display">
<pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace"><span style="font-style: italic">                     labels State Registry                      </span>
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┓
┃<span style="font-weight: bold">      Source Location      </span>┃<span style="font-weight: bold"> Categories </span>┃<span style="font-weight: bold"> scvi-tools Encoding </span>┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━┩
│<span style="color: #0087ff; text-decoration-color: #0087ff"> adata.obs['_scvi_labels'] </span>│<span style="color: #008000; text-decoration-color: #008000">     0      </span>│<span style="color: #af00d7; text-decoration-color: #af00d7">          0          </span>│
└───────────────────────────┴────────────┴─────────────────────┘
</pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>start with training</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO: GPU available: True (cuda), used: True
INFO:lightning.pytorch.utilities.rank_zero:GPU available: True (cuda), used: True
INFO: TPU available: False, using: 0 TPU cores
INFO:lightning.pytorch.utilities.rank_zero:TPU available: False, using: 0 TPU cores
INFO: HPU available: False, using: 0 HPUs
INFO:lightning.pytorch.utilities.rank_zero:HPU available: False, using: 0 HPUs
INFO: LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
INFO:lightning.pytorch.accelerators.cuda:LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Epoch 30000/30000: 100%|██████████| 30000/30000 [1:33:46&lt;00:00,  3.35it/s, v_num=1, elbo_train=1.84e+7]  </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO: `Trainer.fit` stopped: `max_epochs=30000` reached.
INFO:lightning.pytorch.utilities.rank_zero:`Trainer.fit` stopped: `max_epochs=30000` reached.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Epoch 30000/30000: 100%|██████████| 30000/30000 [1:33:46&lt;00:00,  5.33it/s, v_num=1, elbo_train=1.84e+7]
Sampling local variables, batch: 100%|██████████| 1/1 [02:20&lt;00:00, 140.75s/it]
Sampling global variables, sample: 100%|██████████| 999/999 [01:09&lt;00:00, 14.44it/s]</code></pre>
</div>
</div>
<div id="38655df4-9b3c-47a8-a6ca-2f615a4abefb" class="cell" data-execution_count="112">
<div class="sourceCode cell-code" id="cb158"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a>adata_section.write_h5ad(<span class="ss">f'</span><span class="sc">{</span>spatial_wd<span class="sc">}</span><span class="ss">/stdata_deconv.h5ad'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="1cb809a2-a820-4736-bcde-1323d70999b1" class="cell" data-execution_count="113">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a>adata_section.obs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="113">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">brain_section_label</th>
<th data-quarto-table-cell-role="th">feature_matrix_label</th>
<th data-quarto-table-cell-role="th">donor_label</th>
<th data-quarto-table-cell-role="th">donor_genotype</th>
<th data-quarto-table-cell-role="th">donor_sex</th>
<th data-quarto-table-cell-role="th">cluster_alias</th>
<th data-quarto-table-cell-role="th">x</th>
<th data-quarto-table-cell-role="th">y</th>
<th data-quarto-table-cell-role="th">z</th>
<th data-quarto-table-cell-role="th">subclass_confidence_score</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">16 HY MM Glut</th>
<th data-quarto-table-cell-role="th">17 MH-LH Glut</th>
<th data-quarto-table-cell-role="th">18 TH Glut</th>
<th data-quarto-table-cell-role="th">19 MB Glut</th>
<th data-quarto-table-cell-role="th">20 MB GABA</th>
<th data-quarto-table-cell-role="th">25 Pineal Glut</th>
<th data-quarto-table-cell-role="th">30 Astro-Epen</th>
<th data-quarto-table-cell-role="th">31 OPC-Oligo</th>
<th data-quarto-table-cell-role="th">33 Vascular</th>
<th data-quarto-table-cell-role="th">34 Immune</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">cell_label</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">299794030110334465835331169565547130392</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>196</td>
<td>5.873824</td>
<td>1.246374</td>
<td>6.757411</td>
<td>0.951445</td>
<td>...</td>
<td>0.000440</td>
<td>0.000375</td>
<td>0.000383</td>
<td>0.000513</td>
<td>0.001199</td>
<td>0.000208</td>
<td>0.115745</td>
<td>0.034190</td>
<td>0.006665</td>
<td>0.000508</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">66045375924099067525605874805924532486</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>196</td>
<td>5.066814</td>
<td>1.215444</td>
<td>6.757411</td>
<td>0.990889</td>
<td>...</td>
<td>0.000418</td>
<td>0.000054</td>
<td>0.000034</td>
<td>0.000193</td>
<td>0.001392</td>
<td>0.000369</td>
<td>0.000557</td>
<td>0.000258</td>
<td>0.000100</td>
<td>0.002435</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">116791314340621211157298988846240231316</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>196</td>
<td>4.681061</td>
<td>1.287158</td>
<td>6.757411</td>
<td>0.972796</td>
<td>...</td>
<td>0.000155</td>
<td>0.000030</td>
<td>0.000030</td>
<td>0.000185</td>
<td>0.000318</td>
<td>0.000047</td>
<td>0.012837</td>
<td>0.000077</td>
<td>0.004797</td>
<td>0.000160</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13950435409550353848526753205083905255</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>174</td>
<td>4.708601</td>
<td>1.292049</td>
<td>6.757411</td>
<td>0.995666</td>
<td>...</td>
<td>0.000786</td>
<td>0.000336</td>
<td>0.000816</td>
<td>0.001343</td>
<td>0.003183</td>
<td>0.000407</td>
<td>0.040397</td>
<td>0.000437</td>
<td>0.119662</td>
<td>0.021737</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">147436316185400258367147717470263928395</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>174</td>
<td>4.729919</td>
<td>1.285492</td>
<td>6.757411</td>
<td>0.999973</td>
<td>...</td>
<td>0.004426</td>
<td>0.000134</td>
<td>0.000123</td>
<td>0.000602</td>
<td>0.003589</td>
<td>0.000402</td>
<td>0.048852</td>
<td>0.001546</td>
<td>0.000159</td>
<td>0.005458</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">138771139060049685442837890240266124575</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5231</td>
<td>1.083467</td>
<td>4.908302</td>
<td>6.757411</td>
<td>0.999978</td>
<td>...</td>
<td>0.000025</td>
<td>0.000023</td>
<td>0.000028</td>
<td>0.000033</td>
<td>0.000047</td>
<td>0.000024</td>
<td>0.000196</td>
<td>0.174305</td>
<td>0.000264</td>
<td>0.000923</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">293280579073114219480073630362346662120</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5231</td>
<td>1.084075</td>
<td>4.902757</td>
<td>6.757411</td>
<td>0.999968</td>
<td>...</td>
<td>0.000035</td>
<td>0.000023</td>
<td>0.000033</td>
<td>0.000050</td>
<td>0.000103</td>
<td>0.002861</td>
<td>0.009330</td>
<td>0.631577</td>
<td>0.000144</td>
<td>0.000352</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">29789632307409269107735464912172084480</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5231</td>
<td>0.973041</td>
<td>4.805131</td>
<td>6.757411</td>
<td>0.999985</td>
<td>...</td>
<td>0.000018</td>
<td>0.000012</td>
<td>0.000626</td>
<td>0.000096</td>
<td>0.000053</td>
<td>0.000017</td>
<td>0.001566</td>
<td>0.721932</td>
<td>0.025739</td>
<td>0.000318</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">63992048193478335418190347526482691865</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5231</td>
<td>0.865662</td>
<td>4.294161</td>
<td>6.757411</td>
<td>0.999959</td>
<td>...</td>
<td>0.000019</td>
<td>0.000013</td>
<td>0.000011</td>
<td>0.000036</td>
<td>0.000043</td>
<td>0.000017</td>
<td>0.068745</td>
<td>1.068613</td>
<td>0.000251</td>
<td>0.002840</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">217820882131375786118651898970174983535</td>
<td>Zhuang-ABCA-1.080</td>
<td>Zhuang-ABCA-1</td>
<td>Zhuang-ABCA-1</td>
<td>wt/wt</td>
<td>F</td>
<td>5212</td>
<td>0.841118</td>
<td>3.895234</td>
<td>6.757411</td>
<td>0.954223</td>
<td>...</td>
<td>0.000579</td>
<td>0.000088</td>
<td>0.000050</td>
<td>0.000254</td>
<td>0.000727</td>
<td>0.000439</td>
<td>0.000461</td>
<td>1.076670</td>
<td>0.000577</td>
<td>0.000850</td>
</tr>
</tbody>
</table>

<p>37068 rows × 51 columns</p>
</div>
</div>
</div>
<div id="41651555-7421-4035-bbf2-392d8d66843b" class="cell">
<div class="sourceCode cell-code" id="cb160"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a>adata_sc.X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="be766873-5e86-44bc-8602-0ae7aba1c511" class="cell" data-execution_count="114">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Record the end time</span></span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>end_time <span class="op">=</span> time.time()</span>
<span id="cb161-3"><a href="#cb161-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb161-4"><a href="#cb161-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate elapsed time in minutes</span></span>
<span id="cb161-5"><a href="#cb161-5" aria-hidden="true" tabindex="-1"></a>elapsed_time <span class="op">=</span> (end_time <span class="op">-</span> start_time) <span class="op">/</span> <span class="dv">60</span></span>
<span id="cb161-6"><a href="#cb161-6" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Execution Time Tangram: </span><span class="sc">{</span>elapsed_time<span class="sc">:.2f}</span><span class="ss"> minutes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Execution Time Tangram: 159.19 minutes</code></pre>
</div>
</div>
<section id="additional-access-image-in-the-anndata-object" class="level3">
<h3 class="anchored" data-anchor-id="additional-access-image-in-the-anndata-object">Additional: access image in the AnnData object</h3>
<div id="308fd6bd-2050-4773-be52-26260f8280f3" class="cell" data-execution_count="115">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> PIL <span class="im">import</span> Image</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> PIL</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>PIL.Image.MAX_IMAGE_PIXELS <span class="op">=</span> <span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="291a0ceb-87a0-4fd5-91ee-03bd28e34096" class="cell" data-execution_count="116">
<div class="sourceCode cell-code" id="cb164"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.datasets.visium_sge(sample_id<span class="op">=</span><span class="st">"V1_Mouse_Brain_Sagittal_Posterior"</span>)</span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>adata.var_names_make_unique()</span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">"mt"</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">"MT-"</span>)</span>
<span id="cb164-4"><a href="#cb164-4" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>], inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>reading /exports/humgen/bmanzato/spatial_workshop/data/V1_Mouse_Brain_Sagittal_Posterior/filtered_feature_bc_matrix.h5
 (0:00:02)</code></pre>
</div>
</div>
<div id="6b3923c1-5836-48c0-bd42-f9946197308f" class="cell" data-execution_count="117">
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, min_counts<span class="op">=</span><span class="dv">5000</span>)</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, max_counts<span class="op">=</span><span class="dv">35000</span>)</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[adata.obs[<span class="st">"pct_counts_mt"</span>] <span class="op">&lt;</span> <span class="dv">20</span>].copy()</span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"#cells after MT filter: </span><span class="sc">{</span>adata<span class="sc">.</span>n_obs<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_genes(adata, min_cells<span class="op">=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>filtered out 259 cells that have less than 5000 counts
filtered out 291 cells that have more than 35000 counts
#cells after MT filter: 2805
filtered out 15603 genes that are detected in less than 10 cells</code></pre>
</div>
</div>
<div id="027519f2-4ec3-444b-a73c-b258ab669bd0" class="cell" data-execution_count="118">
<div class="sourceCode cell-code" id="cb168"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="118">
<pre><code>AnnData object with n_obs × n_vars = 2805 × 16682
    obs: 'in_tissue', 'array_row', 'array_col', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'n_counts'
    var: 'gene_ids', 'feature_types', 'genome', 'mt', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_cells'
    uns: 'spatial'
    obsm: 'spatial'</code></pre>
</div>
</div>
<div id="073b9aee-547d-4dd2-9842-2d94c6625c2f" class="cell" data-execution_count="119">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">'Number of spots in the GE matrix (section1, IN TISSUE): '</span>, adata.shape[<span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of spots in the GE matrix (section1, IN TISSUE):  2805</code></pre>
</div>
</div>
<figure class="figure">
<img src="img/tissue_hires_image.png" width="700" height="280" alt="Description of the image" class="figure-img">
<figcaption>
From <a href="https://www.10xgenomics.com/datasets/mouse-brain-serial-section-1-sagittal-posterior-1-standard-1-0-0" target="_blank">10X Genomics</a>
</figcaption>
</figure>
<div id="38fc7f0c-d5fd-4914-ae93-27929db23bb2" class="cell" data-execution_count="120">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> (<span class="dv">8</span>, <span class="dv">8</span>)</span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata, img_key<span class="op">=</span><span class="st">"hires"</span>, color<span class="op">=</span>[<span class="st">"total_counts"</span>, <span class="st">"n_genes_by_counts"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="p1_integration_sc_files/figure-html/cell-120-output-1.png" width="1190" height="510" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="5ed59d35-ed53-4633-bb39-54bb4391ae4c" class="cell">
<div class="sourceCode cell-code" id="cb173"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a>sc.pp.pca(adata)</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata)</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span>
<span id="cb173-4"><a href="#cb173-4" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, key_added<span class="op">=</span><span class="st">"clusters"</span>, flavor<span class="op">=</span><span class="st">"igraph"</span>, directed<span class="op">=</span><span class="va">False</span>, n_iterations<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>computing PCA
    with n_comps=50</code></pre>
</div>
</div>
<div id="70bb1ae9-0d24-4a96-88c6-5e36e92e9131" class="cell">
<div class="sourceCode cell-code" id="cb175"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb175-1"><a href="#cb175-1" aria-hidden="true" tabindex="-1"></a>plt.rcParams[<span class="st">"figure.figsize"</span>] <span class="op">=</span> (<span class="dv">8</span>, <span class="dv">8</span>)</span>
<span id="cb175-2"><a href="#cb175-2" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata, img_key<span class="op">=</span><span class="st">"hires"</span>, color<span class="op">=</span>[<span class="st">"clusters"</span>]) <span class="co"># leiden</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="98c7082a-cd5b-49df-9499-e5d57dfaa4be" class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cluster marker genes</span></span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a>adata.var_names_make_unique()</span>
<span id="cb176-3"><a href="#cb176-3" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, <span class="st">"clusters"</span>, method<span class="op">=</span><span class="st">"t-test"</span>)</span>
<span id="cb176-4"><a href="#cb176-4" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_heatmap(adata, groups<span class="op">=</span><span class="st">"9"</span>, n_genes<span class="op">=</span><span class="dv">10</span>, groupby<span class="op">=</span><span class="st">"clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="c4c0e5fc-9415-4ce6-9776-def44dd54ecc" class="cell">
<div class="sourceCode cell-code" id="cb177"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb177-1"><a href="#cb177-1" aria-hidden="true" tabindex="-1"></a>sc.pl.spatial(adata, img_key<span class="op">=</span><span class="st">"hires"</span>, color<span class="op">=</span>[<span class="st">"Ddn"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><span style="color: red; font-size: 15px;"> <strong>EX 16: Replicate the analysis in the last section on another 10X Genomics Visium dataset. Check which datasets are available <a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.datasets.visium_sge.html">here</a></strong></span></p>


</section>
</font></section><font size="3">

</font></font></font></font></font></font></font></font></font></font></font></font></font></main><font size="3"><font size="3"><font size="3"> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/elixir-europe-training\.github\.io\/ELIXIR-SCO-spatial-omics\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</font></font></font></div><font size="3"><font size="3"><font size="3"> <!-- /content -->




</font></font></font></body></html>