{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97537554-152e-4639-8283-3007feb980a8",
   "metadata": {},
   "source": [
    "# Spatial data integration with scRNA-seq\n",
    "## Day 2 - Practical 3\n",
    "\n",
    "#### 23rd January 2025\n",
    "#### Ahmed Mahfouz, Benedetta Manzato (LUMC) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67603fb8-9267-4074-a900-f8e72f97cf94",
   "metadata": {},
   "source": [
    "### Agenda:\n",
    "<font size=\"3\">1. Dataset description\n",
    "\n",
    "<font size=\"3\">2. Spatial imputation with SpaGE and Tangram\n",
    "\n",
    "<font size=\"3\">3. Cell deconvolution with Cell2location\n",
    "\n",
    "<font size=\"3\">4. Access image in AnnData object"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e75a8e87-679c-42b7-8228-cdac0d35a39e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Dataset: MERFISH spatial transcriptomics dataset of a single adult mouse brain (Zhuang-ABCA-1) (Xiaowei Zhuang)\n",
    "\n",
    "<font size=\"3\"> This spatially resolved cell atlas of the whole mouse brain provides a systematic characterization of the spatial organization of transcriptomically defined cell types across the entire adult mouses brain using in situ, single-cell transcriptomic profiling with multiplexed error-robust fluorescence in situ hybridization (MERFISH). A gene panel of 1,122 genes was imaged in 9 million brain cells using **MERFISH**, and spatially resolved, single-cell expression profiling was performed at the whole-transcriptome scale by integrating the MERFISH data with the whole-brain single-cell RNA-sequencing (scRNA-seq) dataset previously generated by the Allen Institute. This resulted in the generation of a comprehensive cell atlas of more than 5,000 transcriptionally distinct cell clusters, belonging to âˆ¼300 major cell types (subclasses), across the whole mouse brain, with high molecular and spatial resolution.\n",
    "The authors performed MERFISH imaging on **245 coronal and sagittal sections**, the cell atlas was then spatially registered to the Allen Mouse Brain Common Coordinate Framework (CCFv3), which allows systematic quantifications of the cell composition and organization in individual brain regions as anatomically delineated in the CCFv3. This cell atlas also reveals molecularly-defined brain regions characterized by distinct cell type compositions, spatial gradients featuring gradual changes in the gene-expression profiles of cells, as well as cell-type-specific cell-cell interactions and the molecular basis and functional implications of these cell-cell interactions.\n",
    "\n",
    "<font size=\"3\"> The collection spans four mouse specimens (two coronal sets and two sagittal sets):\n",
    "- <font size=\"3\">**Zhuang-ABCA-1**: \n",
    "  - **Dataset size**: 4.2 million cells spanning 147 coronal sections with a 1,122 gene panel.\n",
    "  - **Filtered dataset**: 2.8 million cells passed the cell classification confidence score threshold and are displayed in the ABC atlas.\n",
    "  \n",
    "- <font size=\"3\">**Zhuang-ABCA-2**: \n",
    "  - **Dataset size**: 1.9 million cells spanning 66 coronal sections.\n",
    "  - **Filtered dataset**: 1.2 million cells passed the cell classification confidence score threshold.\n",
    "  \n",
    "- <font size=\"3\">**Zhuang-ABCA-3**: \n",
    "  - **Dataset size**: 2.1 million cells spanning 23 sagittal sections.\n",
    "  - **Filtered dataset**: 1.6 million cells passed the confidence score threshold.\n",
    "\n",
    "- <font size=\"3\">**Zhuang-ABCA-4**: \n",
    "  - **Dataset size**: 0.22 million cells spanning 3 sagittal sections.\n",
    "  - **Filtered dataset**: 0.16 million cells passed the confidence score threshold.\n",
    "\n",
    "\n",
    "<font size=\"3\"> We will work with the Brain Coronal 1 collection (Zhuang-ABCA-1); the cell-by-gene matrix of the 8.4 millions cells can be downloaded from the AWS bucket of this animal. The cells are then filtered by cell-classification (label transfer) confidence scores calculated during MERFISH-scRNAseq data integration. 7.0 million cells passed the confidence score threshold for cell subclass label transfer and 5.8 million cells further passed the confidence score threshold for cell cluster label transfer. These 5.8 million cells are included in the cell metadata file that can be downloaded from the the AWS bucket and are displayed on the ABC Atlas. The CCF coordinates of the 5.4 million cells that were registered to the 3D Allen-CCF can be downloaded from the CCF coordinate files in the AWB bucket. The collection spans four mouse specimens (2 coronal sets and 2 sagittal sets). Cells are mapped to the whole mouse brain taxonomy (WMB-taxonomy) and Allen Common Coordinate Framework (Allen-CCF-2020). \n",
    "Refer to Zhang et al, 2023 for more details.\n",
    "\n",
    "\n",
    "For more details: [Molecularly defined and spatially resolved cell atlas of the whole mouse brain (Zhang et al. 2023)](https://www.nature.com/articles/s41586-023-06808-9)\n",
    "\n",
    "From https://github.com/AllenInstitute/abc_atlas_access/blob/main/notebooks/zhuang_merfish_tutorial.ipynb\n",
    "\n",
    "Data Availablity : https://knowledge.brain-map.org/data/5C0201JSVE04WY6DMVC/collections </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a8305d-204c-4780-8c01-738bbe0eeabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "# Record the start time\n",
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c80f5d54-37d4-4375-96c9-48f29a1585a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Importing necessary libraries for spatial data analysis and visualization\n",
    "import os, re\n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "\n",
    "# Display the version of important packages for reproducibility\n",
    "sc.logging.print_header()\n",
    "\n",
    "# Set the default aesthetics for Scanpy plots\n",
    "sc.set_figure_params(facecolor='white', figsize=(8, 8))  # White background, figure size 8x8\n",
    "sc.settings.verbosity = 0  # Set verbosity to show only errors (0) # errors (0), warnings (1), info (2), hints (3)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d917233-09ff-4c30-8941-30a424c4f666",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We create an output folder in your local space to save outputs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "860e1d4a-b95e-4d5f-84e3-ffee578f8f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = './outputs/'\n",
    "if not os.path.exists(folder_path):\n",
    "        os.makedirs(folder_path)\n",
    "        print(f\"Folder created: {folder_path}\")\n",
    "else:\n",
    "    print(f\"Folder already exists: {folder_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f83355-5d90-48d8-b358-e1b7ffc284d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/data/spatial_workshop/day2/practical_3'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9968f120-9669-4d22-ba9d-c30511af5771",
   "metadata": {},
   "source": [
    "<font size=\"3\">We will use the **Zhuang-ABCA-1 dataset**: 4.2 million cell spatial transcriptomics dataset spanning 147 coronal sections with a 1122 gene panel. 2.8 million cells passed cell classification confidence score threshold and displayed in the ABC atlas.\n",
    "\n",
    "<font size=\"3\">We decided to work with **one section only**, the one that captures most cells (section 80).\n",
    "\n",
    "<img src=\"img/ABCAtlas_Zhuang-ABCA-1_Coronal Grid.png\" width=\"800\" height=\"300\">"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9422d000-2e27-4c09-b562-49b2b4060952",
   "metadata": {},
   "source": [
    "### Read data \n",
    "\n",
    "<font size=\"3\">We will read the AnnData object that has been filtered to have only section 80 and QC was performed from the authors of the study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd03c2eb-98e0-4ff3-b3c8-e764013e303e",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section = ad.read_h5ad(f\"{data_dir}/abc_download_root/expression_matrices/Zhuang-ABCA-1/adata_section80.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8367a912-3ae7-4b62-add0-0dcfce2d3362",
   "metadata": {},
   "source": [
    "<font size=\"3\">Spatial coordinates, being metadata about cells/spots are stored either in **`.obs`** or **`.obsm`**\n",
    "\n",
    "<font size=\"3\">Spot metadata also contains info about class, subclass, neurotransmitter, patient metadata and section."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72188ac1-6e72-4f20-b4a8-22117beb3ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa002237-fcb6-4144-bc61-6ab05643f69b",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Some spatial tools require the spatial coordinates can be stored as `spatial` or `X_spatial`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23af698f-cc41-4383-9155-be1719247ccd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add X_spatial to the right field\n",
    "x_spatial = np.array(adata_section.obs[['x','y']])\n",
    "\n",
    "adata_section.obsm['X_spatial'] = x_spatial\n",
    "adata_section.obsm['X_spatial']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39d12ed0-f129-4e08-a823-7024f30296f5",
   "metadata": {},
   "source": [
    "<font size=\"3\">Gene information are stored in **`.var`**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2875c0df-c66c-4fef-9cc1-e6b381a64f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2edc2f71-e0b9-4ad4-8bcd-4cab02b4b939",
   "metadata": {},
   "source": [
    "<font size=\"3\">Plot the section and color them by all the different classifications avaiable in the metadata (class_color, subclass_color, supertype_color, cluster_color):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8470a4f6-e392-42f3-b15f-2a3f95011163",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the color mapping dictionaries\n",
    "class_palette = dict(zip(adata_section.obs['class'].unique(), adata_section.obs['class_color'].unique()))\n",
    "subclass_palette = dict(zip(adata_section.obs['subclass'].unique(), adata_section.obs['subclass_color'].unique()))\n",
    "supertype_palette = dict(zip(adata_section.obs['supertype'].unique(), adata_section.obs['supertype_color'].unique()))\n",
    "cluster_palette = dict(zip(adata_section.obs['cluster'].unique(), adata_section.obs['cluster_color'].unique()))\n",
    "\n",
    "# Create the subplots\n",
    "fig, axes = plt.subplots(2, 2, figsize=(15, 15))\n",
    "fig.suptitle('Scatter Plots of Different Categories', fontsize=16)\n",
    "\n",
    "# Plot 'class' using class_color\n",
    "sns.scatterplot(ax=axes[0, 0], x='x', y=-adata_section.obs['y'], hue='class', palette=class_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[0, 0].set_title('Class')\n",
    "axes[0, 0].set_xlabel('X Coordinate')\n",
    "axes[0, 0].set_ylabel('Y Coordinate')\n",
    "axes[0, 0].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[0, 0].grid(False)  # Remove grid\n",
    "\n",
    "# Plot 'subclass' using subclass_color\n",
    "sns.scatterplot(ax=axes[0, 1], x='x', y=-adata_section.obs['y'], hue='subclass', palette=subclass_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[0, 1].set_title('Subclass')\n",
    "axes[0, 1].set_xlabel('X Coordinate')\n",
    "axes[0, 1].set_ylabel('Y Coordinate')\n",
    "axes[0, 1].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[0, 1].grid(False)  # Remove grid\n",
    "\n",
    "# Plot 'supertype' using supertype_color\n",
    "sns.scatterplot(ax=axes[1, 0], x='x', y=-adata_section.obs['y'], hue='supertype', palette=supertype_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[1, 0].set_title('Supertype')\n",
    "axes[1, 0].set_xlabel('X Coordinate')\n",
    "axes[1, 0].set_ylabel('Y Coordinate')\n",
    "axes[1, 0].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[1, 0].grid(False)  # Remove grid\n",
    "\n",
    "# Plot 'cluster' using cluster_color\n",
    "sns.scatterplot(ax=axes[1, 1], x='x', y=-adata_section.obs['y'], hue='cluster', palette=cluster_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[1, 1].set_title('Cluster')\n",
    "axes[1, 1].set_xlabel('X Coordinate')\n",
    "axes[1, 1].set_ylabel('Y Coordinate')\n",
    "axes[1, 1].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[1, 1].grid(False)  # Remove grid\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.96])  \n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34f11a84-c472-47a7-a3b5-1ee3569b753c",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Perform PCA and UMAP and cluster with Leiden Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77137142-5f01-452a-9015-7dc8b82dfd15",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(adata_section)\n",
    "sc.pp.neighbors(adata_section)\n",
    "sc.tl.umap(adata_section)\n",
    "sc.tl.leiden(\n",
    "    adata_section, \n",
    "    key_added=\"clusters\", \n",
    "    resolution=1.0,\n",
    "    directed=False, \n",
    "    n_iterations=2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "871dc765-9919-444e-ae31-f0c0be0b97c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(adata_section, color=[\"clusters\"], wspace=0.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bcff9f7-97b8-4664-acde-be61cabc2239",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Since the spatial coordinates are stored in  `.obsm[X_spatial]` you can plot the section with `sc.pl.spatial`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e307ded4-983e-48d4-8a7c-42c4f5f6e66a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the spatial plot\n",
    "sc.pl.spatial(adata_section, color='class', color_map='Accent_r', spot_size=0.04, title='Section 80 colored by Class')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff4a5495-ff32-42df-a6e4-76b17d35a6f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the AnnData object to keep only cells belonging to a specific class\n",
    "# and plot them again with sc.pl.spatial\n",
    "adata_section_one_class = adata_section[adata_section.obs['class']=='01 IT-ET Glut']\n",
    "sc.pl.spatial(adata_section_one_class, color='class', color_map='Accent_r', spot_size=0.04, title='Section 80 (one Class)')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eadb60f0-7cbb-45d6-8ad4-8567079ed879",
   "metadata": {},
   "source": [
    "### Plot gene expression in space\n",
    "\n",
    "<font size=\"3\">\n",
    "\n",
    "Since each spot in the spatial data corresponds to a specific location and its associated gene expression profile, we can visualize spatial gene expression.\n",
    "\n",
    "We can do this using the function `sc.pl.spatial`, it requires the name of the variable index to indicate the genes to plot.\n",
    "\n",
    "Let's take a look at `adata_section`.var:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d001c6ad-0581-4e54-bcc8-5ea0db2d147c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b82b4f1-d8d8-44e6-97f2-81f2dfbee5ed",
   "metadata": {},
   "source": [
    "<font size=\"3\">\n",
    "\n",
    "Change the index of the `adata_section.var` to `gene_symbol` for clarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf8740b3-6142-42df-8318-1179a739d04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index to turn the current index into a column\n",
    "adata_section.var.reset_index(inplace=True)\n",
    "adata_section.var.set_index('gene_symbol', inplace=True)\n",
    "adata_section.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11c787c1-911c-4bbb-b621-fee7010bbeb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define genes to plot\n",
    "genes_toplot = ['Zic1',\n",
    " 'Penk',\n",
    " 'Gfap',\n",
    " 'Nr4a2',\n",
    " 'Satb2']\n",
    "\n",
    "sc.pl.spatial(adata_section, color=genes_toplot,spot_size=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c2e379a-79f5-46ff-94a0-654f9026c150",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Plot your favourite gene/s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643caa15-8d5d-472e-8668-97fd010b768e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes_toplot = [.....]\n",
    "\n",
    "#sc.pl.spatial(adata_section, color=genes_toplot,spot_size=0.04)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c35bfe-bcf8-4d3b-8493-ca07c89f5c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f1f79b0-2ec5-4a6d-93a5-13b97885d6d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gc\n",
    "# collect garbage\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a3d2d5-5727-45df-a641-106b5e7c4460",
   "metadata": {},
   "source": [
    "# Prediction of spatial patterns for unmeasured genes using scRNA-seq data with SpaGE\n",
    "[SpaGE: Spatial Gene Enhancement using scRNA-seq (Abdelaal et al. 2020)](https://academic.oup.com/nar/article/48/18/e107/5909530)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b24a6480-c29d-4399-97c5-1ceb21638f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f356ab26-9e1f-4c01-81e4-739efec8e1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from SpaGE.main import SpaGE\n",
    "import scipy.stats as st"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beef66b4-853d-403f-a573-52053b399a49",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <img src=\"img/spage.jpg\" width=\"700\" height=\"280\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://academic.oup.com/nar/article/48/18/e107/5909530\" target=\"_blank\">Abdelaal et al. (2020) NAR</a></figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19eb7650-ac6d-4b90-aa06-6cc781c6311e",
   "metadata": {},
   "source": [
    "<font size=\"3\"> SpaGE takes as input the (targeted) spatial dataset (query) and a single-cell RNA dataset used as reference. The two datasets should be generated under the same conditions (same species, tissue, region) as much as possible.\n",
    "\n",
    "<font size=\"3\"> We will use [Mouse whole-brain transcriptomic cell type atlas (Hongkui Zeng)](https://www.nature.com/articles/s41586-023-06812-z).This scRNAseq dataset profiled 7 million cells (approximately 4.0 million cells passing quality control). More in detail:\n",
    "\n",
    " - <font size=\"3\"> 1.7 million single cell transcriptomes spanning the whole adult mouse brain using 10Xv2 chemistry (WMB-10Xv2)\n",
    "\n",
    " - <font size=\"3\"> 2.3 million single cell transcriptomes spanning the whole adult mouse brain using 10Xv3 chemistry (WMB-10Xv3)\n",
    "\n",
    " - <font size=\"3\"> 1687 single cell transcriptomes spanning the whole adult mouse brain using the 10X Multiome chemistry (WMB-10XMulti)\n",
    "\n",
    "<figure>\n",
    "    <img src=\"img/sc.png\" width=\"700\" height=\"280\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://www.nature.com/articles/s41586-023-06812-z\" target=\"_blank\">Yao et al. (2023) Nature Methods</a></figcaption>\n",
    "</figure>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3314c600-5aff-4998-9fde-c7dd6185a1dd",
   "metadata": {},
   "source": [
    "<font size=\"3\"> The size is of the combined single-cell object from AbcProjectCache is around 280GB. For this reason, the three objects have been previously downloaded, concatenated, preprocessed and subsetted to obtain NNN cells. For your knowledge, you can find all the steps in `obtain_scref.py`.\n",
    "\n",
    "<font size=\"3\"> Let's read the resulting single-cell AnnData object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "926edba6-5733-4042-84f3-7443ed238f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc = ad.read_h5ad(f\"{data_dir}/SpaGE_data/sc_adata_1percent.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08a97c98-d1c9-42e0-8c2c-597b470d2580",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Inspect the object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2ca8b8-830e-472b-bbb5-56db7fff393b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "656a5fd7-5b6b-4d14-bcd2-73e2cf64f7d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0e98222-9420-4181-868d-cd146b5396db",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Add feature metadata to the AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f6b28a7-7c33-446a-9294-aa92289833a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_csv(f\"{data_dir}/abc_download_root/metadata/WMB-10X/gene.csv\",index_col=0)\n",
    "\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc782a1c-66b6-4eb4-aa95-ea4a2eb6914f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.var = genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aec6d0c-dfa4-4586-b887-70330b219f35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index to turn the current index into a column\n",
    "adata_sc.var.reset_index(inplace=True)\n",
    "adata_sc.var.set_index('gene_symbol', inplace=True)\n",
    "adata_sc.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb3b362-419d-462e-93e0-2c7ddeaa5da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect garbage\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4ac2775-5018-4eb9-a852-2d79a7ec95ce",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Set up inputs for SpaGE:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d55535e-9547-4d92-a173-81077340893b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the count data and transform them into a Pandas DataFrame\n",
    "RNA_data = pd.DataFrame(adata_sc.X.todense())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eb576b9-394b-44c3-ad22-49a40e6c962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add gene names to the RNA_data columns \n",
    "RNA_data.columns = adata_sc.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f266e6-82d4-4ced-8a1c-1aa5daa55ef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter lowely expressed genes\n",
    "Genes_count = np.sum(RNA_data > 0, axis=1)\n",
    "RNA_data = RNA_data.loc[Genes_count >=10,:]\n",
    "del Genes_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb28736-c54d-404f-b0e3-b4e09c99ac69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalization\n",
    "def Log_Norm_cpm(x):\n",
    "    return np.log(((x/np.sum(x))*1000000) + 1)\n",
    "RNA_data = RNA_data.apply(Log_Norm_cpm,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77339674-5f9d-4ddb-b4ba-7cbf2278e417",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transpose RNA_data to have rows = genes and columns = cells\n",
    "RNA_data = RNA_data.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6099b6ab-1c53-4ffa-be3c-c4cda0f5c69f",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ac960e-7c2c-471d-8d80-3bf4d82eee2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop any row with at least one NaN value\n",
    "RNA_data = RNA_data.dropna()  # This will drop any row with at least one NaN value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119e8228-3eda-4786-8e83-a959d7757b44",
   "metadata": {},
   "source": [
    "<span style=\"color:; font-size: 15px;\"> **How many genes do the spatial and single-cell data set have in common?**\n",
    "\n",
    "<span style=\"color:; font-size: 15px;\"> **How many genes are measured in the single-cell dataset that are not measured in the spatial dataset?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2846466a-d702-4e33-a9b3-8e6feafbdc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## use .intersection\n",
    "gene_intersection = set(list(adata_section.var.index)).intersection(set(list(adata_sc.var.index)))\n",
    "\n",
    "## use a set operation\n",
    "genes_not_intersection = list(set(RNA_data.index) - set(adata_section.var.index))\n",
    "\n",
    "print(f\"Number of genes measured both in the sc and spatial datasets: {len(gene_intersection)}\")\n",
    "print(f\"Number of genes measured in the sc reference only: {len(genes_not_intersection)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d616ce55-41de-420b-afa3-93d50b8a461c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract count matrix from adata_section and rename index and column names\n",
    "\n",
    "spatial_data = pd.DataFrame(adata_section.X)\n",
    "spatial_data.index = adata_section.obs.index\n",
    "spatial_data.columns = adata_section.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2608515-710a-462c-8ac2-e835eb413c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize spatial_data \n",
    "spatial_data = spatial_data.apply(Log_Norm_cpm,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a7b24db-b2b6-4b53-861a-074d6edf7276",
   "metadata": {},
   "outputs": [],
   "source": [
    "brain_region_specific_genes = [\n",
    "    \"Bcl11b\", \"Cux2\", \"Fezf2\", \"Tbr1\", \"Satb2\", # isocortex\n",
    "    \"Foxp2\",  # Striatum specific\n",
    "    \"Penk\",   # Basal ganglia specific\n",
    "    \"Zic1\",   # Cerebellum specific\n",
    "    \"Hoxb2\",  # Hindbrain specific\n",
    "    \"Dlx6\",   # Subpallium specific\n",
    "]\n",
    "\n",
    "# Checking for genes from brain_region_specific_genes that are measured by spatial data\n",
    "overlap_genes = list(set(list(spatial_data.columns)) & set(brain_region_specific_genes))\n",
    "overlap_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6720795-f4b1-4770-b965-60e510414ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking for genes from brain_region_specific_genes that are NOT measured by spatial data\n",
    "non_overlap_genes = list(set(brain_region_specific_genes) - set(overlap_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c1606da-6019-4423-ad49-8319dc69653d",
   "metadata": {},
   "source": [
    "### Leave-one-gene-out cross validation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d6dbaa-dcbb-427c-97cd-5361a83b9840",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Define the function to predict genes measured in the spatial dataset to have a ground truth and plot the measured expression next to the predicted one."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "040e064f-e652-4831-bc3a-b3e7ad0b4bc9",
   "metadata": {},
   "source": [
    "```python\n",
    "def predicted_GE_LOOCV(Gene_set, spatial_data, RNA_data):\n",
    "    \n",
    "    # Initialize an empty DataFrame to store the predicted values\n",
    "    predicted_df = pd.DataFrame(index=spatial_data.index)\n",
    "\n",
    "    for i in Gene_set:\n",
    "        # Predict the expression of the current gene using the SpaGE method\n",
    "        Imp_Genes = SpaGE(spatial_data.drop(i, axis=1), RNA_data.T, n_pv=30, genes_to_predict=[i])\n",
    "        Imp_Genes.index = spatial_data.index\n",
    "        \n",
    "        # Add the predicted values for the current gene as a new column in the DataFrame\n",
    "        predicted_df[i] = Imp_Genes[i]\n",
    "\n",
    "    return predicted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e6b551-1e69-4bcb-9d28-e002752176b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "predicted_df = pd.read_csv(f\"{data_dir}/SpaGE_data/predicted_GE_loocv_SpaGE.csv\",index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ece7a1cf-e165-494a-9ba0-dda21c0f7cdc",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Since we have a ground truth, create a Series to store Spearman correlation results for each gene.\n",
    "\n",
    "<font size=\"3\"> Our gene_set in this case is a set of genes shared between the spatial and the sc datasets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b94062d-7ddc-40bc-9a49-7bbd1dfd9143",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.gridspec as gridspec  # For precise subplot layout control\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "def plot_measured_vs_predicted_spatial(Gene_set, spatial_data, predicted_df, adata_section):\n",
    "    \"\"\"\n",
    "    Plot measured and predicted spatial expression for genes, with calculated Spearman correlations.\n",
    "    \n",
    "    Parameters:\n",
    "        Gene_set (list): List of genes to visualize.\n",
    "        spatial_data (DataFrame): Measured expression data (columns are genes, rows are spots).\n",
    "        predicted_df (DataFrame): Predicted expression data (columns are genes, rows are spots).\n",
    "        adata_section (AnnData): Spatial data with 'x' and 'y' coordinates.\n",
    "    \n",
    "    Returns:\n",
    "        dict: Spearman correlations for each gene.\n",
    "    \"\"\"\n",
    "    correlations = {}  # Dictionary to store Spearman correlations\n",
    "    \n",
    "    for gene in Gene_set:\n",
    "        # Calculate Spearman correlation\n",
    "        correlation = spearmanr(spatial_data[gene], predicted_df[gene])[0]\n",
    "        correlations[gene] = correlation\n",
    "        \n",
    "        # Create a figure with two scatterplots\n",
    "        fig = plt.figure(figsize=(15, 8))  # Adjust figure size\n",
    "        gs = gridspec.GridSpec(1, 3, width_ratios=[1, 1, 0.05], wspace=0.3)  # Create grid spec for layout\n",
    "        \n",
    "        # --- Measured Gene Expression ---\n",
    "        ax0 = fig.add_subplot(gs[0])\n",
    "        cmap_measured = spatial_data[gene]\n",
    "        cmap_measured = np.clip(cmap_measured, None, np.percentile(cmap_measured, 99))  # Clip to 99th percentile\n",
    "        sc1 = ax0.scatter(adata_section.obs['x'], -adata_section.obs['y'], s=1, c=cmap_measured, cmap=\"viridis\", alpha=1)\n",
    "        \n",
    "        ax0.set_title(f\"Measured Expression: {gene}\", fontsize=12)\n",
    "        ax0.set_xlabel(\"X Coordinate\")\n",
    "        ax0.set_ylabel(\"Y Coordinate\")\n",
    "        ax0.set_aspect(\"equal\", adjustable=\"box\")\n",
    "        ax0.grid(False)\n",
    "        \n",
    "        # --- Predicted Gene Expression ---\n",
    "        ax1 = fig.add_subplot(gs[1])\n",
    "        cmap_predicted = predicted_df[gene]\n",
    "        cmap_predicted = np.clip(cmap_predicted, None, np.percentile(cmap_predicted, 99))  # Clip to 99th percentile\n",
    "        sc2 = ax1.scatter(adata_section.obs['x'], -adata_section.obs['y'], s=1, c=cmap_predicted, cmap=\"viridis\", alpha=1)\n",
    "        \n",
    "        ax1.set_title(f\"Predicted Expression: {gene}\", fontsize=12)\n",
    "        ax1.set_xlabel(\"X Coordinate\")\n",
    "        ax1.set_ylabel(\"Y Coordinate\")\n",
    "        ax1.set_aspect(\"equal\", adjustable=\"box\")\n",
    "        ax1.grid(False)\n",
    "        \n",
    "        # --- Colorbar ---\n",
    "        cbar_ax = fig.add_subplot(gs[2])  # Third subplot for color bar\n",
    "        cbar = fig.colorbar(sc1, cax=cbar_ax)\n",
    "        cbar.set_label(\"Expression\")\n",
    "        \n",
    "        # Adjust layout and show\n",
    "        plt.tight_layout()\n",
    "        plt.show()\n",
    "    \n",
    "    return correlations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cb74f95-ef1d-4b18-b485-ab93cf2cd217",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlations = plot_measured_vs_predicted_spatial(overlap_genes, spatial_data, predicted_df, adata_section)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6778c0bb-5b1f-41f7-b764-1ac748d2bed3",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's take a look at the Spearman Correlations between measured and predicted gene expression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95639806-6b66-4151-b829-de3c5a400507",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since we have a ground truth, create a Series to store Spearman correlation results for each gene\n",
    "# Our gene_set in this case is a set of genes shared between the spatial and the sc datasets\n",
    "correlations_spage = pd.DataFrame(list(correlations.items()), columns=[\"Gene\", \"correlations_spage\"]).set_index(\"Gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6560a27e-b127-4910-badb-961fb1b904af",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlations_spage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9bab07e-c287-4be4-bed0-dbfde2f9729e",
   "metadata": {},
   "source": [
    "### Predict unmeasured genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39d5347c-7c38-4c90-be33-08cb8262939c",
   "metadata": {},
   "source": [
    "<font size=\"3\">  Define a function to predict unmeasured genes in the spatial dataset. \n",
    "\n",
    "<font size=\"3\">  In this case correlations cannot be calculated as there is no ground truth available."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07eb441-c562-4e27-a086-17baaa5fa4f4",
   "metadata": {},
   "source": [
    "```python\n",
    "\n",
    "def predicted_unmeasured_GE(Gene_set, spatial_data, RNA_data):\n",
    "    \n",
    "    # Initialize an empty DataFrame to store the predicted values\n",
    "    predicted_df = pd.DataFrame(index=spatial_data.index)\n",
    "\n",
    "    for i in Gene_set:\n",
    "        # Predict the expression of the current gene using the SpaGE method\n",
    "        Imp_Genes = SpaGE(spatial_data, RNA_data.T, n_pv=30, genes_to_predict=[i])\n",
    "        Imp_Genes.index = spatial_data.index\n",
    "        \n",
    "        # Add the predicted values for the current gene as a new column in the DataFrame\n",
    "        predicted_df[i] = Imp_Genes[i]\n",
    "\n",
    "    return predicted_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32e16719-adcb-4a05-9656-0b7b77559a49",
   "metadata": {},
   "outputs": [],
   "source": [
    "predicted_df = pd.read_csv(f\"{data_dir}/SpaGE_data/predicted_unmeasured_GE_SpaGE.csv\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b9df78-120c-43d3-b850-c77537271492",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_predicted_spatial(Gene_set, predicted_df, adata_section):\n",
    "    \"\"\"\n",
    "    Plot predicted spatial expression for genes.\n",
    "    \n",
    "    Parameters:\n",
    "        Gene_set (list): List of genes to visualize.\n",
    "        predicted_df (DataFrame): Predicted expression data (columns are genes, rows are spots).\n",
    "        adata_section (AnnData): Spatial data with 'x' and 'y' coordinates.\n",
    "    \n",
    "    Returns:\n",
    "        None\n",
    "    \"\"\"\n",
    "    for gene in Gene_set:\n",
    "        # Create a figure for the predicted expression\n",
    "        fig, ax = plt.subplots(figsize=(8, 6))  # Adjust figure size\n",
    "        \n",
    "        # Get predicted values and clip to the 99th percentile\n",
    "        cmap_predicted = predicted_df[gene]\n",
    "        cmap_predicted = np.clip(cmap_predicted, None, np.percentile(cmap_predicted, 99))\n",
    "        \n",
    "        # Plot the predicted expression\n",
    "        scatter = ax.scatter(\n",
    "            adata_section.obs['x'], -adata_section.obs['y'], \n",
    "            s=1, c=cmap_predicted, cmap=\"viridis\", alpha=1\n",
    "        )\n",
    "        \n",
    "        # Add title and labels\n",
    "        ax.set_title(f\"Predicted Expression: {gene}\", fontsize=14)\n",
    "        ax.set_xlabel(\"X Coordinate\")\n",
    "        ax.set_ylabel(\"Y Coordinate\")\n",
    "        ax.set_aspect(\"equal\", adjustable=\"box\")\n",
    "        ax.grid(False)\n",
    "        \n",
    "        # Add a colorbar\n",
    "        cbar = plt.colorbar(scatter, ax=ax, pad=0.01)\n",
    "        cbar.set_label(\"Expression\")\n",
    "        \n",
    "        # Adjust layout and show\n",
    "        plt.tight_layout()\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab92cc4a-9eb4-4a73-b54c-cb8cb86edd2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the set of genes (from the non_overlap_genes list) to be predicted by SpaGE\n",
    "plot_predicted_spatial(non_overlap_genes, predicted_df, adata_section)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06189e2c-8e20-4a83-8b5f-f231dc51b0cc",
   "metadata": {},
   "source": [
    "<font size=\"3\"> There is no definitive ground truth for the real gene expression profiles of these genes. However, an alternative approach could be to plot region-specific genes and verify whether their expression aligns with the expected brain regions.\n",
    "\n",
    "<font size=\"3\"> For example, Tbr1 is known to be Isocortex-specific; does its expression show higher levels in the Isocortex? Similarly, Foxp2 is Striatum-specific; does it exhibit a Striatum-specific expression pattern?\n",
    "\n",
    "<font size=\"3\"> Use [this](https://mouse.brain-map.org/search/index) resource to check region-specific gene expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "371ac2df-aef5-493a-8ff5-38fb8aa29564",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time SpaGE: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948f6c80-567e-4f56-970d-00d8d1bdaf52",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's delete some objects we won't be using anymore to free some RAM:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e35c884-a934-4070-a0cd-c5d556db8d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "del RNA_data, spatial_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0b7b38d-2264-4e77-b264-14fce33f175d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect garbage\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "371a83c3-c3e8-4b01-a94e-9cd30c952ea3",
   "metadata": {},
   "source": [
    "# Unmeasured gene imputation with Tangram\n",
    "[Deep learning and alignment of spatially resolved single-cell transcriptomes with Tangram (Biancalani et al. 2021)](https://www.nature.com/articles/s41592-021-01264-7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "518b2c69-c6c0-4ba9-ada7-f7a1709c594b",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3000981c-012e-404e-b06e-409e7eaaf9cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tangram as tg\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "sc.settings.set_figure_params(dpi=80, facecolor=\"white\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92e8bf4e-ba3e-4559-8b22-fd89efaa0657",
   "metadata": {},
   "source": [
    "#### Common gene set between reference and spatial dataset\n",
    "\n",
    "<font size=\"3\"> We want to select our training genes. These genes are shared between the two datasets and should capture the biological variance between cell types. For this, we first compute marker genes on the single-cell data and then use Tangramâ€™s preprocessing function to subset to those genes that are also present in the spatial data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "674dbcef-8913-42f0-812a-50d342d679f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = list(set.intersection(set(adata_sc.var_names), set(adata_section.var_names)))\n",
    "len(markers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1740106c-6c6c-4dbb-b9d0-9cb5bd85ceb0",
   "metadata": {},
   "source": [
    "<font size=\"3\"> `tg.pp_adatas` computes the overlap between single-cell data and spatial data on the list of genes provided in the genes argument, it then stores the resulting gene set under 'training_genes' in both adata objects under the `.uns` key and enforces consistent ordering of the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12abfa2-3533-4ac3-a16a-55357663742a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tg.pp_adatas(adata_sc, adata_section, genes=markers) # using the same single-cell reference that we used to run SpaGE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc9ff666-e8fa-4f39-995b-7c16102955a8",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Check that the function performs as we expect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ed67270-9beb-48df-8c48-93a59ae24254",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert \"training_genes\" in adata_sc.uns\n",
    "assert \"training_genes\" in adata_section.uns\n",
    "\n",
    "print(f\"Number of training_genes: {len(adata_sc.uns['training_genes'])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59952400-c0aa-49b7-a2fc-c5bfe048b834",
   "metadata": {},
   "source": [
    "#### Computing the map from single-cells to spatial voxels\n",
    "\n",
    "<font size=\"3\"> Having specified the training genes, we can now create the map from dissociated single-cell measurements to the spatial locations. For this, we are going to use the `map_cells_to_space` function from tangram. This function has two different modes, `mode='cells'` and `mode='clusters'`. The latter only maps averaged single-cells which makes the mapping computationally faster and more robust when mapping between specimen."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb7c48f-11ec-47a8-8836-42c91cec39cb",
   "metadata": {},
   "source": [
    "```python\n",
    "ad_map = tg.map_cells_to_space(\n",
    "    adata_sc,\n",
    "    adata_section,\n",
    "    mode=\"cells\",\n",
    "    density_prior=\"rna_count_based\",\n",
    "    num_epochs=500,\n",
    "    device=\"cpu\",  # or: cpu\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9b23b9-ff51-48b3-a89a-a8f1945d373a",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We run `tg.map_cells_to_space` for you (takes a while to run), let's just read the resulting AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8849a82d-8176-4cff-93e0-ab07d8e14cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_map = ad.read_h5ad(f\"{data_dir}/Tangram_data/ad_map_tangram.h5ad\")\n",
    "ad_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c9ff6ac-225a-4c63-9f45-76a7b80660d0",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We observe that Tangramâ€™s mapping from cell i to spatial voxels j is stored in the .X property of ad_map.\n",
    "<font size=\"3\">  The meaning of the .var and .obs also changes:\n",
    "\n",
    "- <font size=\"3\"> in `.var` we have the available metadata of the spatial data, adata_section\n",
    "- <font size=\"3\"> in `.obs` we have the available metadata of the single-cell data, adata_sc\n",
    "- <font size=\"3\"> In addition, the information about the training run is stored in the .uns key, see `.uns['training_genes_df']` and `.uns[training_history]`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc992f7b-4151-487a-954a-913ca9a2b7a1",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Now we can **project the genes present in the single-cell data to the spatial locations**. This is easily achieved by multiplying the mapping matrix stored in ad_map with the original single-cell data stored in adata_sc. Tangram already provides a convenience function which takes in a mapping and the corresponding single-cell data. The result is a **spatial voxel by genes matrix** which technically is identical to the original spatial data adata_section but **contains expression values for all genes**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d93c4214-3d7c-4127-8666-6301d7811564",
   "metadata": {},
   "source": [
    "```python\n",
    "ad_ge = tg.project_genes(adata_map=ad_map, adata_sc=adata_sc)\n",
    "ad_ge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d44756c-050f-4630-adc5-209f74745ff8",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We run `tg.project_genes` for you, let's just read the resulting AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62f70844-0288-433e-879d-70438a1ae988",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_ge = ad.read_h5ad(f\"{data_dir}/Tangram_data/ad_ge.h5ad\")\n",
    "\n",
    "## add .obsm['X_spatial'] so we can use sc.pl.spatial to plot gene expression\n",
    "ad_ge.obsm['X_spatial'] = np.array(ad_ge.obs[['x','y']])\n",
    "ad_ge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c62d08f-d3ba-4f54-b426-78f2de41f449",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Next, we will compare the new spatial data with the original measurements.\n",
    "\n",
    "<font size=\"3\"> To do this we will plot the true expression and the predicted expression of the shared genes, as well as the Spearman correlation between the true and predicted expression in all the spots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd04efe5-5689-409b-bd10-02df368a2563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the same set of genes that we predicted with SpaGE\n",
    "# we need to make the genes all lower-case first\n",
    "overlap_genes = [gene.lower() for gene in overlap_genes]\n",
    "overlap_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5af4c94c-9cc9-476b-9602-4d6c023879e8",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Plot the true expression from adata_section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed36a6d2-750f-42af-ae30-b57791bca2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(adata_section, color=overlap_genes, spot_size=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc740ed-fa95-48e2-aa1f-0ecc22bc94e9",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Plot the predicted expression from ad_ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ac35b9-d0e5-4097-a22f-3b40164cf025",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_ge.var.reset_index(inplace=True)\n",
    "ad_ge.var.set_index('_index', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddf2247b-bdaf-428b-8b18-bfc2d54f77b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(ad_ge, color=overlap_genes, spot_size=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a56f93c8-7f87-440a-bdd2-5666c295012d",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Calculate Spearman correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55bdb832-b668-46ff-8f2a-fe7b87bb0325",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the AnnData objects to include only the genes in overlap_genes\n",
    "# Assuming both AnnData objects have the same gene ordering\n",
    "ad_true  = adata_section[:, adata_section.var_names.isin(overlap_genes)]\n",
    "ad_ge_predicted = ad_ge[:, ad_ge.var_names.isin(overlap_genes)]\n",
    "\n",
    "# Initialize an empty dictionary to store correlations\n",
    "correlation_dict = {}\n",
    "\n",
    "# Loop over each gene to calculate Spearman correlation\n",
    "for gene in overlap_genes:\n",
    "    # Extract gene expression vectors for the current gene\n",
    "    expr1 = ad_true[:, gene].X.flatten()\n",
    "    expr2 = ad_ge_predicted[:, gene].X.flatten()\n",
    "    \n",
    "    # Calculate Spearman correlation for the current gene\n",
    "    corr, _ = st.spearmanr(expr1, expr2)\n",
    "    \n",
    "    # Store the result in the dictionary\n",
    "    correlation_dict[gene] = corr\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "tangram_correlations = pd.DataFrame.from_dict(correlation_dict, orient='index', columns=['correlations_tangram'])\n",
    "# capitalize gene names! \n",
    "tangram_correlations.index = tangram_correlations.index.str.capitalize()\n",
    "tangram_correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2b5b88e-79b8-467d-bd59-bed051161aaf",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Since we predicted the same genes as SpaGE, let's compare the Spearman correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3def41c2-1a5e-48a6-95af-9a97bb26ddd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_correlations = pd.concat([correlations_spage, tangram_correlations],axis=1)\n",
    "all_correlations[['correlations_spage','correlations_tangram']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b77bb9-ea35-4e9b-b80e-84c88bdb9702",
   "metadata": {},
   "source": [
    "<font size=\"3\"> **Which tool results in higher correlation between the true-measured gene expression and the predicted expression of the genes of interest?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db86a1c6-50f9-4762-9214-50e9df47082e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time Tangram: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d1582ca-cfd9-461b-98ef-1ec0518cbf61",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's delete some objects we won't be using anymore to free some RAM:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ad5b9bb-74e1-40d6-9114-6f89aa370a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "del ad_map, ad_ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11f7048-4838-404e-8a13-e45970b3b8c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "del all_correlations, correlations_spage, tangram_correlations, predicted_df, ad_true, ad_ge_predicted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c5ea683-eb3d-4a06-acee-39e37c88db8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d79c3ae8-b6c1-45a4-ae98-f99a13254e2f",
   "metadata": {},
   "source": [
    "# Cell type deconvolution with cell2location\n",
    "[Cell2location maps fine-grained cell types in spatial transcriptomics (Kleshchevnikov et al. 2022)](https://www.nature.com/articles/s41587-021-01139-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dcf3b2c-3b5c-4483-8e45-04e5ef8510f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36bc28b5-9bb2-4fb2-b67d-56b009b92d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import squidpy as sq\n",
    "import cell2location as c2l\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "sc.settings.set_figure_params(dpi=80, facecolor=\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e41ce22b-0a76-4aa9-b8c0-417d2ddc210e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set raw layer as X\n",
    "adata_sc.layers['raw'] = adata_sc.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f02cdba1-fa61-4f02-92b4-36e3230966a6",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Add metadata with cell type annotation to the single-cell AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ebc5d6d-8bf3-4dd7-965f-48abeefd0fdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read csv\n",
    "metadata = pd.read_csv(f\"{data_dir}/abc_download_root/metadata/WMB-10X/views/cell_metadata_with_cluster_annotation.csv\",index_col=0)\n",
    "# filter metadata to keep only obervations in the subsetted SC AnnData object\n",
    "metadata = metadata.loc[metadata.index.intersection(adata_sc.obs.index)]\n",
    "# reindex rows\n",
    "metadata = metadata.reindex(adata_sc.obs.index)\n",
    "# assign metadata to adata_sc.obs (in this way metadata replaces all fields previously in obs)\n",
    "adata_sc.obs = metadata\n",
    "\n",
    "adata_sc.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a612f0f-9842-4314-8bd6-8e6b97e33d6b",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Identify the shared set of genes between spatial and single-cell "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9faae07-e863-4a9f-adcc-abe61138fb09",
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_features = [feature for feature in adata_section.var_names if feature in adata_sc.var_names]\n",
    "\n",
    "# filter \n",
    "adata_sc = adata_sc[:, adata_sc.var.index.isin(shared_features)].copy()\n",
    "adata_section = adata_section[:, shared_features].copy() # probably no need to filter the spatial gene set\n",
    "# reorder genes to have them match between spatial data and sc\n",
    "adata_sc = adata_sc[:, shared_features].copy()\n",
    "\n",
    "adata_section.shape[1] == adata_sc.shape[1]\n",
    "print(f'Number of shared genes between spatial data and reference sc: {adata_section.shape[1]}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "399a4e1b-1b77-4547-948c-050838e3f9aa",
   "metadata": {},
   "source": [
    "<font size=\"3\"> **Fit the reference model**\n",
    "\n",
    "<font size=\"3\"> When your spatial transcriptomics dataset is whole-transcriptome: The default parameters cell_count_cutoff=5, cell_percentage_cutoff2=0.03, nonz_mean_cutoff=1.12 are a good starting point, however, you can increase the cut-off to exclude more genes. To preserve marker genes of rare cell types we recommend low cell_count_cutoff=5, however, cell_percentage_cutoff2 and nonz_mean_cutoff can be increased to select between 8k-16k genes.\n",
    "\n",
    "<font size=\"3\"> In this 2D histogram, orange rectangle highlights genes excluded based on the combination of number of cells expressing that gene (Y-axis) and average RNA count for cells where the gene was detected (X-axis).\n",
    "\n",
    "<font size=\"3\"> This step can be skipped when a limited gene panel is available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2afb3635-7b4f-41cd-ad82-b1bed671ec9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fit the reference model\n",
    "selected = c2l.utils.filtering.filter_genes(\n",
    "    adata_sc, cell_count_cutoff=5, cell_percentage_cutoff2=0.03, nonz_mean_cutoff=1.12\n",
    ")\n",
    "\n",
    "adata_sc = adata_sc[:, selected].copy()\n",
    "adata_section = adata_section[:, selected].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "922e8303-ec18-4064-858e-a31dc7d13426",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"The spatial AnnData object has {adata_section.shape[0]} obs and {adata_section.shape[1]} features.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f970565-e3ee-4103-aac0-134ece4858ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"The single-cell AnnData object has {adata_sc.shape[0]} obs and {adata_sc.shape[1]} features.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dbd7a17-91e5-4e4c-83c2-d9c0e0724903",
   "metadata": {},
   "source": [
    "<font size=\"3\"> **Estimation of reference cell type signatures (NB regression)**\n",
    "\n",
    "<font size=\"3\"> The following two command are computationally intensive and require GPU. We run them for you. You can read the output from `stdata_deconv.h5ad`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76ffd25c-3900-4795-8d74-65414e27b144",
   "metadata": {},
   "source": [
    "<font size=\"3\"> The signatures are estimated from scRNA-seq data, accounting for batch effect, using a Negative binomial regression model.\n",
    "\n",
    "<font size=\"3\"> First, prepare anndata object for the regression model:\n",
    "\n",
    "```python\n",
    "c2l.models.RegressionModel.setup_anndata(\n",
    "    adata=adata_sc,\n",
    "    #batch_key=\"Replicates\", # there's no batch \n",
    "    labels_key=\"class\", # cell type = class\n",
    "    #categorical_covariate_keys=[\"assay\"],\n",
    "    layer=\"raw\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ac8a390-0b83-4355-808a-42ff8f132b1d",
   "metadata": {},
   "source": [
    "```python\n",
    "model = c2l.models.RegressionModel(adata_sc)\n",
    "# default, try on GPU:\n",
    "model.train(max_epochs=250, batch_size=2500, train_size=1, lr=0.002)\n",
    "\n",
    "model.export_posterior(\n",
    "    adata_sc,\n",
    "    sample_kwargs={\"num_samples\": 1000, \"batch_size\": 2500},\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d0bec05-144b-411a-8bd3-96f3ed505fe7",
   "metadata": {},
   "source": [
    "<font size=\"3\"> **Train the model** to estimate the reference cell type signatures:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f229c86-2274-456a-ac9a-35e3ab46a85b",
   "metadata": {},
   "source": [
    "```python\n",
    "# export estimated expression in each cluster\n",
    "if \"means_per_cluster_mu_fg\" in adata_sc.varm.keys():\n",
    "    inf_aver = adata_sc.varm[\"means_per_cluster_mu_fg\"][\n",
    "        [f\"means_per_cluster_mu_fg_{i}\" for i in adata_sc.uns[\"mod\"][\"factor_names\"]]\n",
    "    ].copy()\n",
    "else:\n",
    "    inf_aver = adata_sc.var[\n",
    "        [f\"means_per_cluster_mu_fg_{i}\" for i in adata_sc.uns[\"mod\"][\"factor_names\"]]\n",
    "    ].copy()\n",
    "\n",
    "inf_aver.columns = adata_sc.uns[\"mod\"][\"factor_names\"]\n",
    "inf_aver.head()\n",
    "\n",
    "# cell 2 location cell type mapping\n",
    "print('start with Cell2location')\n",
    "c2l.models.Cell2location.setup_anndata(\n",
    "    adata=adata_section,\n",
    ")\n",
    "\n",
    "model = c2l.models.Cell2location(\n",
    "    adata_section,\n",
    "    cell_state_df=inf_aver,\n",
    "    N_cells_per_location=8,\n",
    ")\n",
    "model.view_anndata_setup()\n",
    "\n",
    "print('start with training')\n",
    "model.train(max_epochs=30000, batch_size=None, train_size=1)#, use_gpu=use_gpu)\n",
    "\n",
    "adata_section = model.export_posterior(\n",
    "    adata_section,\n",
    "    sample_kwargs={\n",
    "        \"num_samples\": 1000,\n",
    "        \"batch_size\": model.adata.n_obs}#,\n",
    "        #\"use_gpu\": use_gpu,\n",
    ")\n",
    "\n",
    "\n",
    "adata_section.obs[adata_section.uns[\"mod\"][\"factor_names\"]] = adata_section.obsm[\"q05_cell_abundance_w_sf\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c0db0b9-dc67-4937-9515-dcf4b33597c1",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Read the pre-computed AnnData object with cell2location results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38655df4-9b3c-47a8-a6ca-2f615a4abefb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section = ad.read_h5ad(f\"{data_dir}/cell2location_data/stdata_deconv.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb809a2-a820-4736-bcde-1323d70999b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### cell2location result\n",
    "adata_section.obs[adata_section.obs['class'].unique()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bead7dec-0f78-44b8-bf2d-da0c54b4b626",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Visualize cell abundance in spatial coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1b028a4-1367-4d66-bbd9-f1d7dc0cb626",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(adata_section, color=['01 IT-ET Glut','04 DG-IMN Glut','30 Astro-Epen'],spot_size=0.04)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be766873-5e86-44bc-8602-0ae7aba1c511",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time cell2location: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4488a9be-81ac-418e-91f4-d73c77667049",
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect garbage\n",
    "gc.collect()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "p1_integration_sc",
   "language": "python",
   "name": "p1_integration_sc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
