<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Niche reconstruction and spatial domain detection – Spatial Omics Data Analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../assets/elixir_favicon.ico" rel="icon">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-bf0ef3c727e58b0ba1c4a46410c3c68e.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<script src="https://unpkg.com/@jupyter-widgets/html-manager@*/dist/embed-amd.js" crossorigin="anonymous"></script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Spatial Omics Data Analysis</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../../../index.html" aria-current="page"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../practicals.html"> 
<span class="menu-text">Practicals</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../presentations.html"> 
<span class="menu-text">Presentations</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/elixir-europe-training/ELIXIR-SCO-spatial-omics/"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Niche reconstruction and spatial domain detection</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../../../index.html" class="sidebar-logo-link">
      <img src="../../../assets/ELIXIR1.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#dataset" id="toc-dataset" class="nav-link active" data-scroll-target="#dataset">Dataset</a></li>
  <li><a href="#cell-neighborhood-detection-via-graph-construction" id="toc-cell-neighborhood-detection-via-graph-construction" class="nav-link" data-scroll-target="#cell-neighborhood-detection-via-graph-construction">1. Cell neighborhood detection via graph construction</a></li>
  <li><a href="#spatial-domain-detection-with-cellcharter" id="toc-spatial-domain-detection-with-cellcharter" class="nav-link" data-scroll-target="#spatial-domain-detection-with-cellcharter">2. Spatial Domain detection with CellCharter</a>
  <ul class="collapse">
  <li><a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction">2.1 Dimensionality reduction</a></li>
  <li><a href="#neighborhood-aggregation" id="toc-neighborhood-aggregation" class="nav-link" data-scroll-target="#neighborhood-aggregation">2.2 Neighborhood aggregation</a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering">2.3 Clustering</a></li>
  <li><a href="#downstream-analyses" id="toc-downstream-analyses" class="nav-link" data-scroll-target="#downstream-analyses">2.4 Downstream analyses</a></li>
  </ul></li>
  <li><a href="#banksy" id="toc-banksy" class="nav-link" data-scroll-target="#banksy">3. BANKSY</a>
  <ul class="collapse">
  <li><a href="#construct-the-k_geom-nn-graph" id="toc-construct-the-k_geom-nn-graph" class="nav-link" data-scroll-target="#construct-the-k_geom-nn-graph">Construct the <span class="math inline">\(k_{geom}\)</span>-NN graph</a></li>
  <li><a href="#generate-spatial-weights-from-distance" id="toc-generate-spatial-weights-from-distance" class="nav-link" data-scroll-target="#generate-spatial-weights-from-distance">Generate spatial weights from distance</a></li>
  <li><a href="#generate-banksy-matrix" id="toc-generate-banksy-matrix" class="nav-link" data-scroll-target="#generate-banksy-matrix">Generate BANKSY matrix</a></li>
  <li><a href="#reduce-dimensions-of-each-data-matrix" id="toc-reduce-dimensions-of-each-data-matrix" class="nav-link" data-scroll-target="#reduce-dimensions-of-each-data-matrix">Reduce dimensions of each data matrix</a></li>
  <li><a href="#cluster-cells-using-a-partition-algorithm" id="toc-cluster-cells-using-a-partition-algorithm" class="nav-link" data-scroll-target="#cluster-cells-using-a-partition-algorithm">Cluster cells using a partition algorithm</a></li>
  <li><a href="#plot-results" id="toc-plot-results" class="nav-link" data-scroll-target="#plot-results">Plot results</a></li>
  <li><a href="#investigate-cell-type-composition-in-each-banksy-defined-spatial-domain" id="toc-investigate-cell-type-composition-in-each-banksy-defined-spatial-domain" class="nav-link" data-scroll-target="#investigate-cell-type-composition-in-each-banksy-defined-spatial-domain">Investigate Cell Type composition in each Banksy-defined Spatial Domain</a></li>
  </ul></li>
  <li><a href="#compare-cellcharter-vs-bansky" id="toc-compare-cellcharter-vs-bansky" class="nav-link" data-scroll-target="#compare-cellcharter-vs-bansky">Compare CellCharter vs BANSKY</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Niche reconstruction and spatial domain detection</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Authors: Francesca Drummer, Marco Varrone</p>
<p>In this notebook we will cover:</p>
<ol type="1">
<li>Graph construction and analysis of spatial transcriptomics data using Squidpy</li>
<li>CellCharter</li>
<li>BANSKY</li>
</ol>
<div id="cell-2" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>load_ext autoreload</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>autoreload <span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-3" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Data analysis and ML imports</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># single-cell imports</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># import squidpy as sq</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> warnings</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>warnings.filterwarnings(<span class="st">"ignore"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="dataset" class="level2">
<h2 class="anchored" data-anchor-id="dataset">Dataset</h2>
<p>We will use the Xenium AD dataset from the previous notebooks here.</p>
<p>As a reminder the dataset consists of 6 coronal mouse brain slices from 2 different conditions (wildtype - ctrl vs TgCRND8 - AD) across 3 timepoints. In this practical, we additionally have information about cell types available in <code>adata.obs['cell_types']</code>. Please note that these annotation are not perfect. For example, there are quite some cells that could not be assigned to a cell type (NaN or “unkown”). These annotations have been made with on leiden clustering and marker genes reported in <a href="https://pages.10xgenomics.com/rs/446-PBO-704/images/10x_LIT000210_App-Note_Xenium-In-Situ_Letter_Digital.pdf">this</a> document.</p>
<p>In this practical we aim to understand the differences of the mouse brain between the two conditions and across the timepoints using niches and spatial domains.</p>
<div id="cell-5" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>PATH <span class="op">=</span> <span class="st">"/home/icb/francesca.drummer/1-Projects/"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-6" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load adata</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(Path(PATH, <span class="st">'xenium_mouse_ad_annotated_rotated.h5ad'</span>))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>AnnData object with n_obs × n_vars = 350209 × 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="cell-7" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating a DataFrame from 'split', 'fov', and 'condition'</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> adata.obs[[<span class="st">'condition'</span>, <span class="st">'time'</span>, <span class="st">'batch_key'</span>]]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>value_counts <span class="op">=</span> pd.DataFrame(df.values, columns<span class="op">=</span>df.columns).value_counts()</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(value_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>condition  time  batch_key
TgCRND8    17_9  2            60825
wildtype   13_4  5            59886
           5_7   4            58683
TgCRND8    5_7   1            58681
wildtype   2_5   3            58230
TgCRND8    2_5   0            53904
Name: count, dtype: int64</code></pre>
</div>
</div>
<div id="cell-8" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add "Unknown" as a category</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"cell_types"</span>] <span class="op">=</span> adata.obs[<span class="st">"cell_types"</span>].cat.add_categories(<span class="st">"Unknown"</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Fill NaN values with "Unknown"</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"cell_types"</span>] <span class="op">=</span> adata.obs[<span class="st">"cell_types"</span>].fillna(<span class="st">"Unknown"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="cell-neighborhood-detection-via-graph-construction" class="level2">
<h2 class="anchored" data-anchor-id="cell-neighborhood-detection-via-graph-construction">1. Cell neighborhood detection via graph construction</h2>
<p>Spatial transcriptomics data can be represented as graphs with cells as nodes and edges as relations. Depending on the technology (imaging-based or sequencing-based) different assumptions can be made for the graph structure. Here we will explore graph construction and cell neighborhood analysis using the squidpy <code>sq.gr.spatial_neighbors</code> module on imaging-based data.</p>
<p><em>Information for graph reconstruciton for sequencing-based data can be found <a href="https://squidpy.readthedocs.io/en/stable/notebooks/examples/graph/compute_spatial_neighbors.html">here</a>.</em></p>
<p>For image-based technologies we construct a graph with <code>coord_type=generic</code>, meaning that the nodes / cells will preserve their spatial location and will not be re-arranged in a spatial grid. For generic graph approaches we can choose to set a fixed radius <code>radius</code> or number of neighbors to connect to (<code>n_neighs</code>).</p>
<p>Below we try both graph construction methods and give them a different <code>key</code> to add to the <code>adata</code> object.</p>
<div id="cell-11" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="kw">del</span> adata.uns[<span class="st">'cell_types_colors'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-12" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'cell_types_colors'</span> <span class="kw">in</span> adata.uns:</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">del</span> adata.uns[<span class="st">'cell_types_colors'</span>]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>sq.gr.spatial_neighbors(adata, n_neighs<span class="op">=</span><span class="dv">10</span>, coord_type<span class="op">=</span><span class="st">"generic"</span>, key_added <span class="op">=</span> <span class="st">'neighs_based_spatial'</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    library_key <span class="op">=</span> <span class="st">'batch_key'</span>,</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"cell_types"</span>],</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    connectivity_key<span class="op">=</span><span class="st">"neighs_based_spatial_connectivities"</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    size <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-9-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-13" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sq.gr.spatial_neighbors(adata, radius<span class="op">=</span><span class="fl">0.3</span>, coord_type<span class="op">=</span><span class="st">"generic"</span>, key_added <span class="op">=</span> <span class="st">'radius_based_spatial'</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>    library_key <span class="op">=</span> <span class="st">'batch_key'</span>, </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">"cell_types"</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>    connectivity_key<span class="op">=</span><span class="st">"radius_based_spatial_connectivities"</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-10-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>First differences across the graphs can be observed by visual inspection, e.i. the cells that are further away from the clear brain structures are connected in the clostest neighbor approach but unconnected in the radius-based approach. Some other differences are harder to observe like dependencies in the more dense connected regions. For this Squidpy provides a number of statistics to better the differences between the connectivities of cells.</p>
<p>Using the cell type information we can formulate some hypothesis from the data.</p>
<p><span style="color: red;"><strong>Task 1:</strong> Try out some analysis from Squidpy to identify distinctions between the graphs e.i. using the <a href="https://squidpy.readthedocs.io/en/stable/notebooks/examples/graph/compute_centrality_scores.html">centrality score</a> and <a href="https://squidpy.readthedocs.io/en/stable/notebooks/examples/graph/compute_nhood_enrichment.html">neighborhood enrichment</a>. </span></p>
<p>Example questions to answer could be: What are the average number of connections per cell type? Which cell types tend to cluster together?</p>
<div id="cell-16" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>sq.gr.nhood_enrichment(adata, cluster_key<span class="op">=</span><span class="st">"cell_types"</span>, library_key <span class="op">=</span> <span class="st">'condition'</span>, connectivity_key <span class="op">=</span> <span class="st">"neighs_based_spatial"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>sq.pl.nhood_enrichment(</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    adata, cluster_key<span class="op">=</span><span class="st">"cell_types"</span>, method<span class="op">=</span><span class="st">"average"</span>, figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">5</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>)  <span class="co"># `method` compute the hierarchic</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"82ee0da3e51447b8ba3f0fa77f1c9eb4","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-11-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-17" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>sq.gr.nhood_enrichment(adata, cluster_key<span class="op">=</span><span class="st">"cell_types"</span>, library_key <span class="op">=</span> <span class="st">'condition'</span>, connectivity_key <span class="op">=</span> <span class="st">"radius_based_spatial"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sq.pl.nhood_enrichment(</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    adata, cluster_key<span class="op">=</span><span class="st">"cell_types"</span>, method<span class="op">=</span><span class="st">"average"</span>, figsize<span class="op">=</span>(<span class="dv">5</span>, <span class="dv">5</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>)  <span class="co"># `method` compute the hierarchic</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<script type="application/vnd.jupyter.widget-view+json">
{"model_id":"99c37b5acf7f471d9f781fbbd6092ded","version_major":2,"version_minor":0,"quarto_mimetype":"application/vnd.jupyter.widget-view+json"}
</script>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-12-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-18" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sq.gr.centrality_scores(adata, cluster_key <span class="op">=</span> <span class="st">"cell_types"</span>, connectivity_key <span class="op">=</span> <span class="st">"neighs_based_spatial"</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>sq.pl.centrality_scores(adata, cluster_key <span class="op">=</span> <span class="st">"cell_types"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-19" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sq.gr.centrality_scores(adata, cluster_key <span class="op">=</span> <span class="st">"cell_types"</span>, connectivity_key <span class="op">=</span> <span class="st">"radius_based_spatial"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>sq.pl.centrality_scores(adata, cluster_key <span class="op">=</span> <span class="st">"cell_types"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-14-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><span style="color: red;"><strong>Task 2:</strong> How do the scores change when increasing or decreasing the number of neighbors <code>n_neighs = 50</code> or <code>radius=200</code>? </span></p>
<p>Once we have a graph construction suitable for the data representation we can perform multiple differential analysis between the different conditions (TgCRND8 vs wildtype).</p>
<section id="comparing-entire-graphs" class="level4">
<h4 class="anchored" data-anchor-id="comparing-entire-graphs">1.1.1 Comparing entire graphs</h4>
<p>GraphCompass introduces two holistic graph comparison metrics:</p>
<ol type="1">
<li>Westfeiler-Lehman Graph Kernels</li>
<li>Filtration curves</li>
</ol>
<p>Both methods compute a graph embedding that can be compared against one another. The graph embeddings are stored in <code>adata.uns["wl_kernel"]</code> and <code>adata.uns["filtration_curves"]</code> respectively.</p>
<div id="cell-24" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">"unknown"</span> <span class="kw">not</span> <span class="kw">in</span> adata.obs[<span class="st">"cell_types"</span>].cat.categories:</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    adata.obs[<span class="st">"cell_types"</span>] <span class="op">=</span> adata.obs[<span class="st">"cell_types"</span>].cat.add_categories(<span class="st">"unknown"</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"cell_types"</span>] <span class="op">=</span> adata.obs[<span class="st">"cell_types"</span>].fillna(<span class="st">"unknown"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-25" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>library_key <span class="op">=</span> <span class="st">"batch_key"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>cluster_key <span class="op">=</span> <span class="st">"cell_types"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>condition_key <span class="op">=</span> <span class="st">"condition"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To lower the computational load we sub-sample the data. Let’s take a subsample from the different conditions but same time points.</p>
<div id="cell-27" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Creating a DataFrame from 'split', 'fov', and 'condition'</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> adata.obs[[<span class="st">'condition'</span>, <span class="st">'time'</span>]]</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>value_counts <span class="op">=</span> pd.DataFrame(df.values, columns<span class="op">=</span>df.columns).value_counts()</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(value_counts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>condition  time
TgCRND8    17_9    60825
wildtype   13_4    59886
           5_7     58683
TgCRND8    5_7     58681
wildtype   2_5     58230
TgCRND8    2_5     53904
Name: count, dtype: int64</code></pre>
</div>
</div>
<div id="cell-28" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sub_adata <span class="op">=</span> adata[adata.obs[<span class="st">'time'</span>] <span class="op">==</span> <span class="st">'5_7'</span>]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>sub_adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>View of AnnData object with n_obs × n_vars = 117364 × 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap', 'neighs_based_spatial_neighbors', 'radius_based_spatial_neighbors', 'cell_types_centrality_scores'
    obsm: 'X_pca', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances', 'neighs_based_spatial_connectivities', 'neighs_based_spatial_distances', 'radius_based_spatial_connectivities', 'radius_based_spatial_distances'</code></pre>
</div>
</div>
</section>
<section id="westfeiler-lehman-graph-kernels" class="level4">
<h4 class="anchored" data-anchor-id="westfeiler-lehman-graph-kernels">1.1.1.1 Westfeiler-Lehman Graph Kernels</h4>
<div id="cell-30" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>gc.tl.wlkernel.compare_conditions(</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>   adata<span class="op">=</span>sub_adata,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>   library_key<span class="op">=</span>library_key,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>   cluster_key<span class="op">=</span>cluster_key,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>   compute_spatial_graphs<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>   kwargs_spatial_neighbors<span class="op">=</span>{</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">'coord_type'</span>: <span class="st">'generic'</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">'delaunay'</span>: <span class="va">True</span>,  </span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Computing spatial graphs...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 1000/1000 [00:02&lt;00:00, 356.64/s]
100%|██████████| 1000/1000 [00:02&lt;00:00, 367.18/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Defining node features...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 2/2 [07:59&lt;00:00, 239.95s/it]
INFO:root:Continuous node features provided, using CONTINUOUS propagation scheme.</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Computing Wasserstein distance between conditions...</code></pre>
</div>
</div>
<div id="cell-31" class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>control_group<span class="op">=</span><span class="st">"normal"</span>  <span class="co"># reference group</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>metric_key<span class="op">=</span><span class="st">"wasserstein_distance"</span> </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>method<span class="op">=</span><span class="st">"wl_kernel"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-32" class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>gc.pl.wlkernel.compare_conditions(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>sub_adata,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span>library_key,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    condition_key<span class="op">=</span>condition_key,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>    control_group<span class="op">=</span>control_group,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>    metric_key<span class="op">=</span>metric_key,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>    method<span class="op">=</span>method,</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">3</span>,<span class="dv">5</span>),</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>    dpi<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="filtration-curves" class="level4">
<h4 class="anchored" data-anchor-id="filtration-curves">1.1.1.2 Filtration curves</h4>
<div id="cell-34" class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>gc.tl.filtration_curves.compare_conditions(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>sub_adata,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span>library_key,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    cluster_key<span class="op">=</span>cluster_key,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    condition_key<span class="op">=</span>condition_key,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    compute_spatial_graphs<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-35" class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>node_labels<span class="op">=</span>[<span class="st">"Astrocytes"</span>, <span class="st">"Microglia"</span>, <span class="st">"Oligodendrocytes"</span>]  <span class="co"># node labels (e.g. cell types) we are interested in visualising</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>metric_key<span class="op">=</span><span class="st">"filtration_curves"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-36" class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>gc.pl.filtration_curves.compare_conditions(</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>    adata<span class="op">=</span>adata,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>    node_labels<span class="op">=</span>node_labels,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    metric_key<span class="op">=</span>metric_key,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>    palette<span class="op">=</span><span class="st">"Set2"</span>,</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    dpi<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">20</span>,<span class="dv">6</span>),</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>    right<span class="op">=</span><span class="dv">4</span>,  <span class="co"># to shorten the x-axis and improve readability</span></span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="spatial-domain-detection-with-cellcharter" class="level2">
<h2 class="anchored" data-anchor-id="spatial-domain-detection-with-cellcharter">2. Spatial Domain detection with CellCharter</h2>
<div id="cell-38" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scvi</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> squidpy <span class="im">as</span> sq</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> cellcharter <span class="im">as</span> cc</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gdown</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> logging</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>logger <span class="op">=</span> logging.getLogger(<span class="st">'pytorch_lightning.utilities.rank_zero'</span>)</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>logger.setLevel(logging.ERROR)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-39" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>scvi.settings.seed <span class="op">=</span> <span class="dv">12345</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO: Seed set to 12345
INFO:lightning.fabric.utilities.seed:Seed set to 12345</code></pre>
</div>
</div>
<div id="cell-40" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>PATH <span class="op">=</span> <span class="st">"/home/varrone/work_curnagl/Projects/ELIXIR-SCO-spatial-omics/day_3/practical_4/data/anndata"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s reload the data. In this way you can see what is required for the analysis independently of the analyses in the previous sections.</p>
<div id="cell-42" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load adata</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> sc.read_h5ad(Path(PATH, <span class="st">'xenium_mouse_ad_annotated_rotated.h5ad'</span>))</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>adata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>AnnData object with n_obs × n_vars = 350209 × 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'cell_types_colors', 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<p>We have the <code>condition</code> and <code>time</code> labels, but we don’t have a single label as an identifier for the samples. <br> We will create a new label <code>sample</code> that will be the concatenation of the <code>condition</code> and <code>time</code> labels.</p>
<div id="cell-44" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'sample'</span>] <span class="op">=</span> adata.obs[<span class="st">'condition'</span>].astype(<span class="bu">str</span>)<span class="op">+</span><span class="st">'_'</span><span class="op">+</span>adata.obs[<span class="st">'time'</span>].astype(<span class="bu">str</span>)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'sample'</span>] <span class="op">=</span> adata.obs[<span class="st">'sample'</span>].astype(<span class="st">'category'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>There may be cells with very low counts. We will filter them out.</p>
<div id="cell-46" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, min_counts<span class="op">=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I have already run the cluster stability analysis using 3 layers of neighbors, as it may take too long for this practical session. (It took 25 minutes on a GPU). <br> The cluster stability analysis works by running clustering multiple times for every possible number of clusters and measuring the similarity (Adjusted Rand Index) between similar numbers of clusters.</p>
<p>We can download the model, plot the stability curve and the look for the peak(s) to identify the optimal number of clusters.</p>
<div id="cell-48" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># gdown.download('https://drive.google.com/file/d/18A9UszEYyIOwC7v7ShBqQulCUKLFlQd7/view?usp=sharing', output=os.path.join(PATH, f'autok_l3.zip'), fuzzy=True)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-49" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!unzip -n -d {PATH} {os.path.join(PATH, f'autok_l3.zip')}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-50" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>autok <span class="op">=</span> cc.tl.ClusterAutoK.load(Path(PATH, <span class="ss">f'autok_l3'</span>))</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>cc.pl.autok_stability(autok)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-33-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Here we don’t see a clear peak, but we can see that the highest stability is for 18 clusters.</p>
<p>We will use 18 clusters for the clustering analysis.</p>
<p>We could use the clustering model generated during the stability analysis, but we will fit a new clustering model to let you see the process.</p>
<section id="dimensionality-reduction" class="level3">
<h3 class="anchored" data-anchor-id="dimensionality-reduction">2.1 Dimensionality reduction</h3>
<p>First, we need to run the dimensionality reduction. We will use scVI for this.</p>
<p>Make sure that <code>adata.X</code> contains count data, as it is required by scVI. <br> In some tutorials you will see that the count data is stored in an AnnData layer called <code>counts</code>. Here we will use the <code>X</code> layer.</p>
<div id="cell-53" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>scvi.model.SCVI.setup_anndata(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">#layer="counts"</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>INFO:jax._src.xla_bridge:Unable to initialize backend 'cuda': 
INFO:jax._src.xla_bridge:Unable to initialize backend 'rocm': module 'jaxlib.xla_extension' has no attribute 'GpuAllocatorConfig'
INFO:jax._src.xla_bridge:Unable to initialize backend 'tpu': INTERNAL: Failed to open libtpu.so: libtpu.so: cannot open shared object file: No such file or directory</code></pre>
</div>
</div>
<div id="cell-54" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>LOAD_MODEL <span class="op">=</span> <span class="va">True</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We set the parameters of the neural network. Here we use 1 layer and an embedding size of 10.<br> These are very common parameters, but you can play with them to see how they affect the results.</p>
<div id="cell-56" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> LOAD_MODEL:</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>    os.makedirs(os.path.join(PATH, <span class="st">'scvi_model'</span>), exist_ok<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>    gdown.download(<span class="st">'https://drive.google.com/file/d/1IDA_VReiI_elV055MSPa03EJNy1_OEks/view?usp=sharing'</span>, output<span class="op">=</span>os.path.join(PATH, <span class="st">'scvi_model'</span>, <span class="st">'model.pt'</span>), fuzzy<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>    model <span class="op">=</span> scvi.model.SCVI.load(os.path.join(PATH, <span class="st">'scvi_model'</span>), adata<span class="op">=</span>adata)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="cf">else</span>:</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>    model <span class="op">=</span> scvi.model.SCVI(</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>        adata,</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>        n_layers<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>        n_latent<span class="op">=</span><span class="dv">10</span>,</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>        use_layer_norm<span class="op">=</span><span class="st">"both"</span>,</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>        use_batch_norm<span class="op">=</span><span class="st">"none"</span>,</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>    model.train(early_stopping<span class="op">=</span><span class="va">True</span>, enable_progress_bar<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading...
From: https://drive.google.com/uc?id=1IDA_VReiI_elV055MSPa03EJNy1_OEks
To: /home/varrone/work_curnagl/Projects/ELIXIR-SCO-spatial-omics/day_3/practical_4/data/anndata/scvi_model/model.pt
100%|██████████| 4.08M/4.08M [00:00&lt;00:00, 97.8MB/s]</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO     File                                                                                                      
         /home/varrone/work_curnagl/Projects/ELIXIR-SCO-spatial-omics/day_3/practical_4/data/anndata/scvi_model/mod
         el.pt already downloaded                                                                                  </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
</div>
<p>To make sure that the training has converged, we can plot the training history.<br> Here we is important to focus on the (validation) reconstruction loss. This shows how well the model is able to reconstruct the original data.</p>
<div id="cell-58" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>fig, axes <span class="op">=</span> plt.subplots(ncols<span class="op">=</span><span class="dv">3</span>, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">5</span>))</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>metrics <span class="op">=</span> [<span class="st">"elbo"</span>, <span class="st">"reconstruction_loss"</span>, <span class="st">"kl_local"</span>]</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> ax, metric <span class="kw">in</span> <span class="bu">zip</span>(axes, metrics):</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    ax.plot(</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>        model.history[<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">_train"</span>],</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>        label<span class="op">=</span><span class="st">"train"</span>,</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>        color<span class="op">=</span><span class="st">"darkgreen"</span>,</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>        linewidth<span class="op">=</span><span class="fl">1.25</span></span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>    ax.plot(</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>        model.history[<span class="ss">f"</span><span class="sc">{</span>metric<span class="sc">}</span><span class="ss">_validation"</span>],</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>        label<span class="op">=</span><span class="st">"validation"</span>,</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>        color<span class="op">=</span><span class="st">"firebrick"</span>,</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>        linewidth<span class="op">=</span><span class="fl">1.25</span></span>
<span id="cb51-15"><a href="#cb51-15" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb51-16"><a href="#cb51-16" aria-hidden="true" tabindex="-1"></a>    ax.legend()</span>
<span id="cb51-17"><a href="#cb51-17" aria-hidden="true" tabindex="-1"></a>    ax.set_title(metric)</span>
<span id="cb51-18"><a href="#cb51-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb51-19"><a href="#cb51-19" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-37-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We extract the latent representation of the data.</p>
<div id="cell-60" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">'X_scVI'</span>] <span class="op">=</span> model.get_latent_representation(adata).astype(np.float32)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="neighborhood-aggregation" class="level3">
<h3 class="anchored" data-anchor-id="neighborhood-aggregation">2.2 Neighborhood aggregation</h3>
<p>Until now, all the analyses have been done ignoring the spatial information. <br> Now we will use the spatial information to perform the clustering.</p>
<p>First, we need to create a network where cells are connected if they are close to each other using the Delaunay triangulation.</p>
<div id="cell-63" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>sq.gr.spatial_neighbors(adata, library_key<span class="op">=</span><span class="st">'sample'</span>, coord_type<span class="op">=</span><span class="st">'generic'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-64" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>    adata, </span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span><span class="st">'sample'</span>,</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>    library_id<span class="op">=</span>adata.obs[<span class="st">'sample'</span>].cat.categories[<span class="dv">0</span>],</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">"sample"</span>, </span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">1</span>, </span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>    connectivity_key<span class="op">=</span><span class="st">"spatial_connectivities"</span>,</span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">1</span></span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-40-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As you can see, the Delaunay triangulation generates very long edges for a few cells. <br> The most appropriate approach would be to estimate the most biologically relevant distance between cells and use it to remove the edges longer than that distance.</p>
<p>A quick alternative solution is that, since those long edges are sort of outlines, the measure the 99th percentile of the edge lengths and remove the edges longer than that distance. <br> This process will lead to some isolated cells, but that’s not a problem.</p>
<div id="cell-66" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>sq.gr.spatial_neighbors(adata, library_key<span class="op">=</span><span class="st">'sample'</span>, coord_type<span class="op">=</span><span class="st">'generic'</span>, percentile<span class="op">=</span><span class="dv">99</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-67" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>    adata, </span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span><span class="st">'sample'</span>,</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>    library_id<span class="op">=</span>adata.obs[<span class="st">'sample'</span>].cat.categories[<span class="dv">0</span>],</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span><span class="st">"sample"</span>, </span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">1</span>, </span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>    connectivity_key<span class="op">=</span><span class="st">"spatial_connectivities"</span>,</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">1</span></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-42-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We construct the neighborhood aggregated representation by combining every cell’s features with the ones of first 3 layers of neighbors.<br> We will use the scVI latent representation as features and the new features will be stored in <code>adata.obsm['X_cellcharter']</code>.</p>
<p>While <code>X_scVI</code> contains 10 features, <code>X_cellcharter</code> will contain 10*(n_layers+1)=40 features.</p>
<div id="cell-69" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>cc.gr.aggregate_neighbors(adata, n_layers<span class="op">=</span><span class="dv">3</span>, use_rep<span class="op">=</span><span class="st">'X_scVI'</span>, out_key<span class="op">=</span><span class="st">'X_cellcharter'</span>, sample_key<span class="op">=</span><span class="st">'sample'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 6/6 [00:05&lt;00:00,  1.01it/s]</code></pre>
</div>
</div>
</section>
<section id="clustering" class="level3">
<h3 class="anchored" data-anchor-id="clustering">2.3 Clustering</h3>
<p>Finally, let’s cluster the cells and find their spatial domains.</p>
<div id="cell-72" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>gmm <span class="op">=</span> cc.tl.Cluster(n_clusters<span class="op">=</span><span class="dv">18</span>, random_state<span class="op">=</span><span class="dv">12345</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>gmm.fit(adata, use_rep<span class="op">=</span><span class="st">'X_cellcharter'</span>)</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'spatial_domain'</span>] <span class="op">=</span> gmm.predict(adata, use_rep<span class="op">=</span><span class="st">'X_cellcharter'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Epoch 51: 100%|██████████| 1/1 [00:00&lt;00:00,  1.81it/s, nll=11.10]
Predicting DataLoader 0: 100%|██████████| 1/1 [00:00&lt;00:00,  4.98it/s]</code></pre>
</div>
</div>
<p>We can now plot domains and cell types back to back</p>
<div id="cell-74" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'spatial_domain_colors'</span> <span class="kw">in</span> adata.uns:</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">del</span> adata.uns[<span class="st">'spatial_domain_colors'</span>]</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>    adata, </span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span><span class="st">'sample'</span>, </span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"spatial_domain"</span>, <span class="st">"cell_types"</span>], </span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),</span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>np.repeat(adata.obs[<span class="st">'sample'</span>].cat.categories, <span class="dv">2</span>),</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">2</span></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-45-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="downstream-analyses" class="level3">
<h3 class="anchored" data-anchor-id="downstream-analyses">2.4 Downstream analyses</h3>
<p>After seeing how we can obtain the spatial domains, let’s load the domains from the ClusterAutoK that have been computed in advance.</p>
<p>In this way, we are going to have all the same results, which will make it easier for the discussion.</p>
<div id="cell-77" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'spatial_domain_18'</span>] <span class="op">=</span> autok.predict(adata, use_rep<span class="op">=</span><span class="st">'X_cellcharter'</span>, k<span class="op">=</span><span class="dv">18</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-78" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'spatial_domain_18_colors'</span> <span class="kw">in</span> adata.uns:</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">del</span> adata.uns[<span class="st">'spatial_domain_18_colors'</span>]</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>    adata, </span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span><span class="st">'sample'</span>, </span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"spatial_domain_18"</span>, <span class="st">"cell_types"</span>], </span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>np.repeat(adata.obs[<span class="st">'sample'</span>].cat.categories, <span class="dv">2</span>),</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">2</span></span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-47-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can see that the domains are different from the ones obtained before, but only slightly.</p>
<p>As you can see, domains tend to correspond to areas with a similar composition of cell types.<br> This is not always true, because there are also areas, like the cortex layers, that seem to have similar composition of cell types.<br> By relying directly on gene expression rather than cell type labels, CellCharter is able to identify domains with similar cell type composition, but with those cell types being different transcriptionally.<br> This is because, even within the same cell type, there can be differences in the transcriptional profile.</p>
<p>Can we quantify this more directly? <br> The cell type enrichment measures the likelihood of a cell type being found in a domain compared to random chance.<br> Domains with similar cell type enrichment are likely to be separated into different domains because those same cell types have different transcriptional profiles.<br> We are not going to test this here.</p>
<div id="cell-81" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>cc.gr.enrichment(</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>    group_key<span class="op">=</span><span class="st">'spatial_domain'</span>,</span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>    label_key<span class="op">=</span><span class="st">'cell_types'</span>,</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>cc.pl.enrichment(</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>    group_key<span class="op">=</span><span class="st">'spatial_domain'</span>,</span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a>    label_key<span class="op">=</span><span class="st">'cell_types'</span>,</span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>    dot_scale<span class="op">=</span><span class="dv">9</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-48-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The <a href="https://atlas.brain-map.org/atlas?atlas=1&amp;plate=100960076#atlas=1&amp;plate=100960520&amp;resolution=6.98&amp;x=5512.001546223959&amp;y=3967.997233072917&amp;zoom=-2">Allen Brain Atlas</a> provides images and annotation of mouse brain samples at multiple depths.</p>
<p>Choose one of the six samples, go to the atlas and try to find the image that better matches the regions shown by CellCharter.<br> 1. Which of the 132 images is the closest one? 2. After identifying the closest image, try to match the domains obtained with the annotated regions in the atlas. 3. Do you find anatomical differences between the two conditions? are there differences in spatial domain compositions? Are there unique domains in one of the two conditions (Alzheimer vs wildtype)? - Focus, in particular on comparing samples from mice of the same age. - If you find difference between the two samples, what do you think is the cause?</p>
<p>Now, save the spatial domains corresponding to 9 and 23 clusters into the <code>adata.obs</code> column <code>spatial_domain_9</code> and <code>spatial_domain_23</code>.</p>
<div id="cell-85" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'spatial_domain_18_colors'</span> <span class="kw">in</span> adata.uns:</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">del</span> adata.uns[<span class="st">'spatial_domain_18_colors'</span>]</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>    adata, </span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span><span class="st">'sample'</span>, </span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"spatial_domain_9"</span>, <span class="st">"spatial_domain"</span>, <span class="st">"spatial_domain_23"</span>], </span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>np.repeat(adata.obs[<span class="st">'sample'</span>].cat.categories, <span class="dv">3</span>),</span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">3</span></span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-50-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The three levels of clustering form a sort of hierarchical structure.<br> 9-cluster domains tend to separate into subdomains in the 18 clusters, and so on.</p>
<p>Go back to the Allen Brain Atlas and look if some of the hierarchical structure is reflected in the atlas.</p>
<section id="shape-characterization" class="level4">
<h4 class="anchored" data-anchor-id="shape-characterization">Shape characterization</h4>
<p>Now, we are going to look at the shapes of the domains we obtained.</p>
<p>We first find the local components of the domains. This is done by first finding the connected components of the spatial graph.</p>
<div id="cell-89" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>cc.gr.connected_components(adata, cluster_key<span class="op">=</span><span class="st">'spatial_domain_18'</span>, min_cells<span class="op">=</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-90" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Hackfix: squidpy's spatial_scatter has some issues with categorical data with NaNs.</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">'component_tmp'</span>] <span class="op">=</span> adata.obs[<span class="st">'component'</span>].astype(<span class="st">'str'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-91" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="st">'component_tmp_colors'</span> <span class="kw">in</span> adata.uns:</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">del</span> adata.uns[<span class="st">'component_tmp_colors'</span>]</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>sq.pl.spatial_scatter(</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    adata[(adata.obs[<span class="st">'sample'</span>].isin([<span class="st">'TgCRND8_17_9'</span>, <span class="st">'TgCRND8_2_5'</span>]))], </span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    shape<span class="op">=</span><span class="va">None</span>, </span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>    library_key<span class="op">=</span><span class="st">'sample'</span>, </span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"component_tmp"</span>, <span class="st">"spatial_domain_18"</span>], </span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>    size<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>    figsize<span class="op">=</span>(<span class="dv">10</span>,<span class="dv">10</span>),</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>    title<span class="op">=</span>np.repeat([<span class="st">'TgCRND8_17_9'</span>, <span class="st">'TgCRND8_2_5'</span>], <span class="dv">2</span>),</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">2</span></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>... storing 'component_tmp' as categorical</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-53-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Then draw a boundary around every component</p>
<div id="cell-93" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>cc.tl.boundaries(adata, alpha_start<span class="op">=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-94" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> cellcharter_utils.utils <span class="im">import</span> plot_boundaries, plot_shape_metrics</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-95" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>plot_boundaries(adata, sample<span class="op">=</span><span class="st">'wildtype_5_7'</span>, show_cells<span class="op">=</span><span class="va">True</span>, cells_radius<span class="op">=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO     Using 'datashader' backend with 'None' as reduction method to speed up plotting. Depending on the         
         reduction method, the value range of the plot might change. Set method to 'matplotlib' do disable this    
         behaviour.                                                                                                
INFO     alpha component of given RGBA value for outline color is discarded, because outline_alpha takes precedent.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-56-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>And finally, by computing some shape metrics that we will use to compare the domains.</p>
<div id="cell-97" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>cc.tl.curl(adata)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>cc.tl.linearity(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at the shape metrics of one of the cortex layers, domain 8.</p>
<div id="cell-99" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>plot_shape_metrics(adata, cluster_key<span class="op">=</span><span class="st">'spatial_domain_18'</span>, figsize<span class="op">=</span>(<span class="dv">6</span>,<span class="dv">3</span>), cluster_id<span class="op">=</span><span class="dv">8</span>, metrics<span class="op">=</span>[<span class="st">'curl'</span>, <span class="st">'linearity'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-58-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Look at the spatial domain 0. - Is it going to have lower, same or higher linearity than domain 8? - What about curl?</p>
<p>After you have answered, plot the shape metrics for domain 0 and check if it matches your expectations.</p>
<p>Tip: you can pass multiple cluster ids to <code>plot_shape_metrics</code> in the form of a list to plot multiple domains at once.</p>
<p>Look at spatial domain 12. - What do you expect to see in terms of curl and linearity? - What about elongation?</p>
<p>After answering, compute the elongation and compare the shape metrics for domains 8 and 12.<br> You can check <a href="https://cellcharter.readthedocs.io/en/latest/tools.html">CellCharter’s documentation</a> for the elongation metric.</p>
</section>
</section>
</section>
<section id="banksy" class="level2">
<h2 class="anchored" data-anchor-id="banksy">3. BANKSY</h2>
<p>In this notebook we are using the Python implementation of <a href="https://github.com/prabhakarlab/Banksy_py">BANSKY</a>.</p>
<p>We run BANKSY on a example section, select a section that you find interesting.</p>
<div id="cell-105" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> adata[(adata.obs[<span class="st">'time'</span>] <span class="op">==</span> <span class="st">'5_7'</span>) <span class="op">&amp;</span> (adata.obs[<span class="st">'condition'</span>] <span class="op">==</span> <span class="st">'wildtype'</span>)]</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>adata_section</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>View of AnnData object with n_obs × n_vars = 58683 × 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample'
    var: 'gene_ids', 'feature_types', 'genome'
    uns: 'dea_leiden_res0_25', 'dendrogram_leiden_res0_25', 'leiden', 'leiden_res0_25_colors', 'leiden_res0_5_colors', 'leiden_res1_colors', 'neighbors', 'umap'
    obsm: 'X_pca', 'X_umap', 'spatial'
    obsp: 'connectivities', 'distances'</code></pre>
</div>
</div>
<div id="cell-106" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>adata_section.obsm[<span class="st">'spatial'</span>][:,<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>ArrayView([ 834.40947266,  854.25923462, 4259.28774414, ...,
           4623.79787598, 4627.65275879, 4577.76008301])</code></pre>
</div>
</div>
<div id="cell-107" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co">## add x and y coordinate to .obs (needed for plot later)</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>adata_section.obs[<span class="st">'x'</span>] <span class="op">=</span> adata_section.obsm[<span class="st">'spatial'</span>][:, <span class="dv">0</span>]</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>adata_section.obs[<span class="st">'y'</span>] <span class="op">=</span> adata_section.obsm[<span class="st">'spatial'</span>][:, <span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-108" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy_utils.load_data <span class="im">import</span> load_adata, display_adata</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy_utils.filter_utils <span class="im">import</span> normalize_total, filter_hvg, print_max_min</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalizes the AnnData object</span></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>adata_section <span class="op">=</span> normalize_total(adata_section)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>--- Max-Min before normalization -----
Displaying max and min of Dataset
Max: 136.0, Min: 0.0

--- Max-Min after normalization -----
Displaying max and min of Dataset
Max: 222.99998474121094, Min: 0.0
</code></pre>
</div>
</div>
<p>BANSKY unifies spatially informed cell type and domain segmentation. In this notebook we will focus on the domain segmentation part.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../figures/BANKSY_fig1A.png" class="img-fluid figure-img"></p>
<figcaption>../figures/BANKSY_fig1A.png</figcaption>
</figure>
</div>
<p>The idea is to create a representation of a cell using 1) its own transcriptomic profile (purple) and 2) local microenvironment (red + light pink). The local microenvironment is represented through a pair of spatial kernels that represent the mean gene expression of the local microenvironment (red) and its gradient calculated with the azimuthal Gabor filter (AGF) (light pink).</p>
<p>The relative contribution of the microenvironment is captured by <span class="math inline">\(\lambda\)</span>. Smaller settings of <span class="math inline">\(\lambda\)</span> decrease the influence of the cells within the microenvironment (<span class="math inline">\(\lambda = 0\)</span> reduces to nonspatial informed clustering). <span class="math inline">\(G(r)\)</span> is a radially symmetric Gaussian kernel that decays from magnitude 1 at distance = 0.</p>
<p>The main BANSKY algorithm requires:</p>
<ol type="1">
<li>Creating a kNN graph by setting a spatial number of neighbords <code>num_neighbors</code> (<code>k_geom</code>) parameter.</li>
<li>Assigning weights to the edges of the conected spatial graph. By default, we use the <code>gaussian decay</code> option, where weights decay as a function of distance to the index cell with <span class="math inline">\(\sigma\)</span> <code>= sigma</code>.</li>
<li>Defining whether to use the Azumithal Gabor Filter kernel (<code>max_m = 1</code>) or just the mean expression (<code>max_m = 0</code>).</li>
</ol>
<p>First, we set the required parameters.</p>
<div id="cell-111" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>coord_keys <span class="op">=</span> (<span class="st">'x'</span>, <span class="st">'y'</span>, <span class="st">'spatial'</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set parameters </span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>plot_graph_weights <span class="op">=</span> <span class="va">True</span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>k_geom <span class="op">=</span> <span class="dv">15</span> <span class="co"># number of neighbors</span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>max_m <span class="op">=</span> <span class="dv">1</span> <span class="co"># azumithal transform up to kth order</span></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>nbr_weight_decay <span class="op">=</span> <span class="st">"scaled_gaussian"</span> <span class="co"># can also be "reciprocal", "uniform" or "ranked"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="construct-the-k_geom-nn-graph" class="level3">
<h3 class="anchored" data-anchor-id="construct-the-k_geom-nn-graph">Construct the <span class="math inline">\(k_{geom}\)</span>-NN graph</h3>
<div id="cell-113" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy.main <span class="im">import</span> median_dist_to_nearest_neighbour</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Find median distance to closest neighbours, the median distance will be `sigma`</span></span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>nbrs <span class="op">=</span> median_dist_to_nearest_neighbour(adata_section, key <span class="op">=</span> coord_keys[<span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Median distance to closest cell = 13.144794274179642

---- Ran median_dist_to_nearest_neighbour in 0.43 s ----
</code></pre>
</div>
</div>
</section>
<section id="generate-spatial-weights-from-distance" class="level3">
<h3 class="anchored" data-anchor-id="generate-spatial-weights-from-distance">Generate spatial weights from distance</h3>
<p>Here, we generate the spatial weights using the gaussian decay function from the median distance to the k-th nearest neighbours as specified earlier.</p>
<div id="cell-115" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy.initialize_banksy <span class="im">import</span> initialize_banksy</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>plt.style.use(<span class="st">'default'</span>)</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>banksy_dict <span class="op">=</span> initialize_banksy(</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>    adata_section,</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>    coord_keys,</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>    k_geom,</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>    nbr_weight_decay<span class="op">=</span>nbr_weight_decay,</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>    max_m<span class="op">=</span>max_m,</span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>    plt_edge_hist<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>    plt_nbr_weights<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>    plt_agf_angles<span class="op">=</span><span class="va">False</span>, <span class="co"># takes long time to plot</span></span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>    plt_theta<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Median distance to closest cell = 13.144794274179642

---- Ran median_dist_to_nearest_neighbour in 0.35 s ----

---- Ran generate_spatial_distance_graph in 0.79 s ----

---- Ran row_normalize in 0.20 s ----

---- Ran generate_spatial_weights_fixed_nbrs in 2.44 s ----

----- Plotting Edge Histograms for m = 0 -----

Edge weights (distances between cells): median = 27.25631288860389, mode = 15.46042177259126

---- Ran plot_edge_histogram in 0.44 s ----


Edge weights (weights between cells): median = 0.05928184183882009, mode = 0.03207643565586688

---- Ran plot_edge_histogram in 0.09 s ----

---- Ran generate_spatial_distance_graph in 1.19 s ----

---- Ran theta_from_spatial_graph in 0.49 s ----

---- Ran row_normalize in 0.20 s ----

---- Ran generate_spatial_weights_fixed_nbrs in 3.37 s ----

----- Plotting Edge Histograms for m = 1 -----

Edge weights (distances between cells): median = 38.16559284558646, mode = 28.827712230132924

---- Ran plot_edge_histogram in 0.09 s ----


Edge weights (weights between cells): median = (-4.8232956030589535e-05-0.0243067639100562j), mode = 0.01005536554634287

---- Ran plot_edge_histogram in 0.18 s ----

----- Plotting Weights Graph -----
Maximum weight: 0.18868491395566991

---- Ran plot_graph_weights in 7.78 s ----

Maximum weight: (0.08484586948704541+0.0052739192523142565j)

---- Ran plot_graph_weights in 14.63 s ----

----- Plotting theta Graph -----</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-66-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-66-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-66-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-66-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-66-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="generate-banksy-matrix" class="level3">
<h3 class="anchored" data-anchor-id="generate-banksy-matrix">Generate BANKSY matrix</h3>
<p>The BANKSY matrix considers a cells transcriptomic profile and local microenvironment (Figure 1).</p>
<p>As mentioned before, <span class="math inline">\(\lambda\)</span> is a mixing parameter that controls the importance of cells’ own expression and neighborhood expression effects, it takes values from 0, being spatial information not used in the clustering, to 1, giving the maximum importance to the neighborhood expression.</p>
<p>o generate the BANKSY matrix, we proceed with the following:</p>
<ol type="1">
<li><p>Matrix multiply sparse CSR weights matrix with cell-gene matrix to get <strong>neighbourhood matrix</strong> and the <strong>AGF matrix</strong> if <code>max_m &gt; 1</code></p></li>
<li><p>Z-score both matrices along <strong>genes</strong></p></li>
<li><p>Multiply each matrix by a weighting factor <span class="math inline">\(\lambda\)</span> (We refer to this parameter as lambda in our manuscript and code)</p></li>
<li><p>Concatenate the matrices along the genes dimension in the form -&gt; <code>horizontal_concat(cell_mat, nbr_mat, agf_mat)</code></p></li>
</ol>
<p>Here, we save all the results in the dictionary (banksy_dict), which contains the results from the subsequent operations for BANKSY.</p>
<div id="cell-117" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy.embed_banksy <span class="im">import</span> generate_banksy_matrix</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co"># The following are the main hyperparameters for BANKSY</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>lambda_list <span class="op">=</span> [<span class="fl">0.6</span>] <span class="co"># list of lambda parameters</span></span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>banksy_dict, banksy_matrix <span class="op">=</span> generate_banksy_matrix(adata_section, banksy_dict, lambda_list, max_m)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Runtime Jan-05-2025-17-26

347 genes to be analysed:
Gene List:
Index(['2010300C02Rik', 'Abca7', 'Acsbg1', 'Acta2', 'Acvrl1', 'Adamts2',
       'Adamtsl1', 'Adgrl4', 'Aldh1a2', 'Aldh1l1',
       ...
       'Unc13c', 'Vat1l', 'Vcan', 'Vim', 'Vip', 'Vwc2l', 'Wfs1', 'Zfp366',
       'Zfp536', 'Zfpm2'],
      dtype='object', length=347)

Check if X contains only finite (non-NAN) values
Decay Type: scaled_gaussian
Weights Object: {'weights': {0: &lt;58683x58683 sparse matrix of type '&lt;class 'numpy.float64'&gt;'
    with 880245 stored elements in Compressed Sparse Row format&gt;, 1: &lt;58683x58683 sparse matrix of type '&lt;class 'numpy.complex128'&gt;'
    with 1760490 stored elements in Compressed Sparse Row format&gt;}}

Nbr matrix | Mean: 0.59 | Std: 1.72
Size of Nbr | Shape: (58683, 347)
Top 3 entries of Nbr Mat:

[[0.47219332 0.         0.48879083]
 [0.43831427 0.         0.46033802]
 [1.52354818 0.53943839 0.02748267]]

AGF matrix | Mean: 0.14 | Std: 0.29
Size of AGF mat (m = 1) | Shape: (58683, 347)
Top entries of AGF:
[[0.2776847  0.         0.16484106]
 [0.22993549 0.         0.1525506 ]
 [0.54008628 0.21168626 0.08738869]]
Ran 'Create BANKSY Matrix' in 0.14 mins

Cell by gene matrix has shape (58683, 347)

Scale factors squared: [0.4 0.4 0.2]
Scale factors: [0.63245553 0.63245553 0.4472136 ]
Shape of BANKSY matrix: (58683, 1041)
Type of banksy_matrix: &lt;class 'anndata._core.anndata.AnnData'&gt;
</code></pre>
</div>
</div>
<div id="cell-118" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy.main <span class="im">import</span> concatenate_all</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>banksy_dict[<span class="st">"nonspatial"</span>] <span class="op">=</span> {</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Here we simply append the nonspatial matrix (adata.X) to obtain the nonspatial clustering results</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>    <span class="fl">0.0</span>: {<span class="st">"adata"</span>: concatenate_all([adata_section.X], <span class="dv">0</span>, adata<span class="op">=</span>adata_section), }</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(banksy_dict[<span class="st">'nonspatial'</span>][<span class="fl">0.0</span>][<span class="st">'adata'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Scale factors squared: [1.]
Scale factors: [1.]
AnnData object with n_obs × n_vars = 58683 × 347
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample', 'x', 'y'
    var: 'gene_ids', 'feature_types', 'genome', 'is_nbr', 'k'</code></pre>
</div>
</div>
</section>
<section id="reduce-dimensions-of-each-data-matrix" class="level3">
<h3 class="anchored" data-anchor-id="reduce-dimensions-of-each-data-matrix">Reduce dimensions of each data matrix</h3>
<p>We utilize two common methods for dimensionality reduction:</p>
<ol type="1">
<li><p>PCA (using <code>scikit-learn</code>), we reduce the size of thematrix from <span class="math inline">\(3 * N_{genes}\)</span> to <code>pca_dims</code>. As a default settings, we reduce to 20 dimensions.</p></li>
<li><p>UMAP (<code>UMAP</code> package), which we use to visualize expressions of clusters in the umap space (2-D space).</p></li>
</ol>
<div id="cell-120" class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="co">## Define hyperparameters</span></span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>resolutions <span class="op">=</span> [<span class="fl">0.1</span>] <span class="co"># clustering resolution for UMAP</span></span>
<span id="cb92-4"><a href="#cb92-4" aria-hidden="true" tabindex="-1"></a>pca_dims <span class="op">=</span> [<span class="dv">20</span>] <span class="co"># Dimensionality in which PCA reduces to</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-121" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy_utils.umap_pca <span class="im">import</span> pca_umap</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>pca_umap(banksy_dict,</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>         pca_dims <span class="op">=</span> pca_dims,</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>         add_umap <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>         plt_remaining_var <span class="op">=</span> <span class="va">False</span>,</span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>         )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Current decay types: ['scaled_gaussian', 'nonspatial']

Reducing dims of dataset in (Index = scaled_gaussian, lambda = 0.6)
==================================================

Setting the total number of PC = 20
Original shape of matrix: (58683, 1041)
Reduced shape of matrix: (58683, 20)
------------------------------------------------------------
min_value = -12.5038547814052, mean = -5.592029328484308e-16, max = 19.625682591950902

Conducting UMAP and adding embeddings to adata.obsm["reduced_pc_20_umap"]
UMAP embedding
------------------------------------------------------------
shape: (58683, 2)


AxisArrays with keys: reduced_pc_20, reduced_pc_20_umap

Reducing dims of dataset in (Index = nonspatial, lambda = 0.0)
==================================================

Setting the total number of PC = 20
Original shape of matrix: (58683, 347)
Reduced shape of matrix: (58683, 20)
------------------------------------------------------------
min_value = -20.638336181640625, mean = 6.7142519810659e-07, max = 36.85270309448242

Conducting UMAP and adding embeddings to adata.obsm["reduced_pc_20_umap"]
UMAP embedding
------------------------------------------------------------
shape: (58683, 2)


AxisArrays with keys: reduced_pc_20, reduced_pc_20_umap</code></pre>
</div>
</div>
</section>
<section id="cluster-cells-using-a-partition-algorithm" class="level3">
<h3 class="anchored" data-anchor-id="cluster-cells-using-a-partition-algorithm">Cluster cells using a partition algorithm</h3>
<p>We then cluster cells using the <strong>leiden</strong> algorithm partition methods. Other clustering algorithms include <em>louvain</em> (another resolution based clustering algorithm), or <em>mclust</em> (a clustering based on gaussian mixture model).</p>
<div id="cell-124" class="cell" data-execution_count="51">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy.cluster_methods <span class="im">import</span> run_Leiden_partition</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>seed <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>results_df, max_num_labels <span class="op">=</span> run_Leiden_partition(</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>    banksy_dict,</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>    resolutions,</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>    num_nn <span class="op">=</span> <span class="dv">50</span>,</span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>    num_iterations <span class="op">=</span> <span class="op">-</span><span class="dv">1</span>,</span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>    partition_seed <span class="op">=</span> seed,</span>
<span id="cb95-9"><a href="#cb95-9" aria-hidden="true" tabindex="-1"></a>    match_labels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb95-10"><a href="#cb95-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Decay type: scaled_gaussian
Neighbourhood Contribution (Lambda Parameter): 0.6
reduced_pc_20 

reduced_pc_20_umap 

PCA dims to analyse: [20]

====================================================================================================
Setting up partitioner for (nbr decay = scaled_gaussian), Neighbourhood contribution = 0.6, PCA dimensions = 20)
====================================================================================================


Nearest-neighbour weighted graph (dtype: float64, shape: (58683, 58683)) has 2934150 nonzero entries.
---- Ran find_nn in 61.94 s ----


Nearest-neighbour connectivity graph (dtype: int16, shape: (58683, 58683)) has 2934150 nonzero entries.

(after computing shared NN)
Allowing nearest neighbours only reduced the number of shared NN from 37737077 to 2932178.


Shared nearest-neighbour (connections only) graph (dtype: int16, shape: (58683, 58683)) has 2861385 nonzero entries.

Shared nearest-neighbour (number of shared neighbours as weights) graph (dtype: int16, shape: (58683, 58683)) has 2861385 nonzero entries.

sNN graph data:
[30 18 12 ...  5 10  9]

---- Ran shared_nn in 2.37 s ----


-- Multiplying sNN connectivity by weights --


shared NN with distance-based weights graph (dtype: float64, shape: (58683, 58683)) has 2861385 nonzero entries.
shared NN weighted graph data: [0.17000748 0.17032101 0.1705099  ... 0.22193401 0.24232573 0.39033699]

Converting graph (dtype: float64, shape: (58683, 58683)) has 2861385 nonzero entries.
---- Ran csr_to_igraph in 0.83 s ----


Resolution: 0.1
------------------------------

---- Partitioned BANKSY graph ----
modularity: 0.78
7 unique labels:
[0 1 2 3 4 5 6]

---- Ran partition in 35.56 s ----

No annotated labels
Decay type: nonspatial
Neighbourhood Contribution (Lambda Parameter): 0.0
reduced_pc_20 

reduced_pc_20_umap 

PCA dims to analyse: [20]

====================================================================================================
Setting up partitioner for (nbr decay = nonspatial), Neighbourhood contribution = 0.0, PCA dimensions = 20)
====================================================================================================


Nearest-neighbour weighted graph (dtype: float64, shape: (58683, 58683)) has 2934150 nonzero entries.
---- Ran find_nn in 74.41 s ----


Nearest-neighbour connectivity graph (dtype: int16, shape: (58683, 58683)) has 2934150 nonzero entries.

(after computing shared NN)
Allowing nearest neighbours only reduced the number of shared NN from 51503123 to 2930357.


Shared nearest-neighbour (connections only) graph (dtype: int16, shape: (58683, 58683)) has 2820200 nonzero entries.

Shared nearest-neighbour (number of shared neighbours as weights) graph (dtype: int16, shape: (58683, 58683)) has 2820200 nonzero entries.

sNN graph data:
[22 17 26 ...  9 15 12]

---- Ran shared_nn in 2.98 s ----


-- Multiplying sNN connectivity by weights --


shared NN with distance-based weights graph (dtype: float64, shape: (58683, 58683)) has 2820200 nonzero entries.
shared NN weighted graph data: [0.19304595 0.19326815 0.19418376 ... 0.2103373  0.21252941 0.25186446]

Converting graph (dtype: float64, shape: (58683, 58683)) has 2820200 nonzero entries.
---- Ran csr_to_igraph in 0.91 s ----


Resolution: 0.1
------------------------------

---- Partitioned BANKSY graph ----
modularity: 0.83
11 unique labels:
[ 0  1  2  3  4  5  6  7  8  9 10]

---- Ran partition in 24.09 s ----

No annotated labels

After sorting Dataframe
Shape of dataframe: (2, 7)

Maximum number of labels = 11
Indices of sorted list: [1 0]

Expanding labels with ids: [0 1 2 3 4 5 6] so that ids range from 0 to 10

Label ids zerod: [0 1 2 3 4 5 6].
0 to be inserted between each id: [0 0 0 0 0 0 0]
4 extra rows to be randomly inserted: [1 1 0 0 1 1]
New ids: [ 0  2  4  5  6  8 10]
---- Ran expand_labels in 0.05 s ----

---- Ran match_labels in 0.01 s ----


Matched Labels</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">decay</th>
<th data-quarto-table-cell-role="th">lambda_param</th>
<th data-quarto-table-cell-role="th">num_pcs</th>
<th data-quarto-table-cell-role="th">resolution</th>
<th data-quarto-table-cell-role="th">num_labels</th>
<th data-quarto-table-cell-role="th">labels</th>
<th data-quarto-table-cell-role="th">adata</th>
<th data-quarto-table-cell-role="th">relabeled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">nonspatial_pc20_nc0.00_r0.10</td>
<td>nonspatial</td>
<td>0.0</td>
<td>20</td>
<td>0.1</td>
<td>11</td>
<td>Label object:\nNumber of labels: 11, number of...</td>
<td>[[[View of AnnData object with n_obs × n_vars ...</td>
<td>Label object:\nNumber of labels: 11, number of...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">scaled_gaussian_pc20_nc0.60_r0.10</td>
<td>scaled_gaussian</td>
<td>0.6</td>
<td>20</td>
<td>0.1</td>
<td>7</td>
<td>Label object:\nNumber of labels: 7, number of ...</td>
<td>[[[View of AnnData object with n_obs × n_vars ...</td>
<td>Label object:\nNumber of labels: 7, number of ...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="plot-results" class="level3">
<h3 class="anchored" data-anchor-id="plot-results">Plot results</h3>
<div id="cell-126" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> banksy.plot_banksy <span class="im">import</span> plot_results</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>c_map <span class="op">=</span>  <span class="st">'tab20'</span> <span class="co"># specify color map</span></span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>weights_graph <span class="op">=</span>  banksy_dict[<span class="st">'scaled_gaussian'</span>][<span class="st">'weights'</span>][<span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-127" class="cell" data-execution_count="53">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>banksy_path <span class="op">=</span> <span class="ss">f'./outputs/banksy_output/'</span> </span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>plot_results(</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>    results_df,</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>    weights_graph,</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>    c_map,</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>    match_labels <span class="op">=</span> <span class="va">True</span>,</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>    coord_keys <span class="op">=</span> coord_keys,</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>    max_num_labels  <span class="op">=</span>  max_num_labels, </span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>    save_path <span class="op">=</span> os.path.join(banksy_path, <span class="st">'tmp_png'</span>),</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>    save_fig <span class="op">=</span> <span class="va">True</span>, <span class="co"># save the spatial map of all clusters</span></span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a>    save_seperate_fig <span class="op">=</span> <span class="va">True</span>, <span class="co"># save the figure of all clusters plotted seperately</span></span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Saving figure at ./outputs/banksy_output/tmp_png/slideseq_mousecerebellum_nonspatial_pc20_nc0.00_r0.10_spatialmap.png
number of labels: 11
---- Ran plot_2d_embeddings in 0.00 s ----

number of labels: 11
---- Ran plot_2d_embeddings in 0.01 s ----

number of labels: 11
---- Ran plot_2d_embeddings in 0.00 s ----

---- Ran row_normalize in 0.00 s ----


matrix multiplying labels x weights x labels-transpose ((11, 58683) x (58683, 58683) x (58683, 11))

Saving figure at ./outputs/banksy_output/tmp_png/slideseq_mousecerebellum_scaled_gaussian_pc20_nc0.60_r0.10_spatialmap.png
number of labels: 7
---- Ran plot_2d_embeddings in 0.00 s ----

number of labels: 7
---- Ran plot_2d_embeddings in 0.01 s ----

number of labels: 7
---- Ran plot_2d_embeddings in 0.00 s ----

---- Ran row_normalize in 0.00 s ----


matrix multiplying labels x weights x labels-transpose ((7, 58683) x (58683, 58683) x (58683, 7))
</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-73-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-73-output-3.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-73-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-73-output-5.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<!-- <div class="alert alert-block alert-danger">
    <b>Task:</b> Explore the robustness of BANSKY. 
    <ol>
        <li>What do you expect would happen when we increase <code>k_geom</code> (to 80)?</li>
        <li>What would you expect the <b>spatial clusters</b> and <b>cell type composition</b> to look like if we increase <code>lambda=0.99</code>?</li>
        <li>How do you expect the <b>spatial clusters</b> and <b>cell type composition</b> to change if <code>lambda</code> is decreasing? Set <code>lambda=0</code> </li>
    </ol>
</div>
 -->
</section>
<section id="investigate-cell-type-composition-in-each-banksy-defined-spatial-domain" class="level3">
<h3 class="anchored" data-anchor-id="investigate-cell-type-composition-in-each-banksy-defined-spatial-domain">Investigate Cell Type composition in each Banksy-defined Spatial Domain</h3>
<div id="cell-130" class="cell" data-execution_count="66">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_sd_vs_cell_type_composition(res_df,idx):</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Plots the cell type composition as a percentage across different SD (standard deviation) values.</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a><span class="co">    The data is visualized as a stacked bar plot.</span></span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Parameters:</span></span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a><span class="co">    - results_df_lambda05: DataFrame containing the data with columns 'labels_scaled_gaussian_pc20_nc0.50_r0.10', 'class', and others.</span></span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a><span class="co">    - idx: string of column of interest in the DataFrame</span></span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a><span class="co">    </span></span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb100-11"><a href="#cb100-11" aria-hidden="true" tabindex="-1"></a><span class="co">    - None</span></span>
<span id="cb100-12"><a href="#cb100-12" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb100-13"><a href="#cb100-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Step 1: Add a 'Count' column to facilitate pivoting (each row contributes a count of 1)</span></span>
<span id="cb100-14"><a href="#cb100-14" aria-hidden="true" tabindex="-1"></a>    res_df.obs[<span class="st">'Count'</span>] <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb100-15"><a href="#cb100-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-16"><a href="#cb100-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Step 2: Create a pivot table with SD as the index, cell types as columns, and the sum of counts as values</span></span>
<span id="cb100-17"><a href="#cb100-17" aria-hidden="true" tabindex="-1"></a>    pivot_df <span class="op">=</span> res_df.obs.pivot_table(</span>
<span id="cb100-18"><a href="#cb100-18" aria-hidden="true" tabindex="-1"></a>        index<span class="op">=</span>idx,  <span class="co"># Group by SD</span></span>
<span id="cb100-19"><a href="#cb100-19" aria-hidden="true" tabindex="-1"></a>        columns<span class="op">=</span><span class="st">'cell_types'</span>,  <span class="co"># Columns represent cell types</span></span>
<span id="cb100-20"><a href="#cb100-20" aria-hidden="true" tabindex="-1"></a>        values<span class="op">=</span><span class="st">'Count'</span>,  <span class="co"># Aggregate the 'Count' column</span></span>
<span id="cb100-21"><a href="#cb100-21" aria-hidden="true" tabindex="-1"></a>        aggfunc<span class="op">=</span><span class="st">'sum'</span>,  <span class="co"># Sum up counts for each combination</span></span>
<span id="cb100-22"><a href="#cb100-22" aria-hidden="true" tabindex="-1"></a>        fill_value<span class="op">=</span><span class="dv">0</span>  <span class="co"># Fill missing combinations with 0</span></span>
<span id="cb100-23"><a href="#cb100-23" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb100-24"><a href="#cb100-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-25"><a href="#cb100-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Ensure SD values are numeric</span></span>
<span id="cb100-26"><a href="#cb100-26" aria-hidden="true" tabindex="-1"></a>    pivot_df.index <span class="op">=</span> pivot_df.index.astype(<span class="bu">float</span>)</span>
<span id="cb100-27"><a href="#cb100-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-28"><a href="#cb100-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Step 3: Convert counts to percentages for each SD</span></span>
<span id="cb100-29"><a href="#cb100-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Divide each row by the row sum to get percentages, then multiply by 100</span></span>
<span id="cb100-30"><a href="#cb100-30" aria-hidden="true" tabindex="-1"></a>    pivot_df <span class="op">=</span> pivot_df.div(pivot_df.<span class="bu">sum</span>(axis<span class="op">=</span><span class="dv">1</span>), axis<span class="op">=</span><span class="dv">0</span>) <span class="op">*</span> <span class="dv">100</span></span>
<span id="cb100-31"><a href="#cb100-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-32"><a href="#cb100-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Step 4: Set up the plot</span></span>
<span id="cb100-33"><a href="#cb100-33" aria-hidden="true" tabindex="-1"></a>    fig, ax <span class="op">=</span> plt.subplots(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">6</span>))  <span class="co"># Define figure size</span></span>
<span id="cb100-34"><a href="#cb100-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-35"><a href="#cb100-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Plot stacked bars</span></span>
<span id="cb100-36"><a href="#cb100-36" aria-hidden="true" tabindex="-1"></a>    bottom <span class="op">=</span> <span class="va">None</span>  <span class="co"># Keeps track of the cumulative height of the bars</span></span>
<span id="cb100-37"><a href="#cb100-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> cell_type <span class="kw">in</span> pivot_df.columns:  <span class="co"># Loop through each cell type</span></span>
<span id="cb100-38"><a href="#cb100-38" aria-hidden="true" tabindex="-1"></a>        ax.bar(</span>
<span id="cb100-39"><a href="#cb100-39" aria-hidden="true" tabindex="-1"></a>            pivot_df.index,  <span class="co"># X-axis: SD values</span></span>
<span id="cb100-40"><a href="#cb100-40" aria-hidden="true" tabindex="-1"></a>            pivot_df[cell_type],  <span class="co"># Y-axis: Percentages for this cell type</span></span>
<span id="cb100-41"><a href="#cb100-41" aria-hidden="true" tabindex="-1"></a>            label<span class="op">=</span>cell_type,  <span class="co"># Legend label</span></span>
<span id="cb100-42"><a href="#cb100-42" aria-hidden="true" tabindex="-1"></a>            bottom<span class="op">=</span>bottom  <span class="co"># Stack on top of previous bars</span></span>
<span id="cb100-43"><a href="#cb100-43" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb100-44"><a href="#cb100-44" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Update 'bottom' to include the current cell type's values</span></span>
<span id="cb100-45"><a href="#cb100-45" aria-hidden="true" tabindex="-1"></a>        bottom <span class="op">=</span> pivot_df[cell_type] <span class="cf">if</span> bottom <span class="kw">is</span> <span class="va">None</span> <span class="cf">else</span> bottom <span class="op">+</span> pivot_df[cell_type]</span>
<span id="cb100-46"><a href="#cb100-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-47"><a href="#cb100-47" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Step 5: Add labels and title</span></span>
<span id="cb100-48"><a href="#cb100-48" aria-hidden="true" tabindex="-1"></a>    ax.set_xlabel(<span class="st">'SD'</span>)  <span class="co"># Label for the x-axis</span></span>
<span id="cb100-49"><a href="#cb100-49" aria-hidden="true" tabindex="-1"></a>    ax.set_ylabel(<span class="st">'Cell Type Composition (%)'</span>)  <span class="co"># Label for the y-axis</span></span>
<span id="cb100-50"><a href="#cb100-50" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="st">'SD vs. Cell Type Composition'</span>)  <span class="co"># Title of the plot</span></span>
<span id="cb100-51"><a href="#cb100-51" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="dv">0</span>, <span class="dv">100</span>)  <span class="co"># Set y-axis limits to [0, 100] to represent percentages</span></span>
<span id="cb100-52"><a href="#cb100-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-53"><a href="#cb100-53" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Add legend</span></span>
<span id="cb100-54"><a href="#cb100-54" aria-hidden="true" tabindex="-1"></a>    plt.legend(</span>
<span id="cb100-55"><a href="#cb100-55" aria-hidden="true" tabindex="-1"></a>        title<span class="op">=</span><span class="st">"Cell Type"</span>,  <span class="co"># Title of the legend</span></span>
<span id="cb100-56"><a href="#cb100-56" aria-hidden="true" tabindex="-1"></a>        bbox_to_anchor<span class="op">=</span>(<span class="fl">1.05</span>, <span class="dv">1</span>),  <span class="co"># Position the legend outside the plot</span></span>
<span id="cb100-57"><a href="#cb100-57" aria-hidden="true" tabindex="-1"></a>        loc<span class="op">=</span><span class="st">'upper left'</span>  <span class="co"># Align the legend at the upper left corner</span></span>
<span id="cb100-58"><a href="#cb100-58" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb100-59"><a href="#cb100-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-60"><a href="#cb100-60" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Adjust layout to prevent overlap</span></span>
<span id="cb100-61"><a href="#cb100-61" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb100-62"><a href="#cb100-62" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Step 6: Show the plot</span></span>
<span id="cb100-63"><a href="#cb100-63" aria-hidden="true" tabindex="-1"></a>    plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-131" class="cell" data-execution_count="67">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>results_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="67">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">decay</th>
<th data-quarto-table-cell-role="th">lambda_param</th>
<th data-quarto-table-cell-role="th">num_pcs</th>
<th data-quarto-table-cell-role="th">resolution</th>
<th data-quarto-table-cell-role="th">num_labels</th>
<th data-quarto-table-cell-role="th">labels</th>
<th data-quarto-table-cell-role="th">adata</th>
<th data-quarto-table-cell-role="th">relabeled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">nonspatial_pc20_nc0.00_r0.10</td>
<td>nonspatial</td>
<td>0.0</td>
<td>20</td>
<td>0.1</td>
<td>11</td>
<td>Label object:\nNumber of labels: 11, number of...</td>
<td>[[[View of AnnData object with n_obs × n_vars ...</td>
<td>Label object:\nNumber of labels: 11, number of...</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">scaled_gaussian_pc20_nc0.60_r0.10</td>
<td>scaled_gaussian</td>
<td>0.6</td>
<td>20</td>
<td>0.1</td>
<td>7</td>
<td>Label object:\nNumber of labels: 7, number of ...</td>
<td>[[[View of AnnData object with n_obs × n_vars ...</td>
<td>Label object:\nNumber of labels: 7, number of ...</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-132" class="cell" data-execution_count="68">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>results_df.loc[idx][<span class="st">'adata'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="68">
<pre><code>AnnData object with n_obs × n_vars = 58683 × 1041
    obs: 'cell_id', 'transcript_counts', 'control_probe_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'region', 'cell_labels', 'condition', 'time', 'batch_key', 'leiden_res0_25', 'leiden_res0_5', 'leiden_res1', 'cell_types', 'sample', 'x', 'y', 'labels_scaled_gaussian_pc20_nc0.60_r0.10', 'Count'
    var: 'gene_ids', 'feature_types', 'genome', 'is_nbr', 'k'
    obsm: 'reduced_pc_20', 'reduced_pc_20_umap', 'spatial'</code></pre>
</div>
</div>
<div id="cell-133" class="cell" data-execution_count="69">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>idx<span class="op">=</span><span class="st">'scaled_gaussian_pc20_nc0.60_r0.10'</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>results_df_lambda05 <span class="op">=</span> results_df.loc[idx][<span class="st">'adata'</span>]</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>label_idx <span class="op">=</span> <span class="ss">f'labels_</span><span class="sc">{</span>idx<span class="sc">}</span><span class="ss">'</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>plot_sd_vs_cell_type_composition(results_df_lambda05, label_idx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="niche_and_domain_reconstruction_files/figure-html/cell-77-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="compare-cellcharter-vs-bansky" class="level2">
<h2 class="anchored" data-anchor-id="compare-cellcharter-vs-bansky">Compare CellCharter vs BANSKY</h2>
<!-- <div class="alert alert-block alert-danger">
    <b>Task:</b> Explore similarities and difference between CellCharter and BANSKY.
    <ol>
        <li>What are common assumptions that both methods make or rely on?</li>
        <li>WWhat advantages or limitations can you find between the methods (e.i. runtime, visualization, spatial domains)?</li>
    </ol>
</div>
 -->
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<p>[1] Varrone, M., Tavernari, D., Santamaria-Martínez, A., Walsh, L. A. &amp; Ciriello, G. CellCharter reveals spatial cell niches associated with tissue remodeling and cell plasticity. Nat Genet 56, 74–84 (2024).</p>
<p>[2] Singhal, V. et al.&nbsp;BANKSY unifies cell typing and tissue domain segmentation for scalable spatial omics data analysis. Nat Genet 56, 431–441 (2024).</p>
<p>[3] https://github.com/BrainOmicsCourse/BrainOmics2024/tree/main/3_Day3. Last access: 18.12.2024</p>
<p>[4] https://github.com/NBISweden/workshop-spatial/blob/main/labs/07b_spatial_domains.ipynb: Last access: 14.01.2025</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/elixir-europe-training\.github\.io\/ELIXIR-SCO-spatial-omics\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>